---
title: "get-started"
author: "Thomas Rauter"
date: "20 September, 2024"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{get-started}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  markdown: 
    wrap: 72
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  dev = "ragg_png"
)

# Ensure ComplexHeatmap’s internal png() calls don’t use Quartz on macOS
if (capabilities("cairo")) options(bitmapType = "cairo")
```

# About this tutorial

This tutorial intends to showcase and explain the capabilities of the
**SplineOmics** package by walking through a real and complete example,
from start to finish.

### Example Overview

The example involves a **time-series proteomics experiment**, where CHO
(chinese hamster ovary) cells were cultivated in three bioreactors
(three biological replicates). The experiment includes the following
setup:

-   Samples were taken both during the **exponential** and **stationary
    growth phases**.
-   Samples were collected in triplicates from each reactor at defined
    timepoints relative to cell feeding:
    -   60 minutes before feeding
    -   15, 60, 90, 120, and 240 minutes after feeding

### Analysis Goals

The main goals of this analysis are:

-   **Identify proteins with significant temporal changes**: Out of 7162
    cellular proteins, the objective is to detect which proteins show a
    significant change over time after the CHO cells were fed (i.e., the
    impact of the feeding).
-   **Cluster hits based on temporal patterns**: The proteins (hits)
    with significant temporal changes will be clustered according to
    their time-based patterns.
-   **Perform gene set enrichment analysis**: For each cluster, a gene
    set enrichment analysis will be performed to determine if specific
    biological processes are up- or downregulated after feeding.

### Note

The documentation of all the **SplineOmics** package functions can be
viewed [here](https://csbg.github.io/SplineOmics/reference)

# Load the packages

```{r setup, eval = TRUE}
library(SplineOmics)
library(here) # For managing filepaths
library(dplyr) # For data manipulation
library(knitr) # For Showing the head of the data and the meta tables.
library(org.Mm.eg.db)    # BioConductor database
```

# Load the files

In this example, the proteomics_data.rds file contains the numeric
values (the intensities) and also the feature descriptions, such as gene
and protein name (= annotation part). Usually, you would load the data
from for example an Excel file, but the .rds file is more compressed,
which is the reason this format was chosen here to limit the size of the
SplineOmics package.

The file meta.xlsx contains the meta information, which are the
descriptions of the columns of the numeric values of data.

(These example files are part of the package and don't have to be
present on your system).

Please note that this dataset is an actual experimental dataset, but the
annotation information, such as gene names, has been removed since it
was not yet published at the time of making the SplineOmics package
public. Instead, the dataset includes randomly generated gene symbols
and gene names corresponding to Cricetulus griseus (Chinese Hamster) for
each row. This is intended to demonstrate the functionality of the
package. Further, the dataset was subsamples to 1165 proteins to limit the size
of this package.

The left part of data contains the numeric values, and the right part
the annotation info, which can be copied in a separate dataframe, as
shown below.

```{r load the files}
data <- readRDS(xzfile(system.file(
  "extdata",
  "proteomics_data.rds.xz",
  package = "SplineOmics"
)))

meta <- read.csv(
  system.file(
    "extdata",
    "proteomics_meta.csv",
    package = "SplineOmics"
  ),
  stringsAsFactors = FALSE
)

# Extract the annotation part from the dataframe.
first_na_col <- which(is.na(data[1, ]))[1]
annotation <- data |>
  dplyr::select((first_na_col + 1):ncol(data)) |>
  dplyr::slice(-c(1:3))
```

## Show top rows of data

```{r Show data head, eval = TRUE}
kable(
  head(data),
  format = "markdown"
  )
```

## Show top rows of meta

```{r Show meta head, eval = TRUE}
kable(
  head(meta),
  format = "markdown"
  )
```

## Show top rows of annotation

```{r Show annotation head, eval = TRUE}
kable(
  head(annotation),
  format = "markdown"
  )
```

Three comments about the characteristics the input data should have:

-   The data must not contain any NA values or other special values, and
    must consist only of numbers. For example, the original proteomics
    data contained some NA values, which were resolved in this case by
    imputation (replacing NA values with numbers).

-   All features of the data should ideally be normally distributed when
    analyzed with `limma`, which fits a linear model to each feature.
    These models rely on statistical tests that assume normality.
    Although `limma` can still function if the data is not normally
    distributed, the resulting p-values may become less reliable. For
    this reason, it is strongly recommended to transform your data using
    techniques such as log2 transformation when the features deviate
    from normality. Proper transformation helps ensure that the
    assumptions underlying the statistical tests are met, leading to
    more accurate and trustworthy results.

-   The samples in the data should be independent of each other. Linear
    models,\
    such as those used in limma, assume that the observations (samples)
    are\
    independent. If there is a dependency between samples (e.g.,
    repeated\
    measurements of the same subject), this assumption is violated,
    which can lead to incorrect statistical inferences.


## Bring the Inputs into the Standardized Format

Since `data` is not in the format required by the **SplineOmics**
package, it needs some processing. The SplineOmics package requires data
to be a numeric matrix, so no element is allowed to be anything else
than a number. This can be done with a few commands in R, but if your
file has a specific structure, the function `extract_data()` can handle
this more easily

### Usage of the extract_data() function

`extract_data()` can:

-   **Extract the data matrix field** by specifying the location of the corners 
    of the matrix.
-   **Create column headers** from the information written in the cells
    above the respective columns of the data matrix field.
-   **Assign rowheaders**:
    -   If no annotation columns are specified, rowheaders will be
        increasing numbers.
    -   If annotation columns are specified (like
        `"First.Protein.Description"` and `"ID"` in this example), these
        will be combined to form the rowheaders (feature names).

### Usage in Plotting

The generated rowheaders will be used to label any plots where a feature
is shown individually, such as:

-   **Spline plots** with the datapoints from an individual feature.

```{r process inputs, eval = TRUE}
data <- SplineOmics::extract_data(
  # The dataframe with the numbers on the left and info on the right.
  data = data,
  # Use this annotation column for the feature names.
  feature_name_columns = c("Gene_name"),
  use_row_index = TRUE,
  top_row = 4,
  bottom_row = 1165,
  right_col = 37,
  left_col = 2
)
```

# Perform EDA (exploratory data analysis)

Now that we have the data in the required format (numeric matrix) we can
go on.

The first step in analyzing data is typically **Exploratory Data
Analysis (EDA)**. EDA involves summarizing the main characteristics of
the data, often through visualizations.

### Common EDA Plots

Some common types of EDA plots include:

-   **Density distributions**
-   **Boxplots**
-   **PCA (Principal Component Analysis)**
-   **Correlation heatmaps**

Again, you can generate those plots yourself with a few lines of R code.
However, if you prefer, for convenience, the `explore_data()` function
can handle this for you.

### Using `explore_data()` for EDA

The **SplineOmics** package provides the function `explore_data()` to
perform EDA. This function requires the following arguments:

-   **data**: The numeric data matrix.
-   **meta**: The metadata table.
-   **condition**: The name of the column in the metadata that contains
    the levels of the experiment (e.g., "Exponential" and "Stationary").
-   **report_info**: A list that contains general information about the
    analysis, such as the name of the analyst and the datatype (e.g.
    proteomics)

### Optional Arguments

In addition to the required arguments, `explore_data()` offers several
optional arguments:

-   **meta_batch_column**: The name of the column that contains the
    first batch effect.

-   **meta_batch2_column**: The name of the column that contains the
    second batch effect.

    If at least one batch column is provided, the function will:

    -   Use the `removeBatchEffect()` function from **limma** to remove
        the batch effect from the data before plotting.
    -   Generate two EDA HTML reports: one for the **uncorrected data**
        and one for the **batch-corrected data**.

### Output and Report Options

-   By default, the reports are saved in the **current working
    directory**, but this location can be changed using the `report_dir`
    argument.
-   The function also **returns all plots** generated during the
    analysis, so that you can modify them according to your own needs.
-   If you do not want a report to be generated, you can set the
    `report` argument to `FALSE` (when you for example just want the
    figures in the R environment)

```{r Load EDA arguments, eval = TRUE}
# Those fields are mandatory, because we believe that when such a report is
# opened after half a year, those infos can be very helpful.
report_info <- list(
  omics_data_type = "PTX",
  data_description = "Proteomics data of CHO cells",
  data_collection_date = "February 2024",
  analyst_name = "Thomas Rauter",
  contact_info = "thomas.rauter@plus.ac.at",
  project_name = "DGTX"
)

report_dir <- here::here(
  "results",
  "explore_data"
)
```

## SplineOmics Object

In the SplineOmics package, multiple functions take the same arguments
as input. To make this easier and to avoid errors, we decided that those
arguments are not provided individually to the functions, but are all
stored in an R6 object (which is of type 'SplineOmics') and then this
object is passed to the functions. Additionally, some functions generate
intermediate output, which is just necessary for the next function in
the workflow, which is then also just passed along by updating the
SplineOmics object. But you don't have to worry about this.

### Functionality

The SplineOmics object can be seen as a container where all necessary
arguments are stored. Each function retrieves the required arguments
from the object and potentially adds new data or results back into it.

### Documentation

The documentation of the function that creates the SplineOmics object
can be found
[here](https://csbg.github.io/SplineOmics/reference/create_splineomics.html)
and the documentation of the function that updates it
[[here](https://csbg.github.io/SplineOmics/reference/update_splineomics.html)

The documentation for each function that takes the SplineOmics object as
input specifies which arguments must be present in the SplineOmics
object when it is passed to the respective function.

## Required Arguments `create_splineomics()`

-   **data**: A matrix with the data
-   **meta**: Metadata associated with the data.
-   **condition**: Meta column name of the levels (e.g., Exponential and
    Stationary).

## Optional Arguments `create_splineomics()`

-   **rna_seq_data**: An object containing the preprocessed RNA-seq
    data, such as the output from `limma::voom` function.
-   **annotation**: A dataframe with the feature descriptions of data.
-   **report_info**: A list containing general information about the
    analysis.
-   **meta_batch_column**: Column for meta batch information.
-   **meta_batch2_column**: Column for secondary meta batch information.
-   **design**: A limma design formula
-   **spline_params**: Parameters for the spline functions.

```{r Create the SplineOmics object, eval = TRUE}
# splineomics now contains the SplineOmics object.
splineomics <- SplineOmics::create_splineomics(
  data = data,
  meta = meta,
  annotation = annotation,
  report_info = report_info,
  condition = "Phase", # Column of meta that contains the levels.
  meta_batch_column = "Reactor" # For batch effect removal
)

# Special print.SplineOmics function leads to selective printing
print(splineomics)
```

Now that we have the SplineOmics object defined, we can perform our
exploratory data analysis.

```{r Run EDA function, eval = TRUE}
plots <- SplineOmics::explore_data(
  splineomics = splineomics, # SplineOmics object
  report_dir = withr::local_tempdir()
)
```

[Here](https://csbg.github.io/SplineOmics_html_reports/explore_data_PTX.html)
you can see the HTML report of the explore_data() function with the NOT
batch-corrected data, and
[here](https://csbg.github.io/SplineOmics_html_reports/explore_batch_corrected_data_PTX.html)
the report for the batch-corrected data.

Note that any report linked in this vignette might have been created with an
older version of SplineOmics, and not updated yet. 

The EDA plots can tell you a range of things. The plots in the HTML
report are grouped into three categories: Distribution and Variability
Analysis, Time Series Analysis, and Dimensionality Reduction and
Clustering.

If you look at the correlation heatmaps in the HTML report, you can see
that the samples E12_TP05_Exponential and E10_TP10_Stationary stick out.
Seeing this, you might want to remove them from the data. You can test
out what happens when you do this, along with testing how other
hyperparameter choices influence the results, with the package function
screen_limma_hyperparams().

# Run limma spline analysis

Now we decide on the parameters and run the limma spline analysis.

For the design formula, you must specify either 'isolated' or
'integrated'. Isolated means limma determines the results for each level
using only the data from that level. Integrated means limma determines
the results for all levels using the full dataset (from all levels). For
the integrated mode, the condition column (here Phase) must be included
in the design. Isolated means that limma only uses the part of the
dataset that belongs to a level to obtain the results for that level.

To generate the limma result categories 2 and 3 ()

```{r Update the SplineOmics object, eval = TRUE}
splineomics <- SplineOmics::update_splineomics(
  splineomics = splineomics,
  design = "~ 1 + Phase*Time + Reactor", # best design formula
  mode = "integrated", # means limma uses the full data for each condition.
  # States explicitly that there is no problem of heteroscedasticity and 
  # therefore, this does not need to be adressed. Setting it to TRUE would mean
  # the opposite, and when setting it to NULL, it means it should be handled 
  # implicitly. For details, see Reference
  # documentation of the create_splineomics() function.
  use_array_weights = FALSE,
  spline_params = list(
    spline_type = c("n"), # natural cubic splines (take these if unsure)
    dof = c(2L) # If you are unsure about which dof, start with 2 and increase
  )
)
```

Run the `run_limma_splines()` function with the updated SplineOmics
object:

```{r limma spline analysis, eval = TRUE}
splineomics <- SplineOmics::run_limma_splines(
  splineomics = splineomics
)
```

The output of the function run_limma_splines() is a named list, where
each element is a specific "category" of results. Refer to [this
document](https://csbg.github.io/SplineOmics/articles/limma_result_categories.html)
for an explanation of the different result categories. Each of those
elements is a list, containing as elements the respective limma
topTables, either for each level or each comparison between two levels.

The element "time_effect" is a list, where each element is the topTable
where the p-value for each feature for the respective level are
reported.

The element "avrg_diff_conditions" is a list that contains as elements
the topTables, that represent the comparison of the average differences
of the levels.

The element "interaction_condition_time" is a list that contains as
elements the topTables, that represent the interaction between the
levels (which includes both time and the average differences)

# Build limma report

The topTables of all three limma result categories can be used to
generate p-value histograms an volcano plots.

```{r build limma report, eval = TRUE}
plots <- SplineOmics::create_limma_report(
  splineomics = splineomics,
  report_dir = withr::local_tempdir()
)
```

You can view the generated analysis report of the create_limma_report
function
[here](https://csbg.github.io/SplineOmics_html_reports/create_limma_report_PTX.html).

This report contains p-value histograms for all three limma result
categories and a volcano plot for category 2. Embedded in this file are
the downloadable limma topTables with the results for category 1 if the
mode was 'isolated' and also with the results of category 2 and 3 if the
mode was 'integrated'.

Note that in the upcoming cluster_hits() function report, the embedded
file will only contain the clustered significant features from result
category 1.

# Cluster the hits (significant features)

After we obtained the limma spline results, we can cluster the hits
based on their temporal pattern (their spline shape). We define what a
hit is by setting an adj. p-value threshold for every level. Hits are
features (e.g. proteins) that have an adj. p-value below the threshold.
Hierarchical clustering is used to place every hit in one of as many
clusters as we have specified for that specific level.

```{r prepare cluster_hits() arguments, eval = TRUE}
# Important note: When you define parameters for the levels, always define them
# in the order those levels appear in the meta condition column! Otherwise, 
# there will be a mixup!

adj_pthresholds <- c( 
  0.05, 
  0.05 
)

# The amount of clusters can be a fixed number (e.g. 6) or a range. When you
# specify a range (e.g. 2:3, which corresponds to 2 3 in the vector) then the 
# cluster_hits() function tries all those cluster numbers and picks the one with
# the highest silhouette score (automatic cluster number identification). When
# you don't want to have a clustering for a level, write 1 for the cluster 
# number for that level.
nr_clusters <- list(
  Exponential = 6,   # specifically 6 clusters for the exponential phase level
  Stationary = 2:3   # range of cluster numbers for the stationary phase level
)

plot_info <- list( # For the spline plots
  y_axis_label = "log2 intensity",
  time_unit = "min", # our measurements were in minutes
  treatment_labels = list(
    Exponential = "feeding",
    Stationary = "feeding",
    double_spline_plots = "feeding"
    ),
  treatment_timepoints = list(
    Exponential = 0,
    Stationary = 0,
    double_spline_plots = 0
    )
)


# Get all the gene names. They are used for generating files
# which contents can be directly used as the input for the Enrichr webtool,
# if you prefer to manually perform the enrichment. Those files are
# embedded in the output HTML report and can be downloaded from there.
gene_column_name <- "Gene_symbol"
genes <- annotation[[gene_column_name]]

plot_options <- list(
  # When meta_replicate_column is not there, all datapoints are blue.
  meta_replicate_column = "Reactor", # Colors the data points based on Reactor
  cluster_heatmap_columns = FALSE # Per default FALSE, just for demonstration
)
```

Run the function:

```{r cluster the hits, eval = TRUE}
report_dir <- file.path(tempdir(), "splineomics_report")
dir.create(report_dir, showWarnings = FALSE, recursive = TRUE)

clustering_results <- SplineOmics::cluster_hits(
  splineomics = splineomics,
  adj_pthresholds = adj_pthresholds,
  nr_clusters = nr_clusters,
  genes = genes,
  plot_info = plot_info,
  plot_options = plot_options,
  report_dir = report_dir,
  adj_pthresh_avrg_diff_conditions = 0.05,
  adj_pthresh_interaction_condition_time = 0.05
)
```

You can view the generated analysis report of the cluster_hits function
[here](https://csbg.github.io/SplineOmics_html_reports/report_clustered_hits_PTX.html).

As discussed before, there are three limma result categories. The
cluster_hits() report shows the results of all three, if they are
present (category 2 and 3 can only be generated when the design formula
contains an interaction effect).

# Perform overrepresentation analysis (ORA)

Once the clustered hits are identified, a subsequent step to gain
biological insights is to perform ORA For this, the respective genes
can be assigned to each clustered hit, and ORA can be carried out. To
proceed, the Enrichr databases of choice need to be downloaded:

```{r download Enrichr databases, eval = FALSE}
# Create a temporary directory for R CMD check
results_dir <- file.path(tempdir(), "ora")
dir.create(report_dir, showWarnings = FALSE, recursive = TRUE)

# Specify which databases you want to download from Enrichr
gene_set_lib <- c(
  "WikiPathways_2019_Human",
  "NCI-Nature_2016",
  "TRRUST_Transcription_Factors_2019",
  "MSigDB_Hallmark_2020",
  "GO_Cellular_Component_2018",
  "CORUM",
  "KEGG_2019_Human",
  "TRANSFAC_and_JASPAR_PWMs",
  "ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X",
  "GO_Biological_Process_2018",
  "GO_Molecular_Function_2018",
  "Human_Gene_Atlas"
)

SplineOmics::download_enrichr_databases(
  gene_set_lib = gene_set_lib,
  output_dir = results_dir,
  filename = "databases.tsv"
)
```

Per default the file is placed in the current working directory, which
is the root dir of the R project.

To run ORA, the downloaded database file has to be loaded as a
dataframe. Further, optionally, the clusterProfiler parameters and the
report dir can be specified. The function run_ora() runs ORA
using clusterProfiler, generates an HTML report and returns the ORA
dotplots in R.

```{r prepare arguments for ORA, eval = FALSE}
# Specify the filepath of the TSV file with the database info
downloaded_dbs_filepath <- file.path(results_dir, "databases.tsv")

# Load the file
databases <- read.delim(
  downloaded_dbs_filepath,
  sep = "\t",
  stringsAsFactors = FALSE
)

# Specify the clusterProfiler parameters
clusterProfiler_params <- list(
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  minGSSize = 10,
  maxGSSize = 500,
  qvalueCutoff = 0.2
)
```

The function below runs the clusterProfiler for all clusters and all
levels, and generates the HTML report:

```{r run ORA, eval = FALSE}
result <- SplineOmics::run_ora(
  cluster_table = clustering_results[["cluster_table"]],
  databases = databases,
  clusterProfiler_params = clusterProfiler_params,
  report_info = report_info,
  cluster_hits_report_name = "report name",
  report_dir =  withr::local_tempdir()
)
```

You can view the generated analysis report of the run_ora function
[here](https://csbg.github.io/SplineOmics_html_reports/run_ora_report_PTX.html).

This report first shows all enrichment results, where more than 2 genes
supported a term, in a tabular format. The table with all the terms with
\< 2 genes supporting it can be downloaded by clicking on a button below
that table.

For the dotplots below that, every row is a term from a specific
database, and the columns are the respective clusters. The color scale
contains the info about the odds ratio and the size the -log10 adj.
p-value. Only terms that have \> 2 genes as support are included in the
plot. Further, for each cluster, just maximally 5 terms are shown (the
terms with the highest odds ratios). Note that when for example cluster
1 already has 5 terms, and cluster 2 does not, and gets a term which was
also found for cluster 1, than this term would be included as the sixth
term for cluster 1, so this is a way the maximum of 5 can be exceeded.

If a phase, like stationary here, does not lead to any enrichment
results, that is stated with a red message.

# Perform ORA with the Bioconductor database

The ORA can also be performed with any other database that is in the format
DB, Geneset, Gene (see documentation of the run_ora function). The BioConductor
databases are an example for that, and SplineOmics also contains a function
to conveniently download these.

```{r download BioConductor mouse database, eval = FALSE}
SplineOmics::download_bioconductor_database(
  organism_db = "org.Mm.eg.db",
  output_dir = results_dir,
  filename = "bioconductor_database.tsv"
)
```

```{r prepare arguments for ORA with BioConductor databases, eval = FALSE}
# Specify the filepath of the TSV file with the database info
downloaded_dbs_filepath <- file.path(
  results_dir,
  "bioconductor_database.tsv"
  )

# Load the file
databases <- read.delim(
  downloaded_dbs_filepath,
  sep = "\t",
  stringsAsFactors = FALSE
)

# Specify the clusterProfiler parameters
clusterProfiler_params <- list(
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  minGSSize = 10,
  maxGSSize = 500,
  qvalueCutoff = 0.2
)

mapping_cfg = list(
  method = "gprofiler",
  from_species = "cgchok1gshd",
  to_species   = "mmusculus"
  )

genes_clean <- genes[!grepl("-", genes)]

enrichGO_cfg = list(
  GO_BP = list(
    OrgDb = org.Mm.eg.db,
    keyType = "SYMBOL",
    ontology = "BP"
  ),
  GO_MF = list(
    OrgDb = org.Mm.eg.db,
    keyType = "SYMBOL",
    ontology = "MF"
  ),
  GO_CC = list(
    OrgDb = org.Mm.eg.db,
    keyType = "SYMBOL",
    ontology = "CC"
  )
)
```

```{r run ORA with BioConductor databases, eval = FALSE}
result <- SplineOmics::run_ora(
  cluster_table = clustering_results[["cluster_table"]],
  databases = databases,
  clusterProfiler_params = clusterProfiler_params,
  universe = genes_clean,
  mapping_cfg = mapping_cfg,
  enrichGO_cfg = enrichGO_cfg,
  report_info = report_info,
  cluster_hits_report_name = "report",
  report_dir = results_dir
)
```

# Conclusion

This example showed most functionalities of the SplineOmics package. You
can also run other datatypes with it, including timeseries RNA-seq and
glycan data (for those, refer to the documentation in the README file on
the GitHub page under Usage/RNA-seq and Glycan Data).

We hope that the SplineOmics package makes your scientific data analysis
easier. If you face any problems (bugs in the code) or are not satisfied
with the documentation, open an issue on GitHub or check out the other
options under the Feedback section of the README on GitHub. Thank you!

# Session Info

```{r sessionInfo, echo=FALSE}
sessionInfo()
```