---
title: "methylation-data-analysis"
author: "Thomas Rauter"
date: "11 July, 2025"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{methylation-data-analysis}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  markdown: 
    wrap: 72
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# About this vignette

This tutorial intends to showcase and explain the capabilities of the
**SplineOmics** package by walking through a real and complete methylation data
analysis example, from start to finish. **SplineOmics** is explained in more
detail in the **get-started** vignette, where a proteomics example is
covered. This vignette is more focused on showing how methylation data can
be used, and because of this, less details about the overall package are
provided here. Further, this vignette is more minimalistic than the other ones,
because the same functionalities than in the other vignettes are used and the
explanations are not repeated here.

## Data Overview

This dataset originates from a **time-series methylation data experiment**
designed to study Chinese Hamster Ovary (CHO) cells. The experiment
involved cultivating cells in **eight bioreactors**, with **four
bioreactors** subjected to a temperature shift after 146 hours
(experimental condition) and the remaining **four bioreactors**
maintained without a temperature shift (control condition).

### Timepoints

Samples were collected at **19 distinct time points** throughout the
experiment, specifically: `"48h"`, `"72h"`, `"96h"`, `"120h"`, `"144h"`,
`"168h"`, `"192h"`, `"216h"`, `"240h"`, `"264h"`, `"268h"`, `"288h"`, `"312h"`,
`"336h"`, `"360h"`, `"384h"`, `"408h"`, `"432h"`, `"456h"`, after cultivation
start. Most time point were sampled from all eight bioreactors, resulting
in a total of **130 samples**.

### Additonal effects in the experiment: Reactor and Group

In this experiment, there are two effects to consider: **Reactor** and **Group**.

1. **Reactor**:
   - This refers to the different bioreactors used for cell cultivation, which can exhibit substantial variability.
   - Each reactor was assigned a single condition: either constant temperature or temperature-shifted. As a result, reactor is a blocked effect.
   - Reactor should not be treated as a fixed effect to simply remove its influence. Instead, it is treated as a **random effect**, which allows us to model its variability appropriately.

2. **Group**:
   - This refers to the two separate groups in which the bioreactors where run.
     Because the apparatus containes 4 bioreactors, only that amount can be run
     simultaneosly. The other 4 were run after the initial 4.
   - Group is considered a **batch effect** with respect to the condition
   (constant temperature vs. temperature-shifted).

Since reactor is a blocked effect, the variability due to the reactors cannot be directly separated from the condition. Instead, **linear mixed models (LMMs)** are used to attribute the reactor as a random effect, allowing us to account for its variability while isolating the effects of the condition. This approach ensures that the analysis appropriately handles the hierarchical structure of the data and avoids incorrect conclusions.


### Note

The documentation of all the **SplineOmics** package functions can be
viewed [here](https://csbg.github.io/SplineOmics/reference)

# Load the packages

```{r setup, eval = TRUE}
library(SplineOmics)
library(readr) # For reading the meta CSV file
library(here)  # For managing filepaths
library(dplyr) # For data manipulation
library(knitr) # For Showing the head of the data and the meta tables.
```

# Load the files

```{r load the files, eval = TRUE}
data_df <- readRDS(xzfile(system.file(
  "extdata",
  "methylation_data.rds.xz",
  package = "SplineOmics"
)))

meta <- readr::read_csv(
  system.file(
    "extdata",
    "methylation_meta.csv",
    package = "SplineOmics"
  ),
  show_col_types = FALSE
)
meta$Condition[meta$Condition == "Temp. Shifted"] <- "Temp_shift"
```

## Show top rows of data

```{r Show data head, eval = TRUE}
kable(
  head(data),
  format = "markdown"
  )
```

## Show top rows of meta

```{r Show meta head, eval = TRUE}
kable(
  head(meta),
  format = "markdown"
  )
```


```{r extract data matrix, eval = TRUE}
data <- SplineOmics::extract_data(
  # The dataframe with the numbers on the left and info on the right.
  data = data_df,
  # Use this annotation column for the feature names.
  feature_name_columns = c("feature_names"),
  top_row = 1,
  bottom_row = 300,
  right_col = 131,
  left_col = 2
)
```

# Perform EDA (exploratory data analysis)

```{r Define EDA arguments, eval = TRUE}
report_info <- list(
  omics_data_type = "EPI",
  data_description = "Methylation data of CHO cells",
  data_collection_date = "December 2024",
  analyst_name = "Thomas Rauter",
  contact_info = "thomas.rauter@plus.ac.at",
  project_name = "DGTX"
)

report_dir <- here::here(
  "results",
  "explore_data"
)
```

```{r Create the SplineOmics object, eval = TRUE}
splineomics <- SplineOmics::create_splineomics(
  data = data,
  meta = meta,
  report_info = report_info,
  condition = "Condition",           # Column of meta that contains the levels.
  meta_batch_column = "Group"        # Remove batch effect for plotting.
)
```

```{r Run EDA function, eval = TRUE}
plots <- SplineOmics::explore_data(
  splineomics = splineomics, 
  report_dir = withr::local_tempdir()
)
```


# Run limma spline analysis

In this example, we are skipping finding the best hyperparameters with
the screen_limma_hyperparams() function, because we already have a clear
idea of what do use.

Lets define our parameters and put them into the `SplineOmics` object:

```{r Update the SplineOmics object, eval = TRUE}
spline_params <- list(
  spline_type = c("n"),    # natural cubic splines
  dof = c(3L)              # Degree of freedom of 3 for the splines.
)

design <- "~ 1 + Condition*Time + Group + (1|Reactor)"

splineomics <- SplineOmics::update_splineomics(
  splineomics = splineomics,
  data = data,
  design = design, 
  mode = "integrated", 
  spline_params = spline_params,
  bp_cfg = c(       # For parallel computing for variancePartition::dream (lmm)
    n_cores = 1,        # For the vignette, just use one core.    
    blas_threads = 1
    )
)
```

Run the `run_limma_splines()` function:

```{r limma spline analysis, eval = TRUE}
splineomics <- SplineOmics::run_limma_splines(
  splineomics = splineomics
)
```

# Build limma report

The topTables of all three limma result categories can be used to
generate p-value histograms an volcano plots.

```{r build limma report, eval = TRUE}
plots <- SplineOmics::create_limma_report(
  splineomics = splineomics,
  report_dir = withr::local_tempdir()
)
```

# Cluster the hits (significant features)

```{r cluster the hits, eval = TRUE}
adj_pthresholds <- c(   
  0.05,   # constant (temperature)   
  0.05    # temp_shift
)

nr_clusters <- list(
  Constant   = 2:6,                    
  Temp_shift = 2:6   
)

report_dir <- here::here(
  "results",
  "clustering_reports"
)

plot_info <- list( # For the spline plots
  y_axis_label = "beta value",
  time_unit = "hours", 
  treatment_labels = list(
    Temp_shift = "temp shift",
    double_spline_plots = "temp shift"
    ),
  treatment_timepoints = list(
    Temp_shift = 146,
    double_spline_plots = 146
    )
)

genes <- rownames(data)

plot_options <- list(
  # When meta_replicate_column is not there, all datapoints are blue.
  meta_replicate_column = "Reactor" # Colors the data points based on Reactor
)

clustering_results <- SplineOmics::cluster_hits(
  splineomics = splineomics,
  adj_pthresholds = adj_pthresholds,
  nr_clusters = nr_clusters,
  genes = genes,
  plot_info = plot_info,
  plot_options = plot_options,
  report_dir = withr::local_tempdir(),
  adj_pthresh_avrg_diff_conditions = 0.05,
  adj_pthresh_interaction_condition_time = 0.05
)
```

# Session Info

```{r sessionInfo, echo=FALSE}
sessionInfo()
```