---
title: "RNA-seq-analysis"
author: "Thomas Rauter"
date: "29 January, 2025"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{RNA-seq-analysis}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  markdown: 
    wrap: 72
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  dev = "ragg_png"
)

# Ensure ComplexHeatmap’s internal png() calls don’t use Quartz on macOS
if (capabilities("cairo")) options(bitmapType = "cairo")
```

# About this vignette

This tutorial intends to showcase and explain the capabilities of the
**SplineOmics** package by walking through a real and complete RNA-seq
example, from start to finish. **SplineOmics** is explained in more
detail in the **get-started** vignette, where a proteomics example is
covered. This vignette is more focused on showing how RNA-seq data can
be used, and because of this, less details about the overall package are
provided here.

## Data Overview

This dataset originates from a **time-series RNA-seq experiment**
designed to study Chinese Hamster Ovary (CHO) cells. The experiment
involved cultivating cells in **eight bioreactors**, with **four
bioreactors** subjected to a temperature shift after 146 hours
(experimental condition) and the remaining **four bioreactors**
maintained without a temperature shift (control condition).

### Timepoints

Samples were collected at **17 distinct time points** throughout the
experiment, specifically: `"72h"`, `"76h"`, `"96h"`, `"120h"`, `"124h"`,
`"144h"`, `"148h"`, `"152h"`, `"168h"`, `"192h"`, `"216h"`, `"220h"`,
`"240h"`, `"264h"`, `"268h"`, `"288h"`, and `"312h"` after cultivation
start. Each time point was sampled from all eight bioreactors, resulting
in a total of **136 samples**.

### Additonal effects in the experiment: Reactor and Plate

In this experiment, there are two effects to consider: **Reactor** and **Plate**.

1. **Reactor**:
   - This refers to the different bioreactors used for cell cultivation, which can exhibit substantial variability.
   - Each reactor was assigned a single condition: either constant temperature or temperature-shifted. As a result, **condition and reactor are confounded**.
   - Reactor should not be treated as a fixed effect to simply remove its influence. Instead, it is treated as a **random effect**, which allows us to model its variability appropriately.

2. **Plate**:
   - This refers to the two separate plates used during the RNA-seq analysis of the samples.
   - A fully random design was employed to distribute samples across the two plates, ensuring no bias in their assignment.
   - Plate is considered a **batch effect** with respect to the condition (constant temperature vs. temperature-shifted).

Since reactor is a blocked effect, the variability due to the reactors cannot be directly separated from the condition. Instead, **linear mixed models (LMMs)** are used to attribute the reactor as a random effect, allowing us to account for its variability while isolating the effects of the condition. This approach ensures that the analysis appropriately handles the hierarchical structure of the data and avoids incorrect conclusions.

In this vignette, we will demonstrate how to use linear mixed models to address these challenges and properly account for both reactor and plate effects.

### Further info

The data matrix comprises **genes as rows** and **samples as columns**,
providing gene expression measurements for all time points. Each sample
was initially sequenced in **three technical replicates** across **two
NovaSeq X flow cells**. These technical replicates were collapsed to
generate the final dataset used in this analysis.

The goal of this experiment is to investigate the effect of a
temperature shift during CHO cell cultivation on gene expression
dynamics over time.

Note: This is not the original dataset, as it has not yet been published
at the time of this vignette's creation. For demonstration purposes, the
genes have been randomly shuffled, and only a subset of the data has
been included to reduce the dataset size.

## Analysis Goals

The main objectives of this analysis are:

-   **Identify genes with significant temporal changes**: Among the
    thousands of genes measured, the goal is to identify those that
    exhibit significant changes in expression over time.

-   **Cluster genes based on temporal patterns**: Genes showing
    significant temporal changes (hits) will be grouped into clusters
    based on their time-dependent expression patterns.

-   **Perform gene set enrichment analysis**: For each cluster, a gene
    set enrichment analysis will be conducted to identify whether
    specific biological pathways or processes are up- or downregulated
    in response to feeding and how these processes are influenced by the
    temperature shift.

-   **Assess the impact of temperature shifts on temporal patterns**:
    The analysis will determine whether the temporal patterns of gene
    expression are affected by the temperature shift, i.e., whether gene
    expression dynamics differ over time under temperature shift
    conditions compared to controls.

### Note

The documentation of all the **SplineOmics** package functions can be
viewed [here](https://csbg.github.io/SplineOmics/reference)

# Load the packages

```{r setup, eval = TRUE}
library(SplineOmics)
library(readr) # For reading the meta CSV file
library(here)  # For managing filepaths
library(dplyr) # For data manipulation
library(knitr) # For Showing the head of the data and the meta tables.
```

# Load the files

```{r load the files, eval = TRUE}
data <- readRDS(xzfile(system.file(
  "extdata",
  "rna_seq_data.rds.xz",
  package = "SplineOmics"
)))

meta <- readr::read_csv(
  system.file(
    "extdata",
    "rna_seq_meta.csv",
    package = "SplineOmics"
  ),
  show_col_types = FALSE
)
```

## Show top rows of data

```{r Show data head, eval = TRUE}
kable(
  head(data),
  format = "markdown"
  )
```

**data**: A numeric matrix where each row represents a gene (features)
and each column corresponds to a sample. The row names of the matrix
contain the gene identifiers, while the columns are aligned with the
sample metadata in meta. The matrix contains expression values for 136
samples. 

Note that the study was conducted in a blinded manner, with samples randomly distributed across two plates for RNA-seq analysis. As a result, the sample numbers (e.g., 1, 2, 3, etc.) are not in sequential order with respect to time, condition, or plate.

For the data analysis involving splines over time, it is essential to sort the samples based on time to establish a valid temporal sequence. Additionally, organizing the data in this way improves clarity and ensures consistency. Within each time point, samples were further sorted by condition (e.g., constant before `temp_shift`) and, subsequently, by plate (e.g., `plate_1` before `plate_2`).


## Show top rows of meta

```{r Show meta head, eval = TRUE}
kable(
  head(meta),
  format = "markdown"
  )
```

**meta**: A data frame containing metadata information about the samples
in data. Each row in meta corresponds to a column in data, ensuring a
1:1 alignment between metadata entries and expression data samples. The
columns in meta describe various attributes of the samples, such as
SampleNr, Reactor, Time, Condition, and Plate.

# Preprocess the data

Filter out data rows (genes) with zero counts across all samples. This step is a standard preprocessing procedure in RNA-seq data analysis, as genes with zero 
counts in all samples provide no information for downstream analyses.

```{r Filter data rows, eval = TRUE}
rows_before <- nrow(data)

# Filter data rows
data <- data[rowSums(data) > 0, ]

rows_after <- nrow(data)
rows_removed <- rows_before - rows_after

cat(sprintf(
  "Rows before filtering: %d\nRows after filtering: %d\nRows removed: %d\n", 
  rows_before,
  rows_after,
  rows_removed
  ))
```

# Perform EDA (exploratory data analysis)

```{r Define EDA arguments, eval = TRUE}
report_info <- list(
  omics_data_type = "RNA",
  data_description = "RNA-seq data of CHO cells",
  data_collection_date = "December 2024",
  analyst_name = "Thomas Rauter",
  contact_info = "thomas.rauter@plus.ac.at",
  project_name = "DGTX"
)

report_dir <- here::here(
  "results",
  "explore_data"
)
```

```{r Create the SplineOmics object, eval = TRUE}
splineomics <- SplineOmics::create_splineomics(
  data = data,
  meta = meta,
  report_info = report_info,
  condition = "Condition",    # Column of meta that contains the levels.
  meta_batch_column = "Plate" # Remove batch effect for plotting.
)
```

```{r Run EDA function, eval = FALSE}
plots <- SplineOmics::explore_data(
  splineomics = splineomics, 
  report_dir = report_dir
)
```

[Here](https://drive.google.com/uc?export=view&id=1mE7FFXFO9ZZ907in3M5Cf1w7yRxRFHiy)
you can see the HTML report of the explore_data() function with the NOT
batch-corrected data, and
[here](https://drive.google.com/uc?export=view&id=1g_Lt4kO4H79GnVh_f7KfXqIzxfTnMrmq)
the report for the batch-corrected data.

# Run limma spline analysis

In this example, we are skipping finding the best hyperparameters with
the screen_limma_hyperparams() function, because we already have a clear
idea of what do use.

Lets define our parameters and put them into the `SplineOmics` object:

```{r Update the SplineOmics object, eval = TRUE}
spline_params <- list(
  spline_type = c("n"),    # natural cubic splines
  dof = c(3L)              # Degree of freedom of 3 for the splines.
)

design <- "~ 1 + Condition*Time + Plate + (1|Reactor)"

splineomics <- SplineOmics::update_splineomics(
  splineomics = splineomics,
  data = data,
  design = design, 
  # dream_params = dream_params,   if we would want to adjust dream()
  # means limma "borrows" statistical power from the other levels
  mode = "integrated", 
  spline_params = spline_params,
  bp_cfg = c(
    n_cores = 1,
    blas_threads = 1
    )    
)
```

Preprocess the RNA-seq data with `limma::voom`. This method transforms raw RNA-seq 
counts into log-counts per million (logCPM) while modeling the mean-variance 
relationship. It assigns precision weights to each observation, ensuring 
accurate linear modeling of RNA-seq data, which often exhibits 
heteroscedasticity (varying variance across expression levels). To normalize 
the data, TMM (Trimmed Mean of M-values) normalization is applied using a 
`DGEList` object from the **edgeR** package, correcting for library size 
differences and compositional biases.

When random effects are included in the design, the function automatically 
uses `voomWithDreamWeights` from the **variancePartition** package instead. 
This method extends `voom` by incorporating random effects into the model, 
allowing for precise handling of complex experimental designs such as repeated 
measures or hierarchical structures. The calculated weights account for both 
fixed and random effects, providing robust results for differential expression 
analysis.

```{r Preprocess RNA-seq data with voom, eval = TRUE}
splineomics <- SplineOmics::preprocess_rna_seq_data(
  splineomics = splineomics
)
```

You can customize the normalization method by providing a specific normalization function through the `normalize_func` argument in the `preprocess_rna_seq_data()` function. For details on how to use this feature, please refer to the function documentation available under 'References' on the website.

Additionally, the use of `preprocess_rna_seq_data()` is optional for your RNA-seq data. Alternatively, you can use the `limma::voom` function directly and pass the resulting `voom` object to the `rna_seq_data` argument of `create_splineomics()` or `update_splineomics()`. Alongside this, you must pass the `$E` data matrix as the `data` argument.

In general, as long as the `data` argument contains the actual data matrix and the `rna_seq_data` argument contains an object compatible with `limma`, your data will be correctly processed.

Run the `run_limma_splines()` function:

```{r limma spline analysis, eval = TRUE}
splineomics <- SplineOmics::run_limma_splines(
  splineomics = splineomics
)
```

The output of the function run_limma_splines() is a named list, where
each element is a specific "category" of results. Refer to [this
document](https://csbg.github.io/SplineOmics/articles/limma_result_categories.html)
for an explanation of the different result categories. Each of those
elements is a list, containing as elements the respective limma
topTables, either for each level or each comparison between two levels.

The element "time_effect" is a list, where each element is the topTable
where the p-value for each feature for the respective level are
reported.

The element "avrg_diff_conditions" is a list that contains as elements
the topTables, that represent the comparison of the average differences
of the levels.

The element "interaction_condition_time" is a list that contains as
elements the topTables, that represent the interaction between the
levels (which includes both time and the average differences)

# Build limma report

The topTables of all three limma result categories can be used to
generate p-value histograms an volcano plots.

```{r build limma report, eval = TRUE}
plots <- SplineOmics::create_limma_report(
  splineomics = splineomics,
  report_dir = withr::local_tempdir()
)
```

You can view the generated analysis report of the create_limma_report
function
[here](https://drive.google.com/uc?export=view&id=1hvKloiaHQA5CO4Y6io8qC2lOpBJ54ypJ).

# Cluster the hits (significant features)

After we obtained the limma spline results, we can cluster the hits
based on their temporal pattern (their spline shape). We define what a
hit is by setting an adj. p-value threshold for every level. Hits are
features (genes here) that have an adj. p-value below the threshold.
Hierarchical clustering is used to place every hit in one of as many
clusters as we have specified for that specific level.

```{r cluster the hits, eval = TRUE}
adj_pthresholds <- c(   
  0.05,    
  0.05     
)

nr_clusters <- list(
  constant = 4, 
  temp_shift = 4   
)

report_dir <- here::here(
  "results",
  "clustering_reports"
)

# treatment_labels allows to place vertical dashed lines into the plots, that
# indicate a treatment, such as "temp shift" in this experiment. For each level,
# the treatment can be specified individually. For the "double spline plots", 
# where two levels are combined into one plot, treatment lines can also be 
# defined. The correct fieldname for those is {first_level}_{second_level}, 
# with first_level being the level of the two occuring first in the respective
# meta column, and second_level the one that occurs after. 
# Here, we don't want a treatment line for the constant level, since no 
# treatment was applied. To achieve that, we simply set it to NA or leave it
# out. 
treatment_labels <- list(
  # constant = NA, 
  temp_shift = "temp shift",  
  constant_temp_shift = "temp shift" 
  )

# treatment_timepoints allows to specify the timepoint, at which the vertical
# dashed treatment line is placed. Again, for the constant level, we don't
# have a treatment, so we also do not specify a timepoint.
treatment_timepoints <- list(
  # constant = NA,
  temp_shift = 146,  
  constant_temp_shift = 146
  )

plot_info <- list( # For the spline plots
  y_axis_label = "log2 counts",
  time_unit = "hours", 
  # When you simply want to add a given treatment line to all conditions (
  # including all double spline plot comparisons) then you can do it in the 
  # following way (commented out here): 
  # treatment_labels = list("temp shift"), # add for all conditions
  # treatment_timepoints = list(146)       # temp shift was at 146 hours.
  treatment_labels = treatment_labels,
  treatment_timepoints = treatment_timepoints
)

genes <- rownames(data)
genes <- sub("_\\d+$", "", genes)   # remove the _1 part from the end

plot_options <- list(
  # When meta_replicate_column is not there, all datapoints are blue.
  meta_replicate_column = "Reactor" # Colors the data points based on Reactor
)

clustering_results <- SplineOmics::cluster_hits(
  splineomics = splineomics,
  adj_pthresholds = adj_pthresholds,
  nr_clusters = nr_clusters,
  genes = genes,
  plot_info = plot_info,
  report = TRUE,
  max_hit_number = 20,          # maximum number of plots per result category
  plot_options = plot_options,
  report_dir = withr::local_tempdir(),
  adj_pthresh_avrg_diff_conditions = 0.05,
  adj_pthresh_interaction_condition_time = 0.05
)
```

You can view the generated analysis report of the cluster_hits function
[here](https://drive.google.com/uc?export=view&id=1EvPiEwHyGnBQFxRets00s87tochgyGK4).

As discussed before, there are three limma result categories. The
cluster_hits() report shows the results of all three, if they are
present (category 2 and 3 can only be generated when the design formula
contains an interaction effect).

# Perform gene set enrichment analysis (GSEA)

Once the clustered hits are identified, a subsequent step to gain biological 
insights is to perform GSEA. For this, the respective genes can be assigned to 
each clustered hit, and GSEA can be carried out. To proceed, the Enrichr 
databases of choice need to be downloaded:

```{r download Enrichr databases, eval = FALSE}
# Specify which databases you want to download from Enrichr
gene_set_lib <- c(
  "WikiPathways_2019_Human",
  "NCI-Nature_2016",
  "TRRUST_Transcription_Factors_2019",
  "MSigDB_Hallmark_2020",
  "GO_Cellular_Component_2018",
  "CORUM",
  "KEGG_2019_Human",
  "TRANSFAC_and_JASPAR_PWMs",
  "ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X",
  "GO_Biological_Process_2018",
  "GO_Molecular_Function_2018",
  "Human_Gene_Atlas"
)

SplineOmics::download_enrichr_databases(
  gene_set_lib = gene_set_lib,
  output_dir = here::here(),   # output into the current working dir (default)
  filename = "databases.tsv"   # just the name of the file, not the full path
)
```

Per default the file is placed in the current working directory, which
is the root dir of the R project.

To run GSEA, the downloaded database file has to be loaded as a
dataframe. Further, optionally, the clusterProfiler parameters and the
report dir can be specified. The function create_gsea_report() runs GSEA
using clusterProfiler, generates an HTML report and returns the GSEA
dotplots in R.

```{r prepare arguments for GSEA, eval = FALSE}
# Specify the filepath of the TSV file with the database info
downloaded_dbs_filepath <- here::here(
  "databases.tsv"
  )

# Load the file
databases <- read.delim(
  downloaded_dbs_filepath,
  sep = "\t",
  stringsAsFactors = FALSE
)

# Specify the clusterProfiler parameters
clusterProfiler_params <- list(
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  minGSSize = 10,
  maxGSSize = 500,
  qvalueCutoff = 0.2
)

report_dir <- here::here(
  "results",
  "gsea_reports"
)
```

The function below runs the clusterProfiler for all clusters and all
levels, and generates the HTML report:

```{r run GSEA, eval = FALSE}
result <- SplineOmics::run_gsea(
  # A dataframe with three columns: feature, cluster, and gene. Feature contains
  # the integer index of the feature, cluster the integer specifying the cluster
  # number, and gene the string of the gene, such as "CLSTN2".
  cluster_table = clustering_results[["cluster_table"]],
  databases = databases,
  clusterProfiler_params = clusterProfiler_params,
  report_info = report_info,
  cluster_hits_report_name = "not meaningful here",
  report_dir = report_dir
)
```

You can view the generated analysis report of the run_gsea function
[here](https://drive.google.com/uc?export=view&id=1EvPiEwHyGnBQFxRets00s87tochgyGK4).

# Session Info

```{r sessionInfo, echo=FALSE}
sessionInfo()
```