---
title: "usage"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{usage}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(SplineOmics)
```

The `SplineOmics` package provides comprehensive tools for analyzing time-course data, focusing on differential expression analysis and clustering. This section covers detailed instructions for utilizing each of the package's main functions.

#### limma_hyperparams_screen {#limma_hyperparams_screen}

Conducts a hyperparameter screening using the `limma` package to identify optimal parameters for differential expression analysis across multiple datasets.

##### Arguments

-   **datas**: A list of expression matrices (data frames or matrices), each representing a distinct dataset.

-   **datas_descr**: A character vector describing each dataset in `datas`. Example: `c("full_data", "outliers_removed")`.

-   **metas**: A list of metadata data frames corresponding to each dataset in `datas`. Each metadata frame should contain experimental conditions and batch information.

-   **designs**: A character vector of model design formulas for differential expression analysis.

-   **modes**: A string indicating the mode of analysis for each design. Example: `"integrated"`.

-   **condition**: The column name of the dataframe meta that contains the levels of the factor of the experiment. A factor is a variable that categorizes the data into different groups that are important for the analysis. A level is a specific value a factor can take.

-   **spline_test_configs**: A list of configurations for testing different spline parameters, including degrees of freedom and spline types.

-   **feature_names**: A character vector of feature names (e.g., gene or protein names).

-   **report_info**: Information that gets printed on top of the report, providing context or additional details about the analysis.

-   **report_dir**: Directory path for saving the output reports. Default is the current working dir.

-   **pthresholds**: A numeric vector of p-value thresholds for determining statistical significance.

-   **meta_batch_column**: A character string specifying the column in `meta` used to remove batch effects. Default is `NA`, which will not remove the batch effect with the help of the removeBatchEffect function from the `limma` package. The batch effect removed data would only be used to generate the plots, because in the limma analysis, limma accounts for the batch effect.

##### Example Usage

``` r
result <- limma_hyperparams_screen(
  datas = datas,
  datas_descr = datas_descr,
  metas = metas,
  designs = designs,
  modes = modes,
  condition = condition,
  spline_test_configs = spline_test_configs,
  feature_names = feature_names,
  report_info = report_info,
  report_dir = report_dir,
  pthresholds = pthresholds,
  meta_batch_column = meta_batch_column
)
```

#### Finding the right parameters for the limma spline analysis

Above you have seen that you can fit splines to your timeseries omics data and determine which features are significantly changed over the timecourse. However, should you use B-splines or natural cubic splines for your dataset, and with which degree of freedom? Should you remove outliers of your dataset? Which limma design formula should you use? Here, we call these decisions hyperparameters. We believe it is not straightforward answering these questions without testing their impact on the actual analysis. That is why the `SplineOmics` package contains a function to streamline the exploration of the different combination of these hyperparameters. With that function, you can very quickly and systematically explore which effect your hyperparameter choices have on your analysis. For example, you are not sure if you should use a degree of freedom for natural cubic splines of 2, 3, 4, or 5, and whether to remove the two potential outliers of your dataset, or just one, or just the other. This already results in a lot of different combinations of your decisions. With the function `limma_hyperparams_screen()` you can just easily plug in all the desired degrees of freedoms and version of your data (outliers removed vs. not removed, etc.) and see the impact in a series of automatically generated HTML reports.


### run_limma_splines {#run_limma_splines}

This function performs differential expression analysis tailored for time-course data using spline models via the `limma` package. It's designed to handle complex experimental designs and time as a continuous variable.

#### Arguments

-   **data**: Expression matrix (either a data frame or matrix) representing gene expression levels where rows correspond to features (e.g., genes, proteins) and columns correspond to samples.

-   **meta**: Metadata frame corresponding to the samples in the `data` matrix. This should include at least the experimental condition and the batch information if any batch effect removal is intended.

-   **design**: A character string specifying the model design formula for the differential expression analysis. The design should incorporate the experimental conditions and may also include interaction terms.

-   **condition**: The name of the experimental condition factor within the `meta` data frame. This specifies the main variable of interest for differential expression analysis.

-   **feature_names**: A character vector containing the names of the features (e.g., gene names, protein IDs) for annotation purposes.

-   **mode**: A character string specifying the mode of analysis, which can be either "integrated" for a comprehensive model including all factors or "isolated" to focus on specific factors or interactions.

-   **spline_params**: A list specifying the spline parameters, including `spline_type` (type of spline) and `DoFs` (degrees of freedom).

-   **padjust_method**: A character string specifying the method for p-value adjustment. Default is "BH" (Benjamini-Hochberg).

#### Example Usage

``` r
result <- run_limma_splines(
  data = data, 
  meta = meta, 
  design = design, 
  spline_params = spline_params,
  condition = condition, 
  feature_names = feature_names, 
  mode = "integrated",
  padjust_method = "BH"
)
```

### cluster_hits {#cluster_hits}

Clusters significant features based on their expression profiles, following differential expression analysis. This function helps identify groups of features (e.g., genes, proteins) that exhibit similar expression patterns across different conditions or time points.

#### Arguments

-   **top_tables**: A list containing the top tables from limma. Each top table should be a data frame with significant features identified by the analysis, including their p-values and other statistics.

-   **data**: An matrix (data frame or matrix) with rows representing features and columns representing samples.

-   **meta**: Metadata frame corresponding to the samples in the `data` matrix. This frame should include variables that describe the experimental conditions, batches, or other grouping factors relevant to the analysis.

-   **condition**: A character string specifying the column in `meta` that contains the experimental conditions or groupings used for the analysis.

-   **report_info**: Information that gets printed on top of the report, providing context or additional details about the analysis.

-   **mode**: A character string specifying the mode of the analysis. Default is `"integrated"`, depending on the limma design.

-   **spline_params**: A list specifying the spline parameters, including `spline_type` (type of spline) and `DoFs` (degrees of freedom).

-   **p_values**: A numeric vector specifying the p-value thresholds for features to be included in the clustering analysis. Only features with p-values below these thresholds in the differential expression analysis will be considered.

-   **clusters**: An integer vector or `"auto"` specifying the number of clusters to generate. This allows for different levels of granularity in the clustering analysis, depending on the desired number of feature groups.

-   **meta_batch_column**: A character string specifying the column in `meta` used to remove batch effects. Default is `NA`.

-   **time_unit**: A character string specifying the time unit for the plot labels. Default is `"m"` (minutes).

-   **report_dir**: The directory path where clustering results, including any plots or summary tables, will be saved. This allows for easy access and review of the clustering outputs.

#### Example Usage

``` r
clustering_results <- cluster_hits(
  top_tables = top_tables, 
  data = data, 
  meta = meta, 
  condition = condition, 
  spline_params = spline_params,
  mode = "integrated",
  p_values = p_values,
  clusters = clusters,
  report_info = report_info,
  meta_batch_column = meta_batch_column,
  report_dir = report_dir
)
```
