<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Thomas Rauter" />


<title>SplineOmics-Tutorial</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">SplineOmics-Tutorial</h1>
<h4 class="author">Thomas Rauter</h4>
<h4 class="date">10 June, 2024</h4>



<div id="about-this-tutorial" class="section level1">
<h1>About this tutorial</h1>
<p>This tutorial intends to showcase and explain the capabilities of the
SplineOmics package by walking through a real and complete example, from
start to the end. The example is a time-series proteomics experiment, in
which CHO cells were cultivated in three bioreactors (three biological
replicates). Both in the exponential and stationary growth phase,
samples were taken from each reactor in triplicates at defined
timepoints relative to a feeding of the cells (60 min before feeding,
and 15, 60, 90, 120, and 240 min after). The goal of this analysis is to
find out which of the 7162 cellular proteins show a significant change
over time after the CHO cells were fed. Further, the hits (proteins with
a significant change over time) are clustered based on their temporal
pattern, and a gene set enrichment analysis is performed with each
cluster to see if there are processes which are up- or downregulated
over time after the feeding.</p>
</div>
<div id="load-the-packages" class="section level1">
<h1>Load the packages</h1>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(SplineOmics)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="co">#&gt; </span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">library</span>(readxl)</span></code></pre></div>
</div>
<div id="load-the-files" class="section level1">
<h1>Load the files</h1>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;data.xlsx&quot;</span>, <span class="at">package =</span> <span class="st">&quot;SplineOmics&quot;</span>))</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>annotation <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;annotation.xlsx&quot;</span>, <span class="at">package =</span> <span class="st">&quot;SplineOmics&quot;</span>))</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>meta <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;meta.xlsx&quot;</span>, <span class="at">package =</span> <span class="st">&quot;SplineOmics&quot;</span>))</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="fu">print</span>(data)</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="co">#&gt; # A tibble: 7,162 × 36</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="co">#&gt;    E09_TP01_Exponential E10_TP01_Exponential E12_TP01_Exponential</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="co">#&gt;                   &lt;dbl&gt;                &lt;dbl&gt;                &lt;dbl&gt;</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="co">#&gt;  1                 15.3                 15.2                 15.3</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="co">#&gt;  2                 14.5                 14.7                 14.6</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a><span class="co">#&gt;  3                 16.0                 16.1                 16.1</span></span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a><span class="co">#&gt;  4                 16.2                 16.1                 16.2</span></span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a><span class="co">#&gt;  5                 15.1                 15.1                 15.2</span></span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a><span class="co">#&gt;  6                 14.6                 14.7                 14.6</span></span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a><span class="co">#&gt;  7                 15.0                 15.0                 15.1</span></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a><span class="co">#&gt;  8                 14.8                 14.7                 14.7</span></span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a><span class="co">#&gt;  9                 20.5                 20.6                 20.7</span></span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a><span class="co">#&gt; 10                 16.4                 16.4                 16.4</span></span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a><span class="co">#&gt; # ℹ 7,152 more rows</span></span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a><span class="co">#&gt; # ℹ 33 more variables: E09_TP02_Exponential &lt;dbl&gt;, E10_TP02_Exponential &lt;dbl&gt;,</span></span>
<span id="cb3-21"><a href="#cb3-21" tabindex="-1"></a><span class="co">#&gt; #   E12_TP02_Exponential &lt;dbl&gt;, E09_TP03_Exponential &lt;dbl&gt;,</span></span>
<span id="cb3-22"><a href="#cb3-22" tabindex="-1"></a><span class="co">#&gt; #   E10_TP03_Exponential &lt;dbl&gt;, E12_TP03_Exponential &lt;dbl&gt;,</span></span>
<span id="cb3-23"><a href="#cb3-23" tabindex="-1"></a><span class="co">#&gt; #   E09_TP04_Exponential &lt;dbl&gt;, E10_TP04_Exponential &lt;dbl&gt;,</span></span>
<span id="cb3-24"><a href="#cb3-24" tabindex="-1"></a><span class="co">#&gt; #   E12_TP04_Exponential &lt;dbl&gt;, E09_TP05_Exponential &lt;dbl&gt;,</span></span>
<span id="cb3-25"><a href="#cb3-25" tabindex="-1"></a><span class="co">#&gt; #   E10_TP05_Exponential &lt;dbl&gt;, E12_TP05_Exponential &lt;dbl&gt;, …</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">print</span>(annotation)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="co">#&gt; # A tibble: 7,162 × 6</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="co">#&gt;    `Potential contaminant` Protein.Group         Protein.Ids Protein.Names Genes</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co">#&gt;    &lt;lgl&gt;                   &lt;chr&gt;                 &lt;chr&gt;       &lt;chr&gt;         &lt;chr&gt;</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co">#&gt;  1 NA                      A0A8C2MV03;A0A8C2N2L3 A0A8C2MV03… A0A8C2MV03_C… Abhd4</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="co">#&gt;  2 NA                      A0A8C2N4G8            A0A8C2N4G8  A0A8C2N4G8_C… Gbe1 </span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="co">#&gt;  3 NA                      G3GR83                G3GR83      G3GR83_CRIGR  Akr1…</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="co">#&gt;  4 NA                      A0A098KX95            A0A098KX95  A0A098KX95_C… Agpa…</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="co">#&gt;  5 NA                      G3HGQ9                G3HGQ9      G3HGQ9_CRIGR  Agpa…</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a><span class="co">#&gt;  6 NA                      A0A8C2MM23            A0A8C2MM23  A0A8C2MM23_C… Agpa…</span></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a><span class="co">#&gt;  7 NA                      G3H8X2                G3H8X2      G3H8X2_CRIGR  Agpa…</span></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a><span class="co">#&gt;  8 NA                      A0A8C2MXW6            A0A8C2MXW6… A0A8C2MXW6_C… &lt;NA&gt; </span></span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a><span class="co">#&gt;  9 NA                      A0A061HUZ2            A0A061HUZ2  A0A061HUZ2_C… H671…</span></span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a><span class="co">#&gt; 10 NA                      A0A8C2MC97            A0A8C2MC97  A0A8C2MC97_C… Plcb3</span></span>
<span id="cb4-15"><a href="#cb4-15" tabindex="-1"></a><span class="co">#&gt; # ℹ 7,152 more rows</span></span>
<span id="cb4-16"><a href="#cb4-16" tabindex="-1"></a><span class="co">#&gt; # ℹ 1 more variable: First.Protein.Description &lt;chr&gt;</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">print</span>(meta)</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="co">#&gt; # A tibble: 36 × 5</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="co">#&gt;    Sample.ID            Reactor Time.Point Phase        Time</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="co">#&gt;    &lt;chr&gt;                &lt;chr&gt;   &lt;chr&gt;      &lt;chr&gt;       &lt;dbl&gt;</span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="co">#&gt;  1 E09_TP01_Exponential E09     TP01       Exponential   -60</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="co">#&gt;  2 E10_TP01_Exponential E10     TP01       Exponential   -60</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="co">#&gt;  3 E12_TP01_Exponential E12     TP01       Exponential   -60</span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="co">#&gt;  4 E09_TP02_Exponential E09     TP02       Exponential    15</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="co">#&gt;  5 E10_TP02_Exponential E10     TP02       Exponential    15</span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="co">#&gt;  6 E12_TP02_Exponential E12     TP02       Exponential    15</span></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="co">#&gt;  7 E09_TP03_Exponential E09     TP03       Exponential    60</span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="co">#&gt;  8 E10_TP03_Exponential E10     TP03       Exponential    60</span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a><span class="co">#&gt;  9 E12_TP03_Exponential E12     TP03       Exponential    60</span></span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a><span class="co">#&gt; 10 E09_TP04_Exponential E09     TP04       Exponential    90</span></span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a><span class="co">#&gt; # ℹ 26 more rows</span></span></code></pre></div>
</div>
<div id="perform-eda-exploratory-data-analysis" class="section level1">
<h1>Perform EDA (exploratory data analysis)</h1>
<p>The first step in analysing data is usually EDA. EDA involves
summarizing the main characteristics of the data, often using plots.
This can involve densitiy distributions, boxplots, PCA, correlation
heatmaps, and more. This process can be carried out by using the package
function explore_data(). The neccessary arguments are the data matrix
(data) the meta tab</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>report_info <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>  <span class="at">omics_data_type =</span> <span class="st">&quot;PTX&quot;</span>,</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>  <span class="at">data_description =</span> <span class="st">&quot;Proteomics data of CHO cells&quot;</span>,</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>  <span class="at">data_collection_date =</span> <span class="st">&quot;February 2024&quot;</span>,</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>  <span class="at">analyst_name =</span> <span class="st">&quot;Thomas Rauter&quot;</span>,</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>  <span class="at">contact_info =</span> <span class="st">&quot;thomas.rauter@plus.ac.at&quot;</span>,</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>  <span class="at">project_name =</span> <span class="st">&quot;DGTX&quot;</span>)</span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a>condition <span class="ot">&lt;-</span> <span class="st">&quot;Phase&quot;</span></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a>meta_batch_column <span class="ot">&lt;-</span> <span class="st">&quot;Reactor&quot;</span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a>report_dir <span class="ot">&lt;-</span> here<span class="sc">::</span><span class="fu">here</span>(<span class="st">&quot;results&quot;</span>, <span class="st">&quot;explore_data&quot;</span>)</span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a>plots <span class="ot">&lt;-</span> <span class="fu">explore_data</span>(<span class="at">data =</span> data,</span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a>                      <span class="at">meta =</span> meta,</span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a>                      <span class="at">condition =</span> condition,</span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a>                      <span class="at">report_info =</span> report_info,</span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a>                      <span class="at">meta_batch_column =</span> meta_batch_column,</span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a>                      <span class="at">report_dir =</span> report_dir)</span>
<span id="cb6-19"><a href="#cb6-19" tabindex="-1"></a><span class="co">#&gt; The time values do not change consistently after the replicates.</span></span>
<span id="cb6-20"><a href="#cb6-20" tabindex="-1"></a><span class="co">#&gt; Column &#39;Reactor&#39; of meta will be used to remove the batch effect for the plotting</span></span>
<span id="cb6-21"><a href="#cb6-21" tabindex="-1"></a><span class="co">#&gt; Making t-SNE plot</span></span>
<span id="cb6-22"><a href="#cb6-22" tabindex="-1"></a><span class="co">#&gt; Performing PCA</span></span>
<span id="cb6-23"><a href="#cb6-23" tabindex="-1"></a><span class="co">#&gt; Read the 36 x 36 data matrix successfully!</span></span>
<span id="cb6-24"><a href="#cb6-24" tabindex="-1"></a><span class="co">#&gt; OpenMP is working. 1 threads.</span></span>
<span id="cb6-25"><a href="#cb6-25" tabindex="-1"></a><span class="co">#&gt; Using no_dims = 2, perplexity = 11.000000, and theta = 0.500000</span></span>
<span id="cb6-26"><a href="#cb6-26" tabindex="-1"></a><span class="co">#&gt; Computing input similarities...</span></span>
<span id="cb6-27"><a href="#cb6-27" tabindex="-1"></a><span class="co">#&gt; Building tree...</span></span>
<span id="cb6-28"><a href="#cb6-28" tabindex="-1"></a><span class="co">#&gt; Done in 0.00 seconds (sparsity = 0.966049)!</span></span>
<span id="cb6-29"><a href="#cb6-29" tabindex="-1"></a><span class="co">#&gt; Learning embedding...</span></span>
<span id="cb6-30"><a href="#cb6-30" tabindex="-1"></a><span class="co">#&gt; Iteration 50: error is 50.965655 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-31"><a href="#cb6-31" tabindex="-1"></a><span class="co">#&gt; Iteration 100: error is 46.219279 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-32"><a href="#cb6-32" tabindex="-1"></a><span class="co">#&gt; Iteration 150: error is 46.250133 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-33"><a href="#cb6-33" tabindex="-1"></a><span class="co">#&gt; Iteration 200: error is 46.878648 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-34"><a href="#cb6-34" tabindex="-1"></a><span class="co">#&gt; Iteration 250: error is 52.125917 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-35"><a href="#cb6-35" tabindex="-1"></a><span class="co">#&gt; Iteration 300: error is 0.569038 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-36"><a href="#cb6-36" tabindex="-1"></a><span class="co">#&gt; Iteration 350: error is 0.328449 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-37"><a href="#cb6-37" tabindex="-1"></a><span class="co">#&gt; Iteration 400: error is 0.237332 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-38"><a href="#cb6-38" tabindex="-1"></a><span class="co">#&gt; Iteration 450: error is 0.052689 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-39"><a href="#cb6-39" tabindex="-1"></a><span class="co">#&gt; Iteration 500: error is 0.044719 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-40"><a href="#cb6-40" tabindex="-1"></a><span class="co">#&gt; Iteration 550: error is 0.044287 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-41"><a href="#cb6-41" tabindex="-1"></a><span class="co">#&gt; Iteration 600: error is 0.046771 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-42"><a href="#cb6-42" tabindex="-1"></a><span class="co">#&gt; Iteration 650: error is 0.042558 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-43"><a href="#cb6-43" tabindex="-1"></a><span class="co">#&gt; Iteration 700: error is 0.046026 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-44"><a href="#cb6-44" tabindex="-1"></a><span class="co">#&gt; Iteration 750: error is 0.045373 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-45"><a href="#cb6-45" tabindex="-1"></a><span class="co">#&gt; Iteration 800: error is 0.044091 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-46"><a href="#cb6-46" tabindex="-1"></a><span class="co">#&gt; Iteration 850: error is 0.046046 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-47"><a href="#cb6-47" tabindex="-1"></a><span class="co">#&gt; Iteration 900: error is 0.046724 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-48"><a href="#cb6-48" tabindex="-1"></a><span class="co">#&gt; Iteration 950: error is 0.046508 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-49"><a href="#cb6-49" tabindex="-1"></a><span class="co">#&gt; Iteration 1000: error is 0.043288 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-50"><a href="#cb6-50" tabindex="-1"></a><span class="co">#&gt; Fitting performed in 0.02 seconds.</span></span>
<span id="cb6-51"><a href="#cb6-51" tabindex="-1"></a><span class="co">#&gt; Report written to: /home/thomas/Documents/PhD/projects/DGTX/R_packages/SplineOmics/results/explore_data/explore_data_PTX_10_06_2024-11_10_25.html</span></span>
<span id="cb6-52"><a href="#cb6-52" tabindex="-1"></a><span class="co">#&gt; Making t-SNE plot</span></span>
<span id="cb6-53"><a href="#cb6-53" tabindex="-1"></a><span class="co">#&gt; Performing PCA</span></span>
<span id="cb6-54"><a href="#cb6-54" tabindex="-1"></a><span class="co">#&gt; Read the 36 x 36 data matrix successfully!</span></span>
<span id="cb6-55"><a href="#cb6-55" tabindex="-1"></a><span class="co">#&gt; OpenMP is working. 1 threads.</span></span>
<span id="cb6-56"><a href="#cb6-56" tabindex="-1"></a><span class="co">#&gt; Using no_dims = 2, perplexity = 11.000000, and theta = 0.500000</span></span>
<span id="cb6-57"><a href="#cb6-57" tabindex="-1"></a><span class="co">#&gt; Computing input similarities...</span></span>
<span id="cb6-58"><a href="#cb6-58" tabindex="-1"></a><span class="co">#&gt; Building tree...</span></span>
<span id="cb6-59"><a href="#cb6-59" tabindex="-1"></a><span class="co">#&gt; Done in 0.00 seconds (sparsity = 0.966049)!</span></span>
<span id="cb6-60"><a href="#cb6-60" tabindex="-1"></a><span class="co">#&gt; Learning embedding...</span></span>
<span id="cb6-61"><a href="#cb6-61" tabindex="-1"></a><span class="co">#&gt; Iteration 50: error is 42.251727 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-62"><a href="#cb6-62" tabindex="-1"></a><span class="co">#&gt; Iteration 100: error is 46.952509 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-63"><a href="#cb6-63" tabindex="-1"></a><span class="co">#&gt; Iteration 150: error is 47.522671 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-64"><a href="#cb6-64" tabindex="-1"></a><span class="co">#&gt; Iteration 200: error is 51.248928 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-65"><a href="#cb6-65" tabindex="-1"></a><span class="co">#&gt; Iteration 250: error is 43.987036 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-66"><a href="#cb6-66" tabindex="-1"></a><span class="co">#&gt; Iteration 300: error is 0.933656 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-67"><a href="#cb6-67" tabindex="-1"></a><span class="co">#&gt; Iteration 350: error is 0.473882 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-68"><a href="#cb6-68" tabindex="-1"></a><span class="co">#&gt; Iteration 400: error is 0.289384 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-69"><a href="#cb6-69" tabindex="-1"></a><span class="co">#&gt; Iteration 450: error is 0.193837 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-70"><a href="#cb6-70" tabindex="-1"></a><span class="co">#&gt; Iteration 500: error is 0.065314 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-71"><a href="#cb6-71" tabindex="-1"></a><span class="co">#&gt; Iteration 550: error is 0.062003 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-72"><a href="#cb6-72" tabindex="-1"></a><span class="co">#&gt; Iteration 600: error is 0.062270 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-73"><a href="#cb6-73" tabindex="-1"></a><span class="co">#&gt; Iteration 650: error is 0.061813 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-74"><a href="#cb6-74" tabindex="-1"></a><span class="co">#&gt; Iteration 700: error is 0.062312 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-75"><a href="#cb6-75" tabindex="-1"></a><span class="co">#&gt; Iteration 750: error is 0.062613 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-76"><a href="#cb6-76" tabindex="-1"></a><span class="co">#&gt; Iteration 800: error is 0.062771 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-77"><a href="#cb6-77" tabindex="-1"></a><span class="co">#&gt; Iteration 850: error is 0.064004 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-78"><a href="#cb6-78" tabindex="-1"></a><span class="co">#&gt; Iteration 900: error is 0.063588 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-79"><a href="#cb6-79" tabindex="-1"></a><span class="co">#&gt; Iteration 950: error is 0.063160 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-80"><a href="#cb6-80" tabindex="-1"></a><span class="co">#&gt; Iteration 1000: error is 0.064223 (50 iterations in 0.00 seconds)</span></span>
<span id="cb6-81"><a href="#cb6-81" tabindex="-1"></a><span class="co">#&gt; Fitting performed in 0.02 seconds.</span></span>
<span id="cb6-82"><a href="#cb6-82" tabindex="-1"></a><span class="co">#&gt; Report written to: /home/thomas/Documents/PhD/projects/DGTX/R_packages/SplineOmics/results/explore_data/explore_batch_corrected_data_PTX_10_06_2024-11_10_25.html</span></span></code></pre></div>
<p>You can view the detailed analysis report <a href="../inst/reports/report.html">here</a>.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
