---
title: "demo"
author: "Thomas Rauter"
date: "15 July, 2024"
output: html_document
editor_options: 
  markdown: 
    wrap: 72
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# About this tutorial

This tutorial demonstrates the capabilities of the SplineOmics package through 
a comprehensive example: a time-series proteomics experiment involving CHO cells
cultivated in three bioreactors (biological replicates). Samples were collected
from each reactor in triplicates at specific time points relative to cell 
feeding (60 min before, and 15, 60, 90, 120, and 240 min after feeding) during
both exponential and stationary growth phases.

The objective is to identify which of the 7162 cellular proteins show 
significant changes over time post-feeding. Proteins with significant temporal
changes are then clustered based on their patterns. A gene set enrichment 
analysis is performed for each cluster to identify processes that are up- or 
downregulated over time after feeding.

Note: For a better understanding of the SplineOmics functions, the required and
optional arguments are documented here. These however are only short forms of 
the full documentation of the arguments, which you can find by selecting a 
function and pressing F2.

# Load the packages

```{r Load the required packages}
library(SplineOmics)   # Functions are marked with SplineOmics::

# Additional packages needed to prepare SplineOmics function inputs
library(readxl)
library(here)
library(readr)
```

To avoid conflicts between functions from the dplyr package and base R 
functions, it is recommended to explicitly state your preferences using the 
conflicted package. This ensures that the intended function is used, preventing
potential errors and improving code clarity.

```{r Load dplyr package}
library(dplyr)
library(conflicted)

# Explicitly state preference of functions
conflicted::conflict_prefer("filter", "dplyr")
conflicted::conflict_prefer("lag", "dplyr")
conflicted::conflict_prefer("intersect", "base")
conflicted::conflict_prefer("setdiff", "base")
conflicted::conflict_prefer("setequal", "base")
conflicted::conflict_prefer("union", "base")
```

# Load the files

In this example, the data.xlsx file contains numeric values (intensities) and
feature descriptions, such as gene and protein names (annotation part). The
meta.xlsx file contains meta information, which describes the columns of the 
numeric values in data.xlsx.

These example files are included in the package and do not need to be present
on your system. For your analysis, create file paths using the here library 
instead of system.file.

```{r Load the files}
data_excel <- readxl::read_excel(
    system.file(
      "extdata",
      "proteomics_data.xlsx",
      package = "SplineOmics"
      )
    )

meta <- readxl::read_excel(
  system.file(
    "extdata",
    "proteomics_meta.xlsx",
    package = "SplineOmics"
    )
  )

# Extract the annotation part from the dataframe.
first_na_col <- which(is.na(data_excel[1,]))[1]
annotation <- data_excel |>
  dplyr::select((first_na_col + 1):ncol(data_excel)) |>
  dplyr::slice(-c(1:3))
```

# Bring the inputs into the standardized (required) format

Since `data_excel` is not in the format required by the SplineOmics package, it
needs processing. This can be done with a few R commands, but if your file looks
like the one here, with the data matrix on the left and annotation info on the
right, separated by an empty column, the `extract_data()` function can handle 
this automatically.

The function identifies the data matrix and converts it into a dataframe. Column
headers are created from the information in the cells above each data matrix 
column. If no annotation columns are specified, row headers are simply 
increasing numbers. In this example, the annotation columns 
"First.Protein.Description" and "ID" are specified to form the row headers
(feature names). These names will be used to label any plots where a feature 
is shown individually, such as spline plots with datapoints from an individual
feature.

## Required Arguments `extract_data()`

- **data**: A dataframe loaded from a tabular file.

## Optional Arguments `extract_data()`

- **feature_name_columns**: A character vector specifying the columns of the                                  dataframe `data` that should be used to construct the                             feature names. If omitted, the feature names are just                             numbers (stored as characters) starting from 1 (1, 2,                             3, etc.).

```{r Process the inputs}
data <- SplineOmics::extract_data(
  data = data_excel,
  feature_name_columns = c(   # Feature names will be a combo out of these col
    "First.Protein.Description",
    "ID"
    )
  )
```

# Perform EDA (exploratory data analysis)

The first step in analyzing data is usually EDA. EDA involves summarizing the
main characteristics of the data, often using plots such as density 
distributions, boxplots, PCA, and correlation heatmaps. This process can be 
carried out using the package function `explore_data()`.

These batch columns are used to run the `removeBatchEffect` function of limma
to remove the batch effect from the data for plotting. When at least one batch
column is provided, the function generates two EDA HTML reports: one for the 
uncorrected data and one for the batch corrected data. 

### Report Generation

The reports are written to the current working directory by default or to a 
specified location using the optional argument `report_dir`. The function also
returns all generated plots. If no report should be generated, set the optional
argument `report` to `FALSE`.

```{r Define info that is written in all the HTML reports}
report_info <- list(
  omics_data_type = "PTX",
  data_description = "Proteomics data of CHO cells",
  data_collection_date = "February 2024",
  analyst_name = "Thomas Rauter",
  contact_info = "thomas.rauter@plus.ac.at",
  project_name = "DGTX"
  )
```

## SplineOmics Object

The SplineOmics object is used because multiple functions in the package take 
the same inputs, and some functions generate "intermediate" output that is used
by subsequent functions in the workflow. 

### Functionality

The SplineOmics object can be seen as a container where all necessary arguments
are stored. Each function retrieves the required arguments from the object and 
potentially adds new data or results back into it.

### Documentation

The documentation for each function specifies which arguments must be present
in the SplineOmics object when it is passed to the respective function.

## Required Arguments `create_splineomics()`

- **data**: A matrix with the data
- **meta**: Metadata associated with the data.
- **condition**: Column name of the levels (e.g., Exponential and Stationary).

## Optional Arguments `create_splineomics()`

- **annotation**: A dataframe with the feature descriptions of data.
- **report_info**: A list containing general information about the analysis.
- **meta_batch_column**: Column for meta batch information.
- **meta_batch2_column**: Column for secondary meta batch information.
- **design**: A limma design formula
- **spline_params**: Parameters for the spline functions.

```{r Create the SplineOmics object}
splineomics <- SplineOmics::create_splineomics(
  data = data,
  meta = meta,
  annotation = annotation,
  report_info = report_info,
  condition = "Phase",
  meta_batch_column = "Reactor"
)
```

## Required Arguments `explore_data()`

- **splineomics**: A SplineOmics object containing the following fields:
  - **data**: A matrix with the data.
  - **meta**: Metadata associated with the data.
  - **condition**: Column name of the levels (e.g., Exponential and Stationary).
  - **report_info**: A list containing general information about the analysis.
  - **meta_batch_column**: Column for meta batch information.
  - **meta_batch2_column**: Column for secondary meta batch information.

## Optional Arguments `explore_data()`

- **report_dir**: The path to the output directory. Default is current work dir.
- **report**: A Boolean TRUE or FALSE value specifying if a report should be generated. Default is TRUE.


```{r Run the EDA function}
report_dir <- here::here(
  "demo_results",
  "explore_data"
  )

plots <- SplineOmics::explore_data(
  splineomics = splineomics,
  report_dir = report_dir
  )

```

The EDA plots can reveal a range of insights. In the HTML report, the plots are
grouped into three categories: distribution and variability analysis, time 
series analysis, and dimensional reduction and clustering.

### Correlation Heatmaps

If you examine the correlation heatmaps in the HTML report, you might notice 
that the samples `E12_TP05_Exponential` and `E10_TP10_Stationary` stand out. 
Based on this observation, you might decide to remove these samples from the 
data. The impact of such decicions can be explored with the 
screen_limma_hyperparams() function.

# Find the best hyperparameters

## Determining the Best Hyperparameters

Before running the limma spline analysis, we need to determine the best 
"hyperparameters." Hyperparameters in this context include the degree of 
freedom, different versions of the data (e.g., outlier removed vs. not removed),
different limma design formulas, etc. Rationally choosing the best combination 
of hyperparameters is challenging, so it is often better to try out multiple 
combinations and select the best one.

### Using `screen_limma_hyperparams()`

The function `screen_limma_hyperparams()` automates this process. For each 
hyperparameter, you specify the values you want to try, and the function runs 
the limma spline analysis with various combinations of these hyperparameters. 
Not every single combination is generated. Instead, there are "inner" and 
"outer" hyperparameters. All combinations are generated for "outer" 
hyperparameters, while specific combinations are generated for "inner" 
hyperparameters.

"Inner" hyperparameters include the adjusted p-value thresholds and spline 
parameters. For example, if you have two versions of a dataset (one with
potential outliers removed and one without), these are considered "outer" 
hyperparameters. The function generates all possible comparisons for the "outer"
hyperparameters, resulting in a single comparison. Then, for each version of the
data, it generates every combination of the "inner" hyperparameters.

### Example

For example, if you specify natural cubic splines with a degree of freedom of 2
or 3, and adjusted p-value thresholds of 0.05 or 0.1, the function will test all
combinations:

- DoF = 2, threshold = 0.05
- DoF = 3, threshold = 0.05
- DoF = 2, threshold = 0.1
- DoF = 3, threshold = 0.1

This systematic approach helps in identifying the best hyperparameters for the
analysis.

```{r Load hyperparameter-screening args}
data1 <- data 
meta1 <- meta

data2 <- data[, !(colnames(data) %in% c(  # Remove potential outliers
  "E12_TP05_Exponential",   
  "E10_TP10_Stationary"
  )
  )]

meta2 <- meta[!meta$`Sample.ID` %in% c(
  "E12_TP05_Exponential", 
  "E10_TP10_Stationary"
  ), ]

datas <- list(data1, data2) 
datas_descr <- c(
  "full_data",
  "outliers_removed"
  ) 

metas <- list(
  meta1,
  meta2
  ) 

designs <- c(
  "~ 1 + Phase*X + Reactor",
  "~ 1 + X + Reactor"
  ) 

pthresholds <- c(
  0.05,
  0.1
  )

# Every row a combo to test.
spline_test_configs <- data.frame(
  spline_type = c("n", "n", "n", "n"),  # All should use natural splines (n)
  degree = c(NA, NA, NA, NA),  # only needed for B-splines (spline_type = b)
  dof = c(2L, 3L, 4L, 5L),     # Test these variations of the DoF.
  # Per default, knots are placed automatically in a central fashion.
  knots = I(list(c(NA), c(NA), c(NA), c(NA))),                                      bknots = I(list(c(NA), c(NA), c(NA), c(NA)))
  )
```

## Required Arguments `screen_limma_hyperparams()`

- **splineomics**: A SplineOmics object containing the following fields:
  - **condition**: Column name of the levels (e.g., Exponential and Stationary).
  - **report_info**: A list containing general information about the analysis.
  - **meta_batch_column**: Column for meta batch information.
  - **meta_batch2_column**: Column for secondary meta batch information.

- **datas**: A list of data frames containing the datasets to be analyzed.
- **datas_descr**: A description object for the data.
- **metas**: A list of data frames containing metadata for each dataset in `datas`.
- **designs**: A character vector of design formulas for the limma analysis.
- **spline_test_configs**: A configuration object for spline tests.

## Optional Arguments `screen_limma_hyperparams()`

- **report_dir**: The path to the output directory. Default is current work dir.
- **adj_pthresholds**: A numeric vector of p-value thresholds for significance determination.
- **time_unit**: A character string specifying the time unit label for plots.
- **padjust_method**: A character string specifying the method for p-value                              adjustment. Default is "BH" (Benjamini-Hochberg).

```{r Perform hyperparameter-screening}
report_dir <- here::here(
  "demo_results",
  "hyperparams_screen_reports"
  ) 

SplineOmics::screen_limma_hyperparams(
  splineomics = splineomics,
  datas = datas,
  datas_descr = datas_descr,
  metas = metas,
  designs = designs,
  spline_test_configs = spline_test_configs,
  report_dir = report_dir,
  adj_pthresholds = pthresholds
  )
```

The last HTML generated by `screen_limma_hyperparams()` describes the meaning of
all short words used in the reports. For example it states that Design_1 =
"~ 1 + Phase*X + Reactor", and Design_2 = "~ 1 + X + Reactor".

# Run limma spline analysis

Once we identify the hyperparameters that are likely the best, we can run the
limma spline analysis with them to obtain the results.

## Required Arguments `update_splineomics()`

- **splineomics**: A SplineOmics object to be updated.

## Optional Arguments `update_splineomics()`

- **...**: Named arguments with new values for fields to be updated or added. Allowed fields include:
  - **data**: A matrix with the data.
  - **meta**: Metadata associated with the data.
  - **condition**: Column name of the levels (e.g., Exponential and Stationary).
  - **annotation**: A dataframe with the feature descriptions of data.
  - **report_info**: A list containing report information such as omics data type, data description, data collection date, analyst name, contact info, and project name.
  - **meta_batch_column**: Column for meta batch information.
  - **meta_batch2_column**: Column for secondary meta batch information.
  - **feature_name_columns**: Columns used to construct feature names.
  - **design**: A limma design formula
  - **spline_params**: Parameters for spline functions.
  - **limma_splines_result**: Results from the limma splines analysis.

```{r Update the SplineOmics object}
splineomics <- SplineOmics::update_splineomics(
  splineomics = splineomics,
  data = data2,   # Currently data1 (data is loaded)
  meta = meta2,   # Currently meta1 (meta is loaded)
  design = "~ 1 + Phase*X + Reactor",
  spline_params = list(     
    spline_type = c("n"),   # Natural splines for all levels
    dof = c(2L)             # Degree of freedom of 2 for all levels
    )
)
```

## Required Arguments `run_limma_splines()`

- **splineomics**: A SplineOmics object containing the following fields:
  - **data**: A matrix with the data.
  - **meta**: Metadata associated with the data.
  - **condition**: Column name of the levels (e.g., Exponential and Stationary).
  - **design**: A limma design formula
  - **spline_params**: Parameters for spline functions.

## Optional Arguments `run_limma_splines()`

- **padjust_method**: A character string specifying the method for p-value                              adjustment. Default is "BH" (Benjamini-Hochberg).

```{r Run the limma spline analysis}
# Run the limma spline analysis
splineomics <- SplineOmics::run_limma_splines(
  splineomics = splineomics
  )
```

The function run_limma_splines() adds a named list to the returned SplineOmics 
object. Each element in this list represents a specific "category" of results. 
These elements are lists containing the respective limma topTables, either for 
each level or for comparisons between two levels.

The element "time_effect" is a list where each element is a topTable reporting 
the p-values for each feature for the respective level.

The element "avrg_diff_conditions" is a list containing topTables that represent
the comparison of the average differences between the levels.

The element "interaction_condition_time" is a list containing topTables that
represent the interaction between the levels, which includes both time and
average differences.

# Build limma report

The topTables of all three categories can be used to generate p-value
histograms an volcano plots.

## Required Arguments `create_limma_report()`

- **splineomics**: A SplineOmics object containing the following fields:
  - **limma_splines_result**: A list containing the results of the limma 
                              analysis with splines. It should have three                                       components: `time_effect`, `avrg_diff_conditions`,                                and `interaction_condition_time`.
  - **condition**: Column name of the levels (e.g., Exponential and Stationary).
  - **annotation**: A dataframe with the feature descriptions of data.
  - **report_info**: A list containing metadata and other information to be                            included in the report.

## Optional Arguments `create_limma_report()`

- **adj_pthresh**: A numeric value specifying the adjusted p-value threshold for                     significance. Default is 0.05.
- **report_dir**: The path to the output directory. Default is current work dir.

```{r Build limma report}
report_dir <- here::here(
  "demo_results",
  "create_limma_reports"
  )

plots <- SplineOmics::create_limma_report(
  splineomics = splineomics,
  report_dir = report_dir
  )
```

# Cluster the hits (significant features)

After obtaining the limma spline results, we can cluster the hits based on their
temporal patterns (spline shapes). A hit is defined by setting an adjusted 
p-value threshold for each level. Hierarchical clustering is then used to assign
each hit to one of the specified number of clusters for that level.

```{r Prepare inputs for the cluster_hits function}
adj_pthresholds <- c(
  0.05,   # threshold for the exponential phase
  0.05    # threshold for the stationary phase
  )

clusters <- list(
  6L,   # 6 clusters for the exponential phase
  3L    # 3 clusters for the stationary phase
  )

plot_info = list(                    # For the spline plots
  y_axis_label = "log2 intensity",   # Unit of the values in the data matrix.
  time_unit = "min",
  treatment_labels = c("Feeding"),
  treatment_timepoints = c(0)       # The feeding occurred at 0 minutes.
)

gene_column_name <- "Genes" 
genes <- data_excel[[gene_column_name]][4:nrow(data_excel)]
```

## Required Arguments `cluster_hits()`

- **splineomics**: An S3 object of class `SplineOmics` that contains all the                         necessary data and parameters for the analysis, including:
  - **data**: A matrix with the data.
  - **meta**: Metadata associated with the data.
  - **design**: A limma design formula.
  - **condition**: Column name of the levels (e.g., Exponential and Stationary).
  - **spline_params**: Parameters for spline functions.
  - **meta_batch_column**: Column for meta batch information.
  - **meta_batch2_column**: Column for secondary meta batch information.
  - **limma_splines_result**: A list of data frames, each representing a 
                              top table from differential expression analysis,                                  containing at least 'adj.P.Val' and expression 
                              data columns.
  - **genes**: A character vector containing the gene names of the features to                   be analyzed.

## Optional Arguments `cluster_hits()`

- **adj_pthresholds**: Numeric vector of p-value thresholds for filtering hits                           in each top table. Default is 0.05.
- **clusters**: Character or integer vector specifying the number of clusters or                  'auto' for automatic estimation.
- **report_info**: A character string to be printed at the top of the report.
- **time_unit**: A character string specifying the time unit label for plots                       (e.g., 'min' for minutes).
- **report_dir**: The path to the output directory. Default is current work dir.
- **report**: Boolean TRUE or FALSE value specifying if a report should be                      generated. Default is TRUE.

```{r Cluster the hits}
report_dir <- here::here(
  "demo_results",
  "clustering_reports"
  )

clustering_results <- SplineOmics::cluster_hits(
  splineomics = splineomics,
  genes = genes,
  analysis_type = "time_effect",   # effect of time on features in level.
  adj_pthresholds = adj_pthresholds,
  clusters = clusters,
  plot_info = plot_info,
  report_dir = report_dir,
  )
```

# Perform gene set enrichment analysis (GSEA)

To each clustered hit, the respective gene can be assigned and GSEA performed. 
For this, the Enrichr databases of choice have to be downloaded:

```{r Define which Enrichr databases to download}
gene_set_lib <- c(
  "WikiPathways_2019_Human",
  "NCI-Nature_2016",
  "TRRUST_Transcription_Factors_2019",
  "MSigDB_Hallmark_2020",
  "GO_Cellular_Component_2018",
  "CORUM",
  "KEGG_2019_Human",
  "TRANSFAC_and_JASPAR_PWMs",
  "ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X",
  "GO_Biological_Process_2018",
  "GO_Molecular_Function_2018",
  "Human_Gene_Atlas"
  )
```

## Required Arguments `download_enrichr_databases()`

- **gene_set_lib**: A char vector of database names to download from Enrichr.

## Optional Arguments `download_enrichr_databases()`

- **output_dir**: The path to the output directory where the .tsv file will be                      saved. Defaults to the current working directory.

```{r Download the Enrichr databases}
output_dir  <- here::here(
  "demo_results",
  "downloaded_databases"
  )

SplineOmics::download_enrichr_databases(
  gene_set_lib = gene_set_lib,
  output_dir = output_dir
  )
```

To run GSEA, a genes vector must be created, containing all the underlying genes
of the features. The downloaded database file should be loaded as a dataframe.
Additionally, the clusterProfiler parameters and the report directory can 
optionally be specified. The function create_gsea_report() runs GSEA using
clusterProfiler, generates an HTML report, and returns the GSEA dot plots in R.

```{r Prepare GSEA inputs}
# Get gene vector. Every gene must be in the standardized format expected by
# the enrichment tools. The subsequent code extracts this part.
# For your analysis, this code needs to be customized based on the format.
gene_column_name <- "Genes"
genes <- annotation[[gene_column_name]][1:nrow(annotation)]
genes <- sub(" .*", "", genes)
genes <- sub(";.*", "", genes)
genes <- sub("_.*", "", genes)
genes <- sub("-.*", "", genes)

# The file has a timestamp, but this code takes it irrespective of it.
downloaded_dbs_filepath <- list.files(
  path = output_dir,
  pattern = "\\.tsv$",
  full.names = TRUE
  )[1]

databases <- readr::read_tsv(
  downloaded_dbs_filepath,
  col_types = readr::cols()
  )

clusterProfiler_params <- list(
  adj_p_value = 0.05,
  pAdjustMethod = "BH",
  minGSSize = 10,
  maxGSSize = 500,
  qvalueCutoff = 0.2
  )
```

## Required Arguments `create_gsea_report()`

- **levels_clustered_hits**: A list of clustered hits for the different levels.
- **genes**: A vector of genes of all features of the dataset
- **databases**: A list of databases to be used in the analysis.
- **report_info**: A list containing information for the report generation.

## Optional Arguments `create_gsea_report()`

- **params**: Additional parameters for the GSEA analysis. Default is NA.
- **plot_titles**: Titles for the plots. Default is NA.
- **background**: Background data. Default is NULL.
- **report_dir**: The path to the output directory where the report will be 
                  saved. Default is the current working directory.

```{r Run GSEA}
report_dir <- here::here(
  "demo_results",
  "gsea_reports"
  )

result <- SplineOmics::create_gsea_report(
  levels_clustered_hits = clustering_results$clustered_hits_levels,
  genes = genes,
  databases = databases,
  params = clusterProfiler_params,
  report_info = report_info,
  report_dir = report_dir
  )
```

In the output HTML, each row in the dot plots represents a term from a specific
database, and the columns correspond to the respective clusters. The color scale
indicates the odds ratio, while the size represents the -log10 adjusted p-value.
Only terms with support from more than 2 genes are included in the plot. For 
each cluster, a maximum of 5 terms with the highest odds ratios are shown.

Note that if, for example, cluster 1 already has 5 terms and cluster 2 receives
a term that was also found for cluster 1, this term will be included as the 
sixth term for cluster 1. This is how the maximum of 5 can be exceeded.
