<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Streamlines the process of analysing omics timeseries data with splines • SplineOmics</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Streamlines the process of analysing omics timeseries data with splines">
<meta property="og:description" content="Timeseries analysis of omics data can be carried out by fitting spline curves to the data and using limma for hypothesis testing. For this, the right spline freedom and further hyperparameters must be identified, and the obtained hits clustered based on the spline shape.The R package splinetime streamlines this whole process and generates reports.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">SplineOmics</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/design_limma_design_formula.html">Designing a Limma Design Formula</a>
    </li>
    <li>
      <a href="articles/Docker_permission_denied.html">Docker_permission_denied</a>
    </li>
    <li>
      <a href="articles/Docker-instructions.html">Docker-instructions</a>
    </li>
    <li>
      <a href="articles/get-started.html">get-started</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="splineomics">SplineOmics<a class="anchor" aria-label="anchor" href="#splineomics"></a>
</h1></div>
<p><img src="https://img.shields.io/badge/version-0.1.0-blue" alt="Version"><a href="./LICENSE"><img src="https://img.shields.io/badge/License-MIT-yellow.svg" alt="License: MIT"></a> <img src="https://img.shields.io/badge/Maintained%3F-yes-brightgreen.svg" alt="Maintained? Yes"><img src="https://img.shields.io/badge/R%20CMD%20check-passed-brightgreen" alt="R CMD Check"><a href="https://hub.docker.com/r/thomasrauter/splineomics" class="external-link"><img src="https://img.shields.io/badge/docker-pull-blue" alt="Docker"></a></p>
<p><img src="reference/figures/hex_logo.png" align="left" width="65"></p>
<p>The R package <code>SplineOmics</code> finds the significant features (hits) of time-series -omics data by using splines and <code>limma</code> for hypothesis testing. It then clusters the hits based on the spline shape while showing all results in summary HTML reports.</p>
<p>The graphical abstract below shows the full workflow streamlined by <code>SplineOmics</code>:</p>
<p><img src="reference/figures/SplineOmics_graphical_abstract.png"></p>
<div class="section level2">
<h2 id="table-of-contents">Table of Contents<a class="anchor" aria-label="anchor" href="#table-of-contents"></a>
</h2>
<ul>
<li><a href="#id_-introduction">📘 Introduction</a></li>
<li>
<a href="#id_-installation">🔧 Installation</a>
<ul>
<li><a href="#id_-docker-container">🐳 Docker Container</a></li>
</ul>
</li>
<li>
<a href="#id_-usage">▶️ Usage</a>
<ul>
<li><a href="#id_-tutorial">Tutorial</a></li>
<li><a href="#id_-functions-in-depth">Functions in Depth</a></li>
<li><a href="#id_-rna-seq-and-glycan-data">RNA-seq and Glycan Data</a></li>
</ul>
</li>
<li><a href="#id_-dependencies">📦 Dependencies</a></li>
<li><a href="#id_-further-reading">📚 Further Reading</a></li>
<li><a href="#id_-getting-help">❓ Getting Help</a></li>
<li><a href="#id_-contributing">🤝 Contributing</a></li>
<li><a href="#id_-feedback">💬 Feedback</a></li>
<li><a href="#id_-license">📜 License</a></li>
<li><a href="#id_-citation">🎓 Citation</a></li>
<li><a href="#id_-contributors">🌟 Contributors</a></li>
<li><a href="#id_-ackknowledgements">🙏 Acknowledgements</a></li>
</ul>
</div>
<div class="section level2">
<h2 id="id_-introduction">📘 Introduction<a class="anchor" aria-label="anchor" href="#id_-introduction"></a>
</h2>
<p>Welcome to <code>SplineOmics</code>, an R package designed to streamline the analysis of -omics time-series data, followed by automated HTML report generation.</p>
<div class="section level3">
<h3 id="is-the-splineomics-package-of-use-for-me">Is the SplineOmics package of use for me?<a class="anchor" aria-label="anchor" href="#is-the-splineomics-package-of-use-for-me"></a>
</h3>
<p>If you have -omics data over time, the package will help you to run <code>limma</code> with splines, decide on which parameters to use, perform the clustering, run GSEA and show result plots in HTML reports. Any time-series data that is a valid input to the <code>limma</code> package is also a valid input to the <code>SplineOmics</code> package (such as transcriptomics, proteomics, phosphoproteomics, metabolomics, glycan fractional abundances, etc.).</p>
</div>
<div class="section level3">
<h3 id="what-do-i-need-precisely">What do I need precisely?<a class="anchor" aria-label="anchor" href="#what-do-i-need-precisely"></a>
</h3>
<ol style="list-style-type: decimal">
<li><p><strong>Data</strong>: A data matrix where each row is a feature (e.g., protein, metabolite, etc.) and each column is a sample taken at a specific time.</p></li>
<li><p><strong>Meta</strong>: A table with metadata on the columns/samples of the data matrix (e.g., batch, time point, etc.)</p></li>
<li><p><strong>Annotation</strong>: A table with identifiers on the rows/features of the data matrix (e.g., gene and protein name).</p></li>
</ol>
</div>
<div class="section level3">
<h3 id="capabilities">Capabilities<a class="anchor" aria-label="anchor" href="#capabilities"></a>
</h3>
<p>With <code>SplineOmics</code>, you can:</p>
<ul>
<li>
<p><strong>Automatically perform exploratory data analysis:</strong></p>
<p>The <code><a href="reference/explore_data.html">explore_data()</a></code> function generates an HTML report, containing various plots, such as density, PCA, and correlation heatmap plots.</p>
</li>
<li>
<p><strong>Explore various limma splines hyperparameters:</strong></p>
<p>Test combinations of hyperparameters, such as different datasets, <code>limma</code> design formulas, degrees of freedom, p-value thresholds, etc., using the <code><a href="reference/screen_limma_hyperparams.html">screen_limma_hyperparams()</a></code> function.</p>
</li>
<li>
<p><strong>Perform limma spline analysis:</strong></p>
<p>Use the <code><a href="reference/run_limma_splines.html">run_limma_splines()</a></code> function to perform the <code>limma</code> analysis with splines once the optimal hyperparameters are identified.</p>
</li>
<li>
<p><strong>Cluster significant features:</strong></p>
<p>Cluster the significant features (hits) identified in the spline analysis with the <code><a href="reference/cluster_hits.html">cluster_hits()</a></code> function.</p>
</li>
<li>
<p><strong>Run GSEA with clustered hits:</strong></p>
<p>Perform gene set enrichment analysis (GSEA) using the clustered hits with the <code>create_gsea_report()</code> function.</p>
</li>
<li>
<p><strong>Generate reports:</strong></p>
<p>Automatically generate reports to showcase all results. See an example report of the cluster_hits() function <a href="https://github.com/csbg/SplineOmics_html_reports/raw/97b73e1cf1353df11a9035f576b98167699cf03c/report_clustered_hits_PTX_19_09_2024-13_47_08.html" class="external-link">here</a>.</p>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="id_-installation">🔧 Installation<a class="anchor" aria-label="anchor" href="#id_-installation"></a>
</h2>
<p>Follow the steps below to install the <code>SplineOmics</code> package from the GitHub repository into your R environment.</p>
<div class="section level4">
<h4 id="prerequisites">Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h4>
<ul>
<li>Ensure <strong>R</strong> is installed on your system. If not, download and install it from <a href="https://cran.r-project.org/" class="external-link">CRAN</a>.</li>
<li>
<strong>RStudio</strong> is recommended for a more user-friendly experience with R. Download and install RStudio from <a href="https://posit.co/download/rstudio-desktop/" class="external-link">posit.co</a>.</li>
</ul>
</div>
<div class="section level4">
<h4 id="installation-steps">Installation Steps<a class="anchor" aria-label="anchor" href="#installation-steps"></a>
</h4>
<ol style="list-style-type: decimal">
<li><p><strong>Open RStudio</strong> or your R console.</p></li>
<li><p><strong>Install <code>SplineOmics</code> from GitHub</strong> with all dependencies.</p></li>
</ol>
<p>Copy and paste the following code block into your R console or run it as a script.</p>
<blockquote>
<p><strong>Note for Windows Users:</strong><br>
Please read the text below this code block before running it!</p>
</blockquote>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Function to ensure a package is installed</span></span>
<span><span class="va">ensure_installed</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">pkg</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="va">pkg</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="va">pkg</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Install packages if not already available</span></span>
<span><span class="fu">ensure_installed</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">ensure_installed</span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">BiocManager</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://remotes.r-lib.org" class="external-link">remotes</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Install Bioconductor dependencies</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"ComplexHeatmap"</span>,</span>
<span>  <span class="st">"limma"</span></span>
<span>  <span class="op">)</span>, force <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Delete once repo is public</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Sys.setenv.html" class="external-link">Sys.setenv</a></span><span class="op">(</span>GITHUB_PAT <span class="op">=</span> <span class="st">"your_GitHub_PAT"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Install SplineOmics from GitHub</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span></span>
<span>  <span class="st">"csbg/SplineOmics@ad35d9aef2e8a8b19572c83ec771f4d92b343a4e"</span>,</span>
<span>  dependencies <span class="op">=</span> <span class="cn">TRUE</span>,  <span class="co"># Install all dependencies</span></span>
<span>  force <span class="op">=</span> <span class="cn">TRUE</span>,         <span class="co"># Force reinstallation</span></span>
<span>  upgrade <span class="op">=</span> <span class="st">"always"</span>,   <span class="co"># Always upgrade dependencies</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Verify the installation</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="st">"SplineOmics"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"SplineOmics installed successfully."</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"SplineOmics installation failed."</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Note that when some installation paths are not writable on <strong>Windows</strong>, it is necessary running <code>RStudio</code> as administrator once for the installation. Otherwise, set up a library path (code block below) for the installation and (re)run the code block above.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a directory for R libraries</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"~/Rlibs"</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set the library path to include the new directory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/libPaths.html" class="external-link">.libPaths</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"~/Rlibs"</span>, <span class="fu"><a href="https://rdrr.io/r/base/libPaths.html" class="external-link">.libPaths</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>
<strong>Load the <code>SplineOmics</code> package</strong>:</li>
</ol>
<p>Once the installation is complete, load the <code>SplineOmics</code> package into your R session or script to start using it:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SplineOmics</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="troubleshooting">Troubleshooting<a class="anchor" aria-label="anchor" href="#troubleshooting"></a>
</h4>
<p>If you encounter errors related to dependencies or package versions during installation, try updating your R and RStudio to the latest versions and repeat the installation steps.</p>
<p>For issues specifically related to the <code>SplineOmics</code> package, check the <a href="https://github.com/%3Cuser%3E/%3Crepo%3E/issues" class="external-link">Issues section</a> of the GitHub repository for similar problems or to post a new issue.</p>
</div>
<div class="section level3">
<h3 id="id_-docker-container">🐳 Docker Container<a class="anchor" aria-label="anchor" href="#id_-docker-container"></a>
</h3>
<p>Alternatively, you can run your analysis in a <code>Docker</code> container. The underlying <code>Docker</code> image encapsulates the <code>SplineOmics</code> package together with the necessary environment and dependencies. This ensures higher levels of reproducibility because the analysis is carried out in a consistent environment, independent of the operating system and its custom configurations.</p>
<p>More information about <code>Docker</code> containers can be found on the <a href="https://www.docker.com/resources/what-container/" class="external-link">official Docker page</a>.</p>
<p>For instructions on downloading the image of the <code>SplineOmics</code> package and running the container, please refer to the <a href="https://raw.githubusercontent.com/csbg/SplineOmics/main/doc/Docker_instructions.html" class="external-link">Docker instructions</a>.</p>
<div class="section level4">
<h4 id="troubleshooting-1">Troubleshooting<a class="anchor" aria-label="anchor" href="#troubleshooting-1"></a>
</h4>
<p>If you face “permission denied” issues on Linux distributions, check this <a href="https://raw.githubusercontent.com/csbg/SplineOmics/main/doc/Docker_permission_denied.html" class="external-link">vignette</a>.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="id_️-usage">▶️ Usage<a class="anchor" aria-label="anchor" href="#id_%EF%B8%8F-usage"></a>
</h2>
<div class="section level3">
<h3 id="tutorial">Tutorial<a class="anchor" aria-label="anchor" href="#tutorial"></a>
</h3>
<p><a href="https://csbg.github.io/SplineOmics/docs/get-started.html" class="external-link">This tutorial</a> covers a real CHO cell time-series proteomics example from start to end.</p>
<p>To open an R Markdown file of the <strong>tutorial</strong> in <code>RStudio</code>, run:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SplineOmics</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/open_tutorial.html">open_tutorial</a></span><span class="op">(</span><span class="op">)</span>  </span></code></pre></div>
<p>To open an R Markdown file in <code>RStudio</code> containing a <strong>template</strong> for your own analysis, run:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SplineOmics</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/open_template.html">open_template</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="functions-in-depth">Functions in Depth<a class="anchor" aria-label="anchor" href="#functions-in-depth"></a>
</h3>
<p>A detailed description of all arguments and outputs of the available package functions can be found <a href="https://raw.githubusercontent.com/csbg/SplineOmics/main/doc/functions-in-depth.html" class="external-link">here</a>.</p>
</div>
<div class="section level3">
<h3 id="design-limma-design-formula">Design <code>limma</code> design formula<a class="anchor" aria-label="anchor" href="#design-limma-design-formula"></a>
</h3>
<p>A quick guide on how to design a <code>limma</code> design formula can be found <a href="https://raw.githubusercontent.com/csbg/SplineOmics/main/doc/design_limma_design_formula.html" class="external-link">here</a></p>
<p>An explanation of the three different <code>limma</code> results can be found <a href="https://raw.githubusercontent.com/csbg/SplineOmics/main/inst/descriptions/limma_result_categories.pdf" class="external-link">here</a></p>
</div>
<div class="section level3">
<h3 id="rna-seq-and-glycan-data">RNA-seq and Glycan Data<a class="anchor" aria-label="anchor" href="#rna-seq-and-glycan-data"></a>
</h3>
<div class="section level4">
<h4 id="rna-seq-data">RNA-seq data<a class="anchor" aria-label="anchor" href="#rna-seq-data"></a>
</h4>
<p>Transcriptomics data must be preprocessed for <code>limma</code>. This is done by setting the preprocess_rna_seq argument to TRUE (see documentation of create_splineomics() function). Then, the raw RNA-seq counts provided in the data matrix will undergo normalization and transformation. The default normalization is performed using TMM (Trimmed Mean of M-values) normalization via the <code>edgeR</code>::calcNormFactors function, followed by the voom transformation from the <code>limma</code> package to obtain log-transformed counts per million (logCPM) with associated precision weights. If you require a different normalization method, you can supply your custom normalization function.</p>
</div>
<div class="section level4">
<h4 id="glycan-fractional-abundance-data">Glycan fractional abundance data<a class="anchor" aria-label="anchor" href="#glycan-fractional-abundance-data"></a>
</h4>
<p>The glycan fractional abundance data matrix, where each row represents a type of glycan and the columns correspond to timepoints, must be transformed before analysis. This preprocessing step is essential due to the compositional nature of the data. In compositional data, an increase in the abundance of one component (glycan) necessarily results in a decrease in others, introducing a dependency among the variables that can bias the analysis. One way to address this issue is by applying the Centered Log Ratio (CLR) transformation to the data with the clr function from the compositions package:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.stat.boogaart.de/compositions/" class="external-link">compositions</a></span><span class="op">)</span></span>
<span><span class="va">clr_transformed_data</span> <span class="op">&lt;-</span> <span class="fu">clr</span><span class="op">(</span><span class="va">data_matrix</span><span class="op">)</span>  <span class="co"># use as SplineOmics input</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="id_-dependencies">📦 Dependencies<a class="anchor" aria-label="anchor" href="#id_-dependencies"></a>
</h2>
<p>The <code>SplineOmics</code> package relies on several other R packages for its functionality. Below is a list of dependencies that will automatically be installed along with <code>SplineOmics</code>. If you already have these packages installed, ensure they are up to date to avoid any compatibility issues.</p>
<ul>
<li>
<strong>ComplexHeatmap</strong>: For creating complex heatmaps with advanced features.</li>
<li>
<strong>base64enc</strong>: For encoding/decoding base64.</li>
<li>
<strong>dendextend</strong>: For extending <code>dendrogram</code> objects in R, allowing for easier manipulation of dendrograms.</li>
<li>
<strong>dplyr</strong>: For data manipulation.</li>
<li>
<strong>ggplot2</strong>: For creating elegant data visualizations using the grammar of graphics.</li>
<li>
<strong>ggrepel</strong>: For better label placement in ggplot2.</li>
<li>
<strong>here</strong>: For constructing paths to your project’s files.</li>
<li>
<strong>limma</strong>: For linear models for microarray data.</li>
<li>
<strong>openxlsx</strong>: For reading, writing, and editing xlsx files.</li>
<li>
<strong>patchwork</strong>: For combining multiple ggplot objects into a single plot.</li>
<li>
<strong>pheatmap</strong>: For creating pretty heatmaps.</li>
<li>
<strong>progress</strong>: For adding progress bars to your loops and apply functions.</li>
<li>
<strong>purrr</strong>: For functional programming tools.</li>
<li>
<strong>rlang</strong>: For tools to work with core language features of R and R’s base types.</li>
<li>
<strong>scales</strong>: For scale functions for visualization.</li>
<li>
<strong>tibble</strong>: For creating tidy data frames that are easy to work with.</li>
<li>
<strong>tidyr</strong>: For tidying your data.</li>
<li>
<strong>zip</strong>: For combining files into a zip file.</li>
</ul>
<div class="section level3">
<h3 id="optional-dependencies">Optional dependencies<a class="anchor" aria-label="anchor" href="#optional-dependencies"></a>
</h3>
<p>These dependencies are only necessary for some functions:</p>
<ul>
<li>
<strong>edgeR</strong>: For preprocessing RNA-seq data in the run_limma_splines() fun.</li>
<li>
<strong>clusterProfiler</strong>: For the run_gsea() function (gene set enrichment).</li>
<li>
<strong>rstudioapi</strong>: For the open_tutorial() and open_template() functions.</li>
</ul>
</div>
<div class="section level3">
<h3 id="r-version">R Version<a class="anchor" aria-label="anchor" href="#r-version"></a>
</h3>
<ul>
<li>Recommended: R 4.3.3 or higher</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="id_-further-reading">📚 Further Reading<a class="anchor" aria-label="anchor" href="#id_-further-reading"></a>
</h2>
<p>For those interested in gaining a deeper understanding of the methodologies used in the <code>SplineOmics</code> package, here are some recommended publications:</p>
<ul>
<li><p><strong>Splines</strong>: To learn more about splines, you can refer to this <a href="https://doi.org/10.1186/s12874-019-0666-3" class="external-link">review</a>.</p></li>
<li><p><strong>limma</strong>: To read about the <code>limma</code> R package, you can refer to this <a href="https://doi.org/10.1093/nar/gkv007" class="external-link">publication</a>.</p></li>
<li><p><strong>Hierarchical clustering</strong>: To get information about hierarchical clustering, you can refer to this <a href="https://towardsdatascience.com/understanding-the-concept-of-hierarchical-clustering-technique-c6e8243758ec" class="external-link">web article</a>.</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="id_-getting-help">❓ Getting Help<a class="anchor" aria-label="anchor" href="#id_-getting-help"></a>
</h2>
<p>If you encounter a bug or have a suggestion for improving the <code>SplineOmics</code> package, we encourage you to <a href="https://github.com/csbg/SplineOmics/issues" class="external-link">open an issue</a> on our GitHub repository. Before opening a new issue, please check to see if your question or bug has already been reported by another user. This helps avoid duplicate reports and ensures that we can address problems efficiently.</p>
<p>For more detailed questions, discussions, or contributions regarding the package’s use and development, please refer to the <a href="https://github.com/csbg/SplineOmics/discussions" class="external-link">GitHub Discussions</a> page for <code>SplineOmics</code>.</p>
</div>
<div class="section level2">
<h2 id="id_-contributing">🤝 Contributing<a class="anchor" aria-label="anchor" href="#id_-contributing"></a>
</h2>
<p>We welcome contributions to the <code>SplineOmics</code> package! Whether you’re interested in fixing bugs, adding new features, or improving documentation, your help is greatly appreciated.</p>
<p>Here’s how you can contribute:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Report a Bug or Request a Feature:</strong> If you encounter a bug or have an idea for a new feature, please <a href="https://github.com/csbg/SplineOmics/issues" class="external-link">open an issue</a> on our GitHub repository. Before opening a new issue, check to see if the issue has already been reported or the feature requested by another user.</p></li>
<li><p><strong>Submit a Pull Request:</strong> If you’ve developed a bug fix or a new feature that you’d like to share, submit a pull request.</p></li>
<li><p><strong>Improve Documentation:</strong> Good documentation is crucial for any project. If you notice missing or incorrect documentation, please feel free to contribute.</p></li>
</ol>
<p>Please adhere to our <a href="CODE_OF_CONDUCT.html">Code of Conduct</a> in all your interactions with the project.</p>
<p>Thank you for considering contributing to <code>SplineOmics</code>. Your efforts are what make the open-source community a fantastic place to learn, inspire, and create.</p>
</div>
<div class="section level2">
<h2 id="id_-feedback">💬 Feedback<a class="anchor" aria-label="anchor" href="#id_-feedback"></a>
</h2>
<p>We appreciate your feedback! Besides raising issues, you can provide feedback in the following ways:</p>
<ul>
<li><p><strong>Direct Email</strong>: Send your feedback directly to <a href="mailto:thomas.rauter@plus.ac.at">Thomas Rauter</a>.</p></li>
<li><p><strong>Anonymous Feedback</strong>: Use <a href="https://forms.gle/jocMXSxLf3GrGBdT9" class="external-link">this Google Form</a> to provide anonymous feedback by answering questions.</p></li>
</ul>
<p>Your feedback helps us improve the project and address any issues you may encounter.</p>
</div>
<div class="section level2">
<h2 id="id_-license">📜 License<a class="anchor" aria-label="anchor" href="#id_-license"></a>
</h2>
<p>This package is licensed under the MIT License: <a href="./LICENSE">LICENSE</a></p>
<p>© 2024 Thomas Rauter. All rights reserved.</p>
</div>
<div class="section level2">
<h2 id="id_-citation">🎓 Citation<a class="anchor" aria-label="anchor" href="#id_-citation"></a>
</h2>
<p>The <code>SplineOmics</code> package is currently not published in a peer-reviewed scientific journal or similar outlet. However, if this package helped you in your work, consider citing this GitHub repository.</p>
<p>To cite this package, you can use the citation information provided in the <a href="./inst/CITATION.cff"><code>inst/CITATION.cff</code></a> file.</p>
<p>You can also generate a citation in various formats using the <code>CITATION.cff</code> file by visiting the top right of this repo and clicking on the “Cite this repository” button.</p>
<p>Also, if you like the package, consider giving the GitHub repository a star. Your support helps us in the continued development and improvement of <code>SplineOmics</code>. Thank you for using our package!</p>
</div>
<div class="section level2">
<h2 id="id_-contributors">🌟 Contributors<a class="anchor" aria-label="anchor" href="#id_-contributors"></a>
</h2>
<ul>
<li>
<a href="https://github.com/Thomas-Rauter" class="external-link">Thomas-Rauter</a> - 🚀 Wrote the package, developed the approach together with VSchaepertoens under guidance from nfortelny and skafdasschaf.</li>
<li>
<a href="https://github.com/nfortelny" class="external-link">nfortelny</a> - 🧠 Principal Investigator, provided guidance and support for the overall approach.</li>
<li>
<a href="https://github.com/skafdasschaf" class="external-link">skafdasschaf</a> - 🔧 Helped reviewing code, delivered improvement suggestions and scientific guidance to develop the approach.</li>
<li>
<a href="https://github.com/VSchaepertoens" class="external-link">VSchaepertoens</a> - ✨ Developed one internal plotting function, as well as some code for the exploratory data analysis plots, and the overall approach together with Thomas-Rauter.</li>
</ul>
</div>
<div class="section level2">
<h2 id="id_-acknowledgements">🙏 Acknowledgements<a class="anchor" aria-label="anchor" href="#id_-acknowledgements"></a>
</h2>
<p>This work was carried out in the context of the DigiTherapeutX project, which was funded by the Austrian Science Fund (FWF). The research was conducted under the supervision of Prof. Nikolaus Fortelny, who leads the Computational Systems Biology working group at the Paris Lodron University of Salzburg, Austria. You can find more information about Prof. Fortelny’s research group <a href="https://www.plus.ac.at/biowissenschaften/der-fachbereich/arbeitsgruppen/fortelny/" class="external-link">here</a>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing SplineOmics</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Thomas Rauter <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Thomas Rauter.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
