<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Streamlines the process of analysing omics timeseries data with splines ‚Ä¢ SplineOmics</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Streamlines the process of analysing omics timeseries data with splines">
<meta name="description" content="Timeseries analysis of omics data can be carried out by fitting spline curves to the data and using limma for hypothesis testing. For this, the right spline freedom and further hyperparameters must be identified, and the obtained hits clustered based on the spline shape.The R package splinetime streamlines this whole process and generates reports.">
<meta property="og:description" content="Timeseries analysis of omics data can be carried out by fitting spline curves to the data and using limma for hypothesis testing. For this, the right spline freedom and further hyperparameters must be identified, and the obtained hits clustered based on the spline shape.The R package splinetime streamlines this whole process and generates reports.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">SplineOmics</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/design_limma_design_formula.html">Designing a Limma Design Formula</a></li>
    <li><a class="dropdown-item" href="articles/Docker_permission_denied.html">Docker_permission_denied</a></li>
    <li><a class="dropdown-item" href="articles/Docker-instructions.html">Docker-instructions</a></li>
    <li><a class="dropdown-item" href="articles/get-started.html">get-started</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="splineomics">SplineOmics<a class="anchor" aria-label="anchor" href="#splineomics"></a>
</h1></div>
<p><img src="https://img.shields.io/badge/version-0.1.0-blue" alt="Version"><a href="./LICENSE"><img src="https://img.shields.io/badge/License-MIT-yellow.svg" alt="License: MIT"></a> <img src="https://img.shields.io/badge/Maintained%3F-yes-brightgreen.svg" alt="Maintained? Yes"><img src="https://img.shields.io/badge/R%20CMD%20check-passed-brightgreen" alt="R CMD Check"><a href="https://hub.docker.com/r/thomasrauter/splineomics" class="external-link"><img src="https://img.shields.io/badge/docker-pull-blue" alt="Docker"></a> <img src="https://img.shields.io/badge/dependencies-19-blue" alt="Dependencies"><img src="https://img.shields.io/badge/platforms-all-brightgreen" alt="Platforms"></p>
<p><img src="reference/figures/hex_logo.png" align="left" width="65"></p>
<p>The R package <code>SplineOmics</code> finds the significant features (hits) of time-series -omics data by using splines and <code>limma</code> for hypothesis testing. It then clusters the hits based on the spline shape while showing all results in summary HTML reports.</p>
<p>The graphical abstract below shows the full workflow streamlined by <code>SplineOmics</code>:</p>
<figure><img src="reference/figures/SplineOmics_graphical_abstract.png" alt="Graphical Abstract of SplineOmics Workflow"><figcaption aria-hidden="true">
Graphical Abstract of SplineOmics Workflow
</figcaption></figure><div class="section level2">
<h2 id="table-of-contents">Table of Contents<a class="anchor" aria-label="anchor" href="#table-of-contents"></a>
</h2>
<ul>
<li><a href="#id_-introduction">üìò Introduction</a></li>
<li>
<a href="#id_-installation">üîß Installation</a>
<ul>
<li><a href="#id_-docker-container">üê≥ Docker Container</a></li>
</ul>
</li>
<li>
<a href="#id_-usage">‚ñ∂Ô∏è Usage</a>
<ul>
<li><a href="#id_-tutorial">Tutorial</a></li>
<li><a href="#id_-details">Details</a></li>
<li><a href="#id_-rna-seq-and-glycan-data">RNA-seq and Glycan Data</a></li>
</ul>
</li>
<li><a href="#id_-dependencies">üì¶ Dependencies</a></li>
<li><a href="#id_-further-reading">üìö Further Reading</a></li>
<li><a href="#id_-getting-help">‚ùì Getting Help</a></li>
<li><a href="#id_-contributing">ü§ù Contributing</a></li>
<li><a href="#id_-feedback">üí¨ Feedback</a></li>
<li><a href="#id_-license">üìú License</a></li>
<li><a href="#id_-citation">üéì Citation</a></li>
<li><a href="#id_-contributors">üåü Contributors</a></li>
<li><a href="#id_-acknowledgements">üôè Acknowledgements</a></li>
</ul>
</div>
<div class="section level2">
<h2 id="id_-introduction">üìò Introduction<a class="anchor" aria-label="anchor" href="#id_-introduction"></a>
</h2>
<p>Welcome to <code>SplineOmics</code>, an R package designed to streamline the analysis of -omics time-series data, followed by automated HTML report generation.</p>
<div class="section level3">
<h3 id="is-the-splineomics-package-of-use-for-me">Is the SplineOmics package of use for me?<a class="anchor" aria-label="anchor" href="#is-the-splineomics-package-of-use-for-me"></a>
</h3>
<p>If you have -omics data over time, the package will help you to run <code>limma</code> with splines, decide on which parameters to use, perform the clustering, run GSEA and show result plots in HTML reports. Any time-series data that is a valid input to the <code>limma</code> package is also a valid input to the <code>SplineOmics</code> package (such as transcriptomics, proteomics, phosphoproteomics, metabolomics, glycan fractional abundances, etc.).</p>
</div>
<div class="section level3">
<h3 id="what-do-i-need-precisely">What do I need precisely?<a class="anchor" aria-label="anchor" href="#what-do-i-need-precisely"></a>
</h3>
<ol style="list-style-type: decimal">
<li><p><strong>Data</strong>: A data matrix where each row is a feature (e.g., protein, metabolite, etc.) and each column is a sample taken at a specific time. The data must have no NA values and should have normally distributed features.</p></li>
<li><p><strong>Meta</strong>: A table with metadata on the columns/samples of the data matrix (e.g., batch, time point, etc.)</p></li>
<li><p><strong>Annotation</strong> (optional): A table with identifiers on the rows/features of the data matrix (e.g., gene and protein name).</p></li>
</ol>
</div>
<div class="section level3">
<h3 id="capabilities">Capabilities<a class="anchor" aria-label="anchor" href="#capabilities"></a>
</h3>
<p>With <code>SplineOmics</code>, you can:</p>
<ul>
<li>
<p><strong>Automatically perform exploratory data analysis:</strong></p>
<p>The <code><a href="reference/explore_data.html">explore_data()</a></code> function generates an HTML report, containing various plots, such as density, PCA, and correlation heatmap plots (<a href="https://csbg.github.io/SplineOmics_html_reports/explore_data_PTX.html">example report</a>).</p>
</li>
<li>
<p><strong>Explore various limma splines hyperparameters:</strong></p>
<p>Test combinations of hyperparameters, such as different datasets, <code>limma</code> design formulas, degrees of freedom, p-value thresholds, etc., using the <code><a href="reference/screen_limma_hyperparams.html">screen_limma_hyperparams()</a></code> function (<a href="https://csbg.github.io/SplineOmics_html_reports/Data_1_Design_1_vs_Data_1_Design_2_PTX.html">example report</a> (along with the <a href="https://csbg.github.io/SplineOmics_html_reports/hyperparams_screen_meta_table.html">encoding</a>)).</p>
</li>
<li>
<p><strong>Perform limma spline analysis:</strong></p>
<p>Use the <code><a href="reference/run_limma_splines.html">run_limma_splines()</a></code> function to perform the <code>limma</code> analysis with splines once the optimal hyperparameters are identified (<a href="https://csbg.github.io/SplineOmics_html_reports/create_limma_report_PTX.html">example report</a>).</p>
</li>
<li>
<p><strong>Cluster significant features:</strong></p>
<p>Cluster the significant features (hits) identified in the spline analysis with the <code><a href="reference/cluster_hits.html">cluster_hits()</a></code> function (<a href="https://csbg.github.io/SplineOmics_html_reports/report_clustered_hits_PTX.html">example report</a>).</p>
</li>
<li>
<p><strong>Run GSEA with clustered hits:</strong></p>
<p>Perform gene set enrichment analysis (GSEA) using the clustered hits with the <code>create_gsea_report()</code> function (<a href="https://csbg.github.io/SplineOmics_html_reports/create_gsea_report_PTX.html">example report</a>).</p>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="id_-installation">üîß Installation<a class="anchor" aria-label="anchor" href="#id_-installation"></a>
</h2>
<p>Follow the steps below to install the <code>SplineOmics</code> package from the GitHub repository into your R environment.</p>
<div class="section level4">
<h4 id="prerequisites">Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h4>
<ul>
<li>Ensure <strong>R</strong> is installed on your system. If not, download and install it from <a href="https://cran.r-project.org/" class="external-link">CRAN</a>.</li>
<li>
<strong>RStudio</strong> is recommended for a more user-friendly experience with R. Download and install RStudio from <a href="https://posit.co/download/rstudio-desktop/" class="external-link">posit.co</a>.</li>
</ul>
</div>
<div class="section level4">
<h4 id="installation-steps">Installation Steps<a class="anchor" aria-label="anchor" href="#installation-steps"></a>
</h4>
<blockquote>
<p><strong>Note for Windows Users:</strong></p>
</blockquote>
<p>Note that some installation paths potentially are not writable on <strong>Windows</strong>. Therefore, it can be necessary to set up a library path and use that path for the installations:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define the custom library path and expand the tilde (~)</span></span>
<span><span class="va">custom_lib_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/path.expand.html" class="external-link">path.expand</a></span><span class="op">(</span><span class="st">"~/Rlibs"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create the directory if it doesn't exist</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="va">custom_lib_path</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">custom_lib_path</span>, showWarnings <span class="op">=</span> <span class="cn">FALSE</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Set the library path to include the new directory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/libPaths.html" class="external-link">.libPaths</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">custom_lib_path</span>, <span class="fu"><a href="https://rdrr.io/r/base/libPaths.html" class="external-link">.libPaths</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Check if the new library path is added successfully</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="va">custom_lib_path</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/libPaths.html" class="external-link">.libPaths</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"Library path set to: "</span>, <span class="va">custom_lib_path</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="st">"Failed to set library path."</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Alternatively, you can run <code>RStudio</code> as administrator once for the installation (which is however generally not recommended, because it is a security risk).</p>
<ol style="list-style-type: decimal">
<li><p><strong>Open RStudio</strong> or your R console.</p></li>
<li><p><strong>Install <code>BiocManager</code></strong> for Bioconductor dependencies (if not already installed)</p></li>
</ol>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span></span>
<span>  <span class="st">"BiocManager"</span>, </span>
<span>  <span class="co"># lib = custom_lib_path</span></span>
<span><span class="op">)</span></span></code></pre></div>
<ol start="3" style="list-style-type: decimal">
<li>
<strong>Install Bioconductor dependencies</strong> separately using <code>BiocManager</code>
</li>
</ol>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">BiocManager</a></span><span class="op">)</span>  <span class="co"># load the BiocManager package</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"ComplexHeatmap"</span>,</span>
<span>  <span class="st">"limma"</span></span>
<span>  <span class="op">)</span>, </span>
<span>  force <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="co"># lib = custom_lib_path </span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<ol start="4" style="list-style-type: decimal">
<li>
<strong>Install</strong> the <strong><code>remotes</code></strong> package for GitHub downloads (if not already installed)</li>
</ol>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span></span>
<span>  <span class="st">"remotes"</span>,</span>
<span>  <span class="co"># lib = custom_lib_path</span></span>
<span><span class="op">)</span></span></code></pre></div>
<ol start="5" style="list-style-type: decimal">
<li>
<strong>Install</strong> the <strong><code>SplineOmics</code></strong> package from GitHub and all its non-Bioconductor dependencies, using <code>remotes</code>
</li>
</ol>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://remotes.r-lib.org" class="external-link">remotes</a></span><span class="op">)</span>  <span class="co"># load the remotes package</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span></span>
<span>  <span class="st">"csbg/SplineOmics"</span>,   <span class="co"># GitHub repository</span></span>
<span>  ref <span class="op">=</span> <span class="st">"0.1.0"</span>,       <span class="co"># Specify the tag to install</span></span>
<span>  dependencies <span class="op">=</span> <span class="cn">TRUE</span>,  <span class="co"># Install all dependencies</span></span>
<span>  force <span class="op">=</span> <span class="cn">TRUE</span>,         <span class="co"># Force reinstallation</span></span>
<span>  upgrade <span class="op">=</span> <span class="st">"always"</span>,   <span class="co"># Always upgrade dependencies</span></span>
<span>  <span class="co"># lib = custom_lib_path </span></span>
<span><span class="op">)</span></span></code></pre></div>
<ol start="6" style="list-style-type: decimal">
<li>
<strong>Verify</strong> the <strong>installation</strong> of the <code>SplineOmics</code> package</li>
</ol>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="st">"SplineOmics"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"SplineOmics installed successfully."</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"SplineOmics installation failed."</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="troubleshooting">Troubleshooting<a class="anchor" aria-label="anchor" href="#troubleshooting"></a>
</h4>
<p>If you encounter errors related to dependencies or package versions during installation, try updating your R and RStudio to the latest versions and repeat the installation steps.</p>
<p>For issues specifically related to the <code>SplineOmics</code> package, check the <a href="https://github.com/csbg/SplineOmics/issues" class="external-link">Issues section</a> of the GitHub repository for similar problems or to post a new issue.</p>
</div>
<div class="section level3">
<h3 id="id_-docker-container">üê≥ Docker Container<a class="anchor" aria-label="anchor" href="#id_-docker-container"></a>
</h3>
<p>Alternatively, you can run your analysis in a <code>Docker</code> container. The underlying <code>Docker</code> image encapsulates the <code>SplineOmics</code> package together with the necessary environment and dependencies. This ensures higher levels of reproducibility because the analysis is carried out in a consistent environment, independent of the operating system and its custom configurations.</p>
<p>More information about <code>Docker</code> containers can be found on the <a href="https://www.docker.com/resources/what-container/" class="external-link">official Docker page</a>.</p>
<p>For instructions on downloading the image of the <code>SplineOmics</code> package and running the container, please refer to the <a href="https://csbg.github.io/SplineOmics/articles/Docker-instructions.html">Docker instructions</a>.</p>
<div class="section level4">
<h4 id="troubleshooting-1">Troubleshooting<a class="anchor" aria-label="anchor" href="#troubleshooting-1"></a>
</h4>
<p>If you face ‚Äúpermission denied‚Äù issues on Linux distributions, check this <a href="https://csbg.github.io/SplineOmics/articles/Docker_permission_denied.html">vignette</a>.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="id_-usage">‚ñ∂ Usage<a class="anchor" aria-label="anchor" href="#id_-usage"></a>
</h2>
<div class="section level3">
<h3 id="tutorial">Tutorial<a class="anchor" aria-label="anchor" href="#tutorial"></a>
</h3>
<p><a href="https://csbg.github.io/SplineOmics/articles/get-started.html">This tutorial</a> covers a real CHO cell time-series proteomics example from start to end.</p>
<p>To open an R Markdown file of the <strong>tutorial</strong> in <code>RStudio</code>, run:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://csbg.github.io/SplineOmics">SplineOmics</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/open_tutorial.html">open_tutorial</a></span><span class="op">(</span><span class="op">)</span>  </span></code></pre></div>
<p>To open an R Markdown file in <code>RStudio</code> containing a <strong>template</strong> for your own analysis, run:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://csbg.github.io/SplineOmics">SplineOmics</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/open_template.html">open_template</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a>
</h3>
<p>A detailed description of all arguments and outputs of all the functions in the package (exported and internal functions) can be found <a href="https://csbg.github.io/SplineOmics/reference/">here</a>.</p>
<div class="section level4">
<h4 id="design-limma-design-formula">Design <code>limma</code> design formula<a class="anchor" aria-label="anchor" href="#design-limma-design-formula"></a>
</h4>
<p>A quick guide on how to design a <code>limma</code> design formula can be found <a href="https://csbg.github.io/SplineOmics/articles/design_limma_design_formula.html">here</a></p>
<p>An explanation of the three different <code>limma</code> results is <a href="https://csbg.github.io/SplineOmics/articles/limma_result_categories.html">here</a></p>
</div>
</div>
<div class="section level3">
<h3 id="rna-seq-and-glycan-data">RNA-seq and Glycan Data<a class="anchor" aria-label="anchor" href="#rna-seq-and-glycan-data"></a>
</h3>
<div class="section level4">
<h4 id="rna-seq-data">RNA-seq data<a class="anchor" aria-label="anchor" href="#rna-seq-data"></a>
</h4>
<p>Transcriptomics data must be preprocessed for <code>limma</code>. You need to provide an appropriate object, such as a <code>voom</code> object, in the <code>rna_seq_data</code> argument of the <code>SplineOmics</code> object (see <a href="https://csbg.github.io/SplineOmics/reference/create_splineomics.html">documentation</a>). Along with this, the normalized matrix (e.g., the <code>$E</code> slot of the <code>voom</code> object) must be passed to the <code>data</code> argument. This allows flexibility in preprocessing; you can use any method you prefer as long as the final object and matrix are compatible with limma. One way to preprocess your RNA-seq data is by using the <code><a href="reference/preprocess_rna_seq_data.html">preprocess_rna_seq_data()</a></code> function included in the <code>SplineOmics</code> package (see <a href="https://csbg.github.io/SplineOmics/reference/preprocess_rna_seq_data.html">documentation</a>).</p>
</div>
<div class="section level4">
<h4 id="glycan-fractional-abundance-data">Glycan fractional abundance data<a class="anchor" aria-label="anchor" href="#glycan-fractional-abundance-data"></a>
</h4>
<p>The glycan fractional abundance data matrix, where each row represents a type of glycan and the columns correspond to timepoints, must be transformed before analysis. This preprocessing step is essential due to the compositional nature of the data. In compositional data, an increase in the abundance of one component (glycan) necessarily results in a decrease in others, introducing a dependency among the variables that can bias the analysis. One way to address this issue is by applying the Centered Log Ratio (CLR) transformation to the data with the clr function from the compositions package:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.stat.boogaart.de/compositions/" class="external-link">compositions</a></span><span class="op">)</span></span>
<span><span class="va">clr_transformed_data</span> <span class="op">&lt;-</span> <span class="fu">clr</span><span class="op">(</span><span class="va">data_matrix</span><span class="op">)</span>  <span class="co"># use as SplineOmics input</span></span></code></pre></div>
<p>The results from clr transformed data can be harder to understand and interpret however. If you prefer ease of interpretation and are fine that the results contain some artifacts due to the compositional nature of the data, log2 transform your data instead and use that as input for the <code>SplineOmics</code> package.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">log2_transformed_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">data_matrix</span><span class="op">)</span>  <span class="co"># use as SplineOmics input</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="id_-dependencies">üì¶ Dependencies<a class="anchor" aria-label="anchor" href="#id_-dependencies"></a>
</h2>
<p>The <code>SplineOmics</code> package relies on several other R packages for its functionality. Below is a list of dependencies that will automatically be installed along with <code>SplineOmics</code>. If you already have these packages installed, ensure they are up to date to avoid any compatibility issues.</p>
<ul>
<li>
<strong>ComplexHeatmap</strong> (&gt;= 2.18.0): For creating complex heatmaps with advanced features.</li>
<li>
<strong>base64enc</strong> (&gt;= 0.1-3): For encoding/decoding base64.</li>
<li>
<strong>dendextend</strong> (&gt;= 1.17.1): For extending <code>dendrogram</code> objects, allowing for easier manipulation of dendrograms.</li>
<li>
<strong>dplyr</strong> (&gt;= 1.1.4): For data manipulation.</li>
<li>
<strong>ggplot2</strong> (&gt;= 3.5.1): For creating elegant data visualizations using the grammar of graphics.</li>
<li>
<strong>ggrepel</strong> (&gt;= 0.9.5): For better label placement in <code>ggplot2</code>.</li>
<li>
<strong>here</strong> (&gt;= 1.0.1): For constructing paths to your project‚Äôs files.</li>
<li>
<strong>limma</strong> (&gt;= 3.58.1): For linear models in microarray and RNA-seq analysis.</li>
<li>
<strong>openxlsx</strong> (&gt;= 4.2.6.1): For reading, writing, and editing <code>.xlsx</code> files.</li>
<li>
<strong>patchwork</strong> (&gt;= 1.2.0): For combining multiple <code>ggplot</code> objects into a single plot.</li>
<li>
<strong>pheatmap</strong> (&gt;= 1.0.12): For creating aesthetically pleasing heatmaps.</li>
<li>
<strong>progress</strong> (&gt;= 1.2.3): For adding progress bars to loops and apply functions.</li>
<li>
<strong>purrr</strong> (&gt;= 1.0.2): For functional programming tools.</li>
<li>
<strong>rlang</strong> (&gt;= 1.1.4): For working with core language features of R.</li>
<li>
<strong>scales</strong> (&gt;= 1.3.0): For scale functions in visualizations.</li>
<li>
<strong>svglite</strong> (&gt;= 2.1.3): For creating high-quality vector graphics (SVG).</li>
<li>
<strong>tibble</strong> (&gt;= 3.2.1): For creating tidy data frames.</li>
<li>
<strong>tidyr</strong> (&gt;= 1.3.1): For tidying data.</li>
<li>
<strong>zip</strong> (&gt;= 2.3.1): For compressing and combining files into zip archives.</li>
</ul>
<div class="section level3">
<h3 id="optional-dependencies">Optional Dependencies<a class="anchor" aria-label="anchor" href="#optional-dependencies"></a>
</h3>
<p>These packages are optional and are only needed for specific functionality:</p>
<ul>
<li>
<strong>edgeR</strong> (&gt;= 4.0.16): For preprocessing RNA-seq data in the <code><a href="reference/preprocess_rna_seq_data.html">preprocess_rna_seq_data()</a></code> function.</li>
<li>
<strong>clusterProfiler</strong> (&gt;= 4.10.1): For the <code><a href="reference/run_gsea.html">run_gsea()</a></code> function (gene set enrichment analysis).</li>
<li>
<strong>rstudioapi</strong> (&gt;= 0.16.0): For the <code><a href="reference/open_tutorial.html">open_tutorial()</a></code> and <code><a href="reference/open_template.html">open_template()</a></code> functions.</li>
</ul>
</div>
<div class="section level3">
<h3 id="r-version">R Version<a class="anchor" aria-label="anchor" href="#r-version"></a>
</h3>
<ul>
<li>Recommended: R 4.3.3 or higher</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="id_-further-reading">üìö Further Reading<a class="anchor" aria-label="anchor" href="#id_-further-reading"></a>
</h2>
<p>For those interested in gaining a deeper understanding of the methodologies used in the <code>SplineOmics</code> package, here are some recommended publications:</p>
<ul>
<li><p><strong>Splines</strong>: To learn more about splines, you can refer to this <a href="https://doi.org/10.1186/s12874-019-0666-3" class="external-link">review</a>.</p></li>
<li><p><strong>limma</strong>: To read about the <code>limma</code> R package, you can refer to this <a href="https://doi.org/10.1093/nar/gkv007" class="external-link">publication</a>.</p></li>
<li><p><strong>Hierarchical clustering</strong>: To get information about hierarchical clustering, you can refer to this <a href="https://towardsdatascience.com/understanding-the-concept-of-hierarchical-clustering-technique-c6e8243758ec" class="external-link">web article</a>.</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="id_-getting-help">‚ùì Getting Help<a class="anchor" aria-label="anchor" href="#id_-getting-help"></a>
</h2>
<p>If you encounter a bug or have a suggestion for improving the <code>SplineOmics</code> package, we encourage you to <a href="https://github.com/csbg/SplineOmics/issues" class="external-link">open an issue</a> on our GitHub repository. Before opening a new issue, please check to see if your question or bug has already been reported by another user. This helps avoid duplicate reports and ensures that we can address problems efficiently.</p>
<p>For more detailed questions, discussions, or contributions regarding the package‚Äôs use and development, please refer to the <a href="https://github.com/csbg/SplineOmics/discussions" class="external-link">GitHub Discussions</a> page for <code>SplineOmics</code>.</p>
</div>
<div class="section level2">
<h2 id="id_-contributing">ü§ù Contributing<a class="anchor" aria-label="anchor" href="#id_-contributing"></a>
</h2>
<p>We welcome contributions to the <code>SplineOmics</code> package! Whether you‚Äôre interested in fixing bugs, adding new features, or improving documentation, your help is greatly appreciated.</p>
<p>Here‚Äôs how you can contribute:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Report a Bug or Request a Feature:</strong> If you encounter a bug or have an idea for a new feature, please <a href="https://github.com/csbg/SplineOmics/issues" class="external-link">open an issue</a> on our GitHub repository. Before opening a new issue, check to see if the issue has already been reported or the feature requested by another user.</p></li>
<li><p><strong>Submit a Pull Request:</strong> If you‚Äôve developed a bug fix or a new feature that you‚Äôd like to share, submit a pull request.</p></li>
<li><p><strong>Improve Documentation:</strong> Good documentation is crucial for any project. If you notice missing or incorrect documentation, please feel free to contribute.</p></li>
</ol>
<p>Please adhere to our <a href="CODE_OF_CONDUCT.html">Code of Conduct</a> in all your interactions with the project.</p>
<p>Thank you for considering contributing to <code>SplineOmics</code>. Your efforts are what make the open-source community a fantastic place to learn, inspire, and create.</p>
</div>
<div class="section level2">
<h2 id="id_-feedback">üí¨ Feedback<a class="anchor" aria-label="anchor" href="#id_-feedback"></a>
</h2>
<p>We appreciate your feedback! Besides raising issues, you can provide feedback in the following ways:</p>
<ul>
<li><p><strong>Direct Email</strong>: Send your feedback directly to <a href="mailto:thomas.rauter@plus.ac.at">Thomas Rauter</a>.</p></li>
<li><p><strong>Anonymous Feedback</strong>: Use <a href="https://forms.gle/jocMXSxLf3GrGBdT9" class="external-link">this Google Form</a> to provide anonymous feedback by answering questions.</p></li>
</ul>
<p>Your feedback helps us improve the project and address any issues you may encounter.</p>
</div>
<div class="section level2">
<h2 id="id_-license">üìú License<a class="anchor" aria-label="anchor" href="#id_-license"></a>
</h2>
<p>This package is licensed under the MIT License: <a href="./LICENSE">LICENSE</a></p>
<p>¬© 2024 Thomas Rauter. All rights reserved.</p>
</div>
<div class="section level2">
<h2 id="id_-citation">üéì Citation<a class="anchor" aria-label="anchor" href="#id_-citation"></a>
</h2>
<p>The <code>SplineOmics</code> package is currently not published in a peer-reviewed scientific journal or similar outlet. However, if this package helped you in your work, consider citing this GitHub repository.</p>
<p>To cite this package, you can use the citation information provided in the <a href="./CITATION.cff"><code>CITATION.cff</code></a> file.</p>
<p>You can also generate a citation in various formats using the <code>CITATION.cff</code> file by visiting the top right of this repo and clicking on the ‚ÄúCite this repository‚Äù button.</p>
<p>Also, if you like the package, consider giving the GitHub repository a star. Your support helps us in the continued development and improvement of <code>SplineOmics</code>. Thank you for using our package!</p>
</div>
<div class="section level2">
<h2 id="id_-contributors">üåü Contributors<a class="anchor" aria-label="anchor" href="#id_-contributors"></a>
</h2>
<ul>
<li>
<a href="https://github.com/Thomas-Rauter" class="external-link">Thomas-Rauter</a> - üöÄ Wrote the package, developed the approach together with VSchaepertoens under guidance from nfortelny and skafdasschaf.</li>
<li>
<a href="https://github.com/nfortelny" class="external-link">nfortelny</a> - üß† Principal Investigator, provided guidance and support for the overall approach.</li>
<li>
<a href="https://github.com/skafdasschaf" class="external-link">skafdasschaf</a> - üîß Helped reviewing code, delivered improvement suggestions and scientific guidance to develop the approach.</li>
<li>
<a href="https://github.com/VSchaepertoens" class="external-link">VSchaepertoens</a> - ‚ú® Developed one internal plotting function, as well as some code for the exploratory data analysis plots, and the overall approach together with Thomas-Rauter.</li>
</ul>
</div>
<div class="section level2">
<h2 id="id_-acknowledgements">üôè Acknowledgements<a class="anchor" aria-label="anchor" href="#id_-acknowledgements"></a>
</h2>
<p>This work was carried out in the context of the DigiTherapeutX project, which was funded by the Austrian Science Fund (FWF). The research was conducted under the supervision of Prof.¬†Nikolaus Fortelny, who leads the Computational Systems Biology working group at the Paris Lodron University of Salzburg, Austria. You can find more information about Prof.¬†Fortelny‚Äôs research group <a href="https://www.plus.ac.at/biowissenschaften/der-fachbereich/arbeitsgruppen/fortelny/" class="external-link">here</a>.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing SplineOmics</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Thomas Rauter <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Thomas Rauter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
