<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Glycan-analysis • SplineOmics</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Glycan-analysis">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SplineOmics</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/CHANGELOG.html">Changelog</a></li>
    <li><a class="dropdown-item" href="../articles/design_limma_design_formula.html">Designing a Limma Design Formula</a></li>
    <li><a class="dropdown-item" href="../articles/Docker_permission_denied.html">Docker_permission_denied</a></li>
    <li><a class="dropdown-item" href="../articles/Docker-instructions.html">Docker-instructions</a></li>
    <li><a class="dropdown-item" href="../articles/get-started.html">get-started</a></li>
    <li><a class="dropdown-item" href="../articles/glycan-analysis.html">Glycan-analysis</a></li>
    <li><a class="dropdown-item" href="../articles/limma_result_categories.html">limma_result_categories</a></li>
    <li><a class="dropdown-item" href="../articles/methylation_data_analysis.html">methylation-data-analysis</a></li>
    <li><a class="dropdown-item" href="../articles/PVC_analysis.html">PVC_analysis</a></li>
    <li><a class="dropdown-item" href="../articles/RNA-seq-analysis.html">RNA-seq-analysis</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../articles/CHANGELOG.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/csbg/SplineOmics/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Glycan-analysis</h1>
                        <h4 data-toc-skip class="author">Thomas
Rauter</h4>
            
            <h4 data-toc-skip class="date">13 February, 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/csbg/SplineOmics/blob/HEAD/vignettes/glycan-analysis.Rmd" class="external-link"><code>vignettes/glycan-analysis.Rmd</code></a></small>
      <div class="d-none name"><code>glycan-analysis.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="about-this-vignette">About this vignette<a class="anchor" aria-label="anchor" href="#about-this-vignette"></a>
</h2>
<p>This tutorial intends to showcase and explain the capabilities of the
<strong>SplineOmics</strong> package by walking through a real and
complete glycan analysis example, from start to finish.
<strong>SplineOmics</strong> is explained in more detail in the
<strong>get-started</strong> vignette, where a proteomics example is
covered. This vignette is more focused on showing how glycan data can be
used, and because of this, less details about the overall package are
provided here.</p>
<div class="section level3">
<h3 id="data-overview">Data Overview<a class="anchor" aria-label="anchor" href="#data-overview"></a>
</h3>
<p>This dataset originates from a <strong>time-series glycan
experiment</strong> designed to study Chinese Hamster Ovary (CHO) cells.
The data is compositional, meaning that the observed glycoform
abundances represent parts of a whole and are constrained by a
constant-sum constraint, making standard statistical analyses
inappropriate without appropriate log-ratio transformations. The
experiment involved cultivating cells in <strong>eight
bioreactors</strong>, with <strong>four bioreactors</strong> subjected
to a temperature shift after 146 hours (experimental condition) and the
remaining <strong>four bioreactors</strong> maintained without a
temperature shift (control condition).</p>
<div class="section level4">
<h4 id="timepoints">Timepoints<a class="anchor" aria-label="anchor" href="#timepoints"></a>
</h4>
<p>Samples were collected at <strong>7 distinct time points</strong>
throughout the experiment, specifically: <code>"120h"</code>,
<code>"144h"</code>, <code>"168h"</code>, <code>"192h"</code>,
<code>"216h"</code>, <code>"288h"</code>, and <code>"336h"</code> after
cultivation start. Each time point was sampled from all eight
bioreactors, but E17_336 is missing, therefore resulting in a total of
<strong>55 samples</strong>.</p>
</div>
<div class="section level4">
<h4 id="effects-in-the-experiment-reactor-and-batch">Effects in the Experiment: Reactor and Batch<a class="anchor" aria-label="anchor" href="#effects-in-the-experiment-reactor-and-batch"></a>
</h4>
<p>In this experiment, there are two effects to consider:
<strong>Reactor</strong> and <strong>Batch</strong>.</p>
<ol style="list-style-type: decimal">
<li>
<strong>Reactor</strong>:
<ul>
<li>This refers to the different bioreactors used for cell cultivation,
which can exhibit substantial variability.</li>
<li>Each reactor was assigned a single condition: either constant
temperature or temperature-shifted. As a result, <strong>condition and
reactor are confounded</strong>.</li>
<li>Reactor should not be treated as a fixed effect to simply remove its
influence. Instead, it is treated as a <strong>random effect</strong>,
which allows us to model its variability appropriately.</li>
</ul>
</li>
<li>
<strong>Batch</strong>:
<ul>
<li>This refers to the two separate glycan analysis batches.</li>
<li>Batch is considered a <strong>batch effect</strong> with respect to
the condition (constant temperature vs. temperature-shifted).</li>
</ul>
</li>
</ol>
<p>Since Condition and Reactor are confounded, the variability due to
the reactors cannot be directly separated from the condition. Instead,
<strong>linear mixed models (LMMs)</strong> are used to attribute the
reactor as a random effect, allowing us to account for its variability
while isolating the effects of the condition. This approach ensures that
the analysis appropriately handles the hierarchical structure of the
data and avoids incorrect conclusions.</p>
<p>In this vignette, we will demonstrate how to use linear mixed models
to address these challenges and properly account for both reactor and
plate effects.</p>
</div>
<div class="section level4">
<h4 id="further-info">Further info<a class="anchor" aria-label="anchor" href="#further-info"></a>
</h4>
<p>The data matrix comprises <strong>glycoforms as rows</strong> (such
as G0/G0 and none/G0F) and <strong>samples as columns</strong>,
providing glycoform measurements for all time points. Each glycoform as
a row stands for a combination of sugars (glycans) that can be attached
to the left and right side of the product antibody, that we produced in
our CHO cell cultivation. For example G0/G0 means that the glycan G0 was
attached to both sides, and none/G0F means at the left side, there was
no glycan, and on the right side, there was the G0F glycan.</p>
<p>The goal of this experiment is to investigate the effect of a
temperature shift during CHO cell cultivation on the antibody glycan
dynamics over time.</p>
<p>Note: This is not the original dataset, as it has not yet been
published at the time of this vignette’s creation. For demonstration
purposes, the glycans have been randomly shuffled, and 2% has been
randomly added or substracted off each value.</p>
</div>
</div>
<div class="section level3">
<h3 id="analysis-goals">Analysis Goals<a class="anchor" aria-label="anchor" href="#analysis-goals"></a>
</h3>
<p>The main objectives of this analysis are:</p>
<ul>
<li><p><strong>Identify glycans with significant temporal
changes</strong>: Among the glycoforms measured, the goal is to identify
those that exhibit significant changes in abundance over time.</p></li>
<li><p><strong>Cluster glycans based on temporal patterns</strong>:
Glycoforms showing significant temporal changes (hits) will be grouped
into clusters based on their time-dependent expression
patterns.</p></li>
<li><p><strong>Assess the impact of temperature shifts on temporal
patterns</strong>: The analysis will determine whether the temporal
patterns of glycoform abundance are affected by the temperature shift,
i.e., whether glycoform abundance dynamics differ over time under
temperature shift conditions compared to controls.</p></li>
</ul>
<div class="section level4">
<h4 id="note">Note<a class="anchor" aria-label="anchor" href="#note"></a>
</h4>
<p>The documentation of all the <strong>SplineOmics</strong> package
functions can be viewed <a href="https://csbg.github.io/SplineOmics/reference">here</a></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="load-the-packages">Load the packages<a class="anchor" aria-label="anchor" href="#load-the-packages"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://csbg.github.io/SplineOmics">SplineOmics</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://here.r-lib.org/" class="external-link">here</a></span><span class="op">)</span> <span class="co"># For managing filepaths</span></span>
<span><span class="co">#&gt; here() starts at /home/thomas/Documents/PhD/projects/DGTX/SplineOmics_hub/SplineOmics</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span> <span class="co"># For data manipulation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yihui.org/knitr/" class="external-link">knitr</a></span><span class="op">)</span> <span class="co"># For Showing the head of the data and the meta tables.</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.stat.boogaart.de/compositions/" class="external-link">compositions</a></span><span class="op">)</span>  <span class="co"># For clr transforming the glycan data</span></span>
<span><span class="co">#&gt; Welcome to compositions, a package for compositional data analysis.</span></span>
<span><span class="co">#&gt; Find an intro with "? compositions"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'compositions'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anova, cor, cov, dist, var</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:graphics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     segments</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     %*%, norm, scale, scale.default</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-the-files">Load the files<a class="anchor" aria-label="anchor" href="#load-the-files"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>    <span class="st">"extdata"</span>,</span>
<span>    <span class="st">"glycan_data.csv"</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"SplineOmics"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>    <span class="st">"extdata"</span>,</span>
<span>    <span class="st">"glycan_meta.csv"</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"SplineOmics"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set the first column as row names and remove it from the dataframe</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>  <span class="co"># Assign first column as row names</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span>  <span class="co"># Remove the first column</span></span>
<span></span>
<span></span>
<span><span class="co"># Make data to a numeric matrix (required by SplineOmics)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.matrix.html" class="external-link">data.matrix</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>   </span></code></pre></div>
<div class="section level3">
<h3 id="show-top-rows-of-data">Show top rows of data<a class="anchor" aria-label="anchor" href="#show-top-rows-of-data"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>,</span>
<span>  format <span class="op">=</span> <span class="st">"markdown"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<table style="width:100%;" class="table">
<colgroup>
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
<col width="1%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="right">E13_120</th>
<th align="right">E13_144</th>
<th align="right">E13_168</th>
<th align="right">E13_192</th>
<th align="right">E13_216</th>
<th align="right">E13_288</th>
<th align="right">E13_336</th>
<th align="right">E14_120</th>
<th align="right">E14_144</th>
<th align="right">E14_168</th>
<th align="right">E14_192</th>
<th align="right">E14_216</th>
<th align="right">E14_288</th>
<th align="right">E14_336</th>
<th align="right">E15_120</th>
<th align="right">E15_144</th>
<th align="right">E15_168</th>
<th align="right">E15_192</th>
<th align="right">E15_216</th>
<th align="right">E15_288</th>
<th align="right">E15_336</th>
<th align="right">E16_120</th>
<th align="right">E16_144</th>
<th align="right">E16_168</th>
<th align="right">E16_192</th>
<th align="right">E16_216</th>
<th align="right">E16_288</th>
<th align="right">E16_336</th>
<th align="right">E17_120</th>
<th align="right">E17_144</th>
<th align="right">E17_168</th>
<th align="right">E17_192</th>
<th align="right">E17_216</th>
<th align="right">E17_288</th>
<th align="right">E18_120</th>
<th align="right">E18_144</th>
<th align="right">E18_168</th>
<th align="right">E18_192</th>
<th align="right">E18_216</th>
<th align="right">E18_288</th>
<th align="right">E18_336</th>
<th align="right">E19_120</th>
<th align="right">E19_144</th>
<th align="right">E19_168</th>
<th align="right">E19_192</th>
<th align="right">E19_216</th>
<th align="right">E19_288</th>
<th align="right">E19_336</th>
<th align="right">E20_120</th>
<th align="right">E20_144</th>
<th align="right">E20_168</th>
<th align="right">E20_192</th>
<th align="right">E20_216</th>
<th align="right">E20_288</th>
<th align="right">E20_336</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">none/G0F</td>
<td align="right">1.0558960</td>
<td align="right">1.6535404</td>
<td align="right">0.6493199</td>
<td align="right">0.4020102</td>
<td align="right">0.2991287</td>
<td align="right">0.2952674</td>
<td align="right">0.4744168</td>
<td align="right">1.0534692</td>
<td align="right">1.093678</td>
<td align="right">0.5601851</td>
<td align="right">0.5350867</td>
<td align="right">0.3788699</td>
<td align="right">0.4035115</td>
<td align="right">0.3757274</td>
<td align="right">0.8211157</td>
<td align="right">0.8911038</td>
<td align="right">0.3801903</td>
<td align="right">0.4426778</td>
<td align="right">0.3843382</td>
<td align="right">0.3318367</td>
<td align="right">0.3398791</td>
<td align="right">0.7996867</td>
<td align="right">1.571295</td>
<td align="right">1.4926070</td>
<td align="right">0.4777732</td>
<td align="right">0.6603077</td>
<td align="right">0.3305801</td>
<td align="right">0.2346606</td>
<td align="right">0.7880709</td>
<td align="right">1.0201295</td>
<td align="right">0.7577885</td>
<td align="right">0.1905728</td>
<td align="right">0.3328567</td>
<td align="right">0.2802764</td>
<td align="right">0.7150925</td>
<td align="right">0.7753927</td>
<td align="right">0.9080968</td>
<td align="right">0.3811781</td>
<td align="right">0.2896830</td>
<td align="right">0.3272081</td>
<td align="right">0.2351659</td>
<td align="right">1.3833084</td>
<td align="right">0.9376732</td>
<td align="right">1.1173998</td>
<td align="right">0.6469200</td>
<td align="right">0.3043375</td>
<td align="right">0.2228878</td>
<td align="right">0.3791463</td>
<td align="right">1.0784873</td>
<td align="right">0.8032912</td>
<td align="right">0.3756631</td>
<td align="right">0.3559814</td>
<td align="right">0.2675088</td>
<td align="right">0.3162241</td>
<td align="right">0.2643130</td>
</tr>
<tr class="even">
<td align="left">none/G1F</td>
<td align="right">0.5323374</td>
<td align="right">0.1912681</td>
<td align="right">0.3820570</td>
<td align="right">0.1920724</td>
<td align="right">0.3369441</td>
<td align="right">0.6384174</td>
<td align="right">0.8548695</td>
<td align="right">0.5733259</td>
<td align="right">0.567492</td>
<td align="right">0.7163753</td>
<td align="right">0.4688813</td>
<td align="right">0.4682832</td>
<td align="right">0.5383108</td>
<td align="right">0.6551322</td>
<td align="right">0.4758381</td>
<td align="right">0.3954215</td>
<td align="right">0.3811519</td>
<td align="right">0.3056876</td>
<td align="right">0.4411230</td>
<td align="right">0.5827720</td>
<td align="right">0.6590168</td>
<td align="right">0.4662977</td>
<td align="right">0.621623</td>
<td align="right">0.4692532</td>
<td align="right">0.4053633</td>
<td align="right">0.4627089</td>
<td align="right">0.5262178</td>
<td align="right">0.4367194</td>
<td align="right">0.2626777</td>
<td align="right">0.4807747</td>
<td align="right">0.4591137</td>
<td align="right">0.0997751</td>
<td align="right">0.2675994</td>
<td align="right">0.7528157</td>
<td align="right">0.3142561</td>
<td align="right">0.2728702</td>
<td align="right">0.4900672</td>
<td align="right">0.3237094</td>
<td align="right">0.3904656</td>
<td align="right">0.5668250</td>
<td align="right">0.4722005</td>
<td align="right">0.5748934</td>
<td align="right">0.6408565</td>
<td align="right">0.5941058</td>
<td align="right">0.3463255</td>
<td align="right">0.2640809</td>
<td align="right">0.4779125</td>
<td align="right">0.8081467</td>
<td align="right">0.3307047</td>
<td align="right">0.3004908</td>
<td align="right">0.2121393</td>
<td align="right">0.3430701</td>
<td align="right">0.3647285</td>
<td align="right">0.5845660</td>
<td align="right">0.6514764</td>
</tr>
<tr class="odd">
<td align="left">none/G2F</td>
<td align="right">1.5324508</td>
<td align="right">1.8521343</td>
<td align="right">0.8379884</td>
<td align="right">0.7293874</td>
<td align="right">0.4902723</td>
<td align="right">0.5382350</td>
<td align="right">0.6436627</td>
<td align="right">1.5629840</td>
<td align="right">1.289147</td>
<td align="right">0.9681731</td>
<td align="right">0.8381506</td>
<td align="right">0.7008038</td>
<td align="right">0.6532061</td>
<td align="right">0.6719115</td>
<td align="right">1.2138739</td>
<td align="right">1.1464261</td>
<td align="right">0.8146055</td>
<td align="right">0.7811525</td>
<td align="right">0.6912260</td>
<td align="right">0.5626390</td>
<td align="right">0.4371626</td>
<td align="right">1.2651715</td>
<td align="right">1.989103</td>
<td align="right">1.5335954</td>
<td align="right">0.7930346</td>
<td align="right">0.9061705</td>
<td align="right">0.4856366</td>
<td align="right">0.4323194</td>
<td align="right">1.4655321</td>
<td align="right">1.3453025</td>
<td align="right">1.1834043</td>
<td align="right">0.5523239</td>
<td align="right">0.5450371</td>
<td align="right">0.5750546</td>
<td align="right">1.0783679</td>
<td align="right">1.1487218</td>
<td align="right">1.1215321</td>
<td align="right">0.6694769</td>
<td align="right">0.5519309</td>
<td align="right">0.5116042</td>
<td align="right">0.4199132</td>
<td align="right">1.7598936</td>
<td align="right">1.1849219</td>
<td align="right">1.3638929</td>
<td align="right">0.7197493</td>
<td align="right">0.5763007</td>
<td align="right">0.2844577</td>
<td align="right">0.5380468</td>
<td align="right">1.5920806</td>
<td align="right">1.3495185</td>
<td align="right">0.8517591</td>
<td align="right">0.6509213</td>
<td align="right">0.6691747</td>
<td align="right">0.6211062</td>
<td align="right">0.6178701</td>
</tr>
<tr class="even">
<td align="left">G0/G0</td>
<td align="right">12.4576400</td>
<td align="right">10.2160346</td>
<td align="right">10.2302200</td>
<td align="right">6.5323299</td>
<td align="right">4.5557592</td>
<td align="right">4.6899314</td>
<td align="right">4.7203765</td>
<td align="right">12.3929924</td>
<td align="right">10.510602</td>
<td align="right">11.4393911</td>
<td align="right">9.4008800</td>
<td align="right">9.4694548</td>
<td align="right">8.1328422</td>
<td align="right">7.6767012</td>
<td align="right">12.1137799</td>
<td align="right">11.9553981</td>
<td align="right">11.2695201</td>
<td align="right">9.3657469</td>
<td align="right">8.0426315</td>
<td align="right">6.5983448</td>
<td align="right">5.2429438</td>
<td align="right">12.0570776</td>
<td align="right">10.660712</td>
<td align="right">11.3814227</td>
<td align="right">10.8993364</td>
<td align="right">9.5825590</td>
<td align="right">5.9291533</td>
<td align="right">4.4084289</td>
<td align="right">10.3195660</td>
<td align="right">8.3433472</td>
<td align="right">9.5614397</td>
<td align="right">7.4586037</td>
<td align="right">6.6427808</td>
<td align="right">8.4710449</td>
<td align="right">11.0041856</td>
<td align="right">9.8884089</td>
<td align="right">10.6157702</td>
<td align="right">8.8851596</td>
<td align="right">7.7355069</td>
<td align="right">6.4930713</td>
<td align="right">6.3931685</td>
<td align="right">10.3100238</td>
<td align="right">8.9498072</td>
<td align="right">9.7648715</td>
<td align="right">7.6610699</td>
<td align="right">6.2008262</td>
<td align="right">3.8324832</td>
<td align="right">3.4344767</td>
<td align="right">10.5438400</td>
<td align="right">8.6848225</td>
<td align="right">9.8761378</td>
<td align="right">8.7988318</td>
<td align="right">8.1356305</td>
<td align="right">7.7008655</td>
<td align="right">8.5356119</td>
</tr>
<tr class="odd">
<td align="left">G0/G0F</td>
<td align="right">34.1841189</td>
<td align="right">38.3372702</td>
<td align="right">41.0179628</td>
<td align="right">49.0643714</td>
<td align="right">57.0485903</td>
<td align="right">55.2748321</td>
<td align="right">51.4400433</td>
<td align="right">35.7867219</td>
<td align="right">39.951440</td>
<td align="right">41.3072610</td>
<td align="right">45.0540699</td>
<td align="right">45.5449446</td>
<td align="right">46.7403271</td>
<td align="right">46.3506974</td>
<td align="right">37.0730784</td>
<td align="right">39.6951101</td>
<td align="right">40.6190535</td>
<td align="right">46.7785275</td>
<td align="right">48.3297400</td>
<td align="right">55.0964653</td>
<td align="right">55.9160088</td>
<td align="right">34.3909667</td>
<td align="right">35.142868</td>
<td align="right">38.1036796</td>
<td align="right">41.7931083</td>
<td align="right">44.5759546</td>
<td align="right">56.0226674</td>
<td align="right">63.5428148</td>
<td align="right">36.5324214</td>
<td align="right">40.7206353</td>
<td align="right">41.4229164</td>
<td align="right">48.7461735</td>
<td align="right">52.2302885</td>
<td align="right">49.1391718</td>
<td align="right">40.4162712</td>
<td align="right">43.1801464</td>
<td align="right">42.3152053</td>
<td align="right">47.0698711</td>
<td align="right">48.8438299</td>
<td align="right">50.1411749</td>
<td align="right">51.0158271</td>
<td align="right">35.9886203</td>
<td align="right">40.0629781</td>
<td align="right">40.8438051</td>
<td align="right">46.0248557</td>
<td align="right">52.3444655</td>
<td align="right">58.5929353</td>
<td align="right">57.2609213</td>
<td align="right">36.2618340</td>
<td align="right">41.6920256</td>
<td align="right">42.9619517</td>
<td align="right">45.8755552</td>
<td align="right">45.5816057</td>
<td align="right">45.9113688</td>
<td align="right">46.1812359</td>
</tr>
<tr class="even">
<td align="left">G0F/G0F</td>
<td align="right">1.5218235</td>
<td align="right">0.7862833</td>
<td align="right">1.1317116</td>
<td align="right">0.3237543</td>
<td align="right">-0.1691198</td>
<td align="right">0.9544022</td>
<td align="right">1.2641091</td>
<td align="right">1.8948004</td>
<td align="right">1.654238</td>
<td align="right">1.2341346</td>
<td align="right">1.2849757</td>
<td align="right">1.1123587</td>
<td align="right">1.1001650</td>
<td align="right">0.6284302</td>
<td align="right">1.5736525</td>
<td align="right">1.3480386</td>
<td align="right">1.6721168</td>
<td align="right">1.0828775</td>
<td align="right">0.8432043</td>
<td align="right">0.8164667</td>
<td align="right">1.0990374</td>
<td align="right">2.2216793</td>
<td align="right">1.773861</td>
<td align="right">1.5704807</td>
<td align="right">1.3096889</td>
<td align="right">1.2557005</td>
<td align="right">0.8205026</td>
<td align="right">0.5531342</td>
<td align="right">1.1133060</td>
<td align="right">0.8061578</td>
<td align="right">1.1689967</td>
<td align="right">0.3151786</td>
<td align="right">0.2796022</td>
<td align="right">-0.1211154</td>
<td align="right">1.1292473</td>
<td align="right">0.9209109</td>
<td align="right">1.3967621</td>
<td align="right">1.0388824</td>
<td align="right">0.7251202</td>
<td align="right">0.7248511</td>
<td align="right">0.5504109</td>
<td align="right">1.3079025</td>
<td align="right">0.7601992</td>
<td align="right">1.1123447</td>
<td align="right">0.5752250</td>
<td align="right">0.4825608</td>
<td align="right">0.0937523</td>
<td align="right">0.2893026</td>
<td align="right">1.2427219</td>
<td align="right">0.7371158</td>
<td align="right">0.8688332</td>
<td align="right">0.8953497</td>
<td align="right">0.6032681</td>
<td align="right">0.8568143</td>
<td align="right">1.0963033</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="show-top-rows-of-meta">Show top rows of meta<a class="anchor" aria-label="anchor" href="#show-top-rows-of-meta"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">meta</span><span class="op">)</span>,</span>
<span>  format <span class="op">=</span> <span class="st">"markdown"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">sample_name</th>
<th align="left">Reactor</th>
<th align="right">Time</th>
<th align="left">Condition</th>
<th align="right">Batch</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">E13_120</td>
<td align="left">E13</td>
<td align="right">120</td>
<td align="left">constant</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">E13_144</td>
<td align="left">E13</td>
<td align="right">144</td>
<td align="left">constant</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">E13_168</td>
<td align="left">E13</td>
<td align="right">168</td>
<td align="left">constant</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">E13_192</td>
<td align="left">E13</td>
<td align="right">192</td>
<td align="left">constant</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">E13_216</td>
<td align="left">E13</td>
<td align="right">216</td>
<td align="left">constant</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">E13_288</td>
<td align="left">E13</td>
<td align="right">288</td>
<td align="left">constant</td>
<td align="right">1</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level2">
<h2 id="perform-eda-exploratory-data-analysis">Perform EDA (exploratory data analysis)<a class="anchor" aria-label="anchor" href="#perform-eda-exploratory-data-analysis"></a>
</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Those fields are mandatory, because we believe that when such a report is</span></span>
<span><span class="co"># opened after half a year, those infos can be very helpful.</span></span>
<span><span class="va">report_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  omics_data_type <span class="op">=</span> <span class="st">"Glycan"</span>,</span>
<span>  data_description <span class="op">=</span> <span class="st">"clr transformed timeseries fractional abundance glycoform data of CHO cells, batch corrected"</span>,</span>
<span>  data_collection_date <span class="op">=</span> <span class="st">"September 2024"</span>,</span>
<span>  analyst_name <span class="op">=</span> <span class="st">"Thomas Rauter"</span>,</span>
<span>  contact_info <span class="op">=</span> <span class="st">"thomas.rauter@plus.ac.at"</span>,</span>
<span>  project_name <span class="op">=</span> <span class="st">"DGTX"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">report_dir</span> <span class="op">&lt;-</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html" class="external-link">here</a></span><span class="op">(</span></span>
<span>  <span class="st">"results"</span>,</span>
<span>  <span class="st">"explore_data"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Compositional data, such as glycan profiles expressed as relative
abundances, are inherently constrained by a constant sum and therefore
reside in a simplex rather than Euclidean space. This violates the
assumptions of standard statistical methods. To address this, we applied
a centered log-ratio (CLR) transformation, which appropriately maps the
data into real space by accounting for the compositional structure.
While some studies apply a simple log-transformation to stabilize
variance, this approach ignores the relative nature of the data and may
lead to biased or misleading results. CLR transformation is thus the
more appropriate and statistically sound choice for downstream
modeling.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># TRANSPOSE: make samples rows, features columns</span></span>
<span><span class="va">data_t</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Handle zeros (simple pseudocount, crude but common)</span></span>
<span><span class="va">data_t</span><span class="op">[</span><span class="va">data_t</span> <span class="op">==</span> <span class="fl">0</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1e-6</span></span>
<span></span>
<span><span class="co"># Apply CLR</span></span>
<span><span class="va">clr_data_t</span> <span class="op">&lt;-</span> <span class="fu">compositions</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/compositions/man/clr.html" class="external-link">clr</a></span><span class="op">(</span><span class="va">data_t</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Transpose back to original shape: features as rows, samples as columns</span></span>
<span><span class="va">clr_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">clr_data_t</span><span class="op">)</span></span>
<span><span class="va">clr_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">unclass</a></span><span class="op">(</span><span class="va">clr_data</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># splineomics now contains the SplineOmics object.</span></span>
<span><span class="va">splineomics</span> <span class="op">&lt;-</span> <span class="fu">SplineOmics</span><span class="fu">::</span><span class="fu"><a href="../reference/create_splineomics.html">create_splineomics</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">clr_data</span>,</span>
<span>  meta <span class="op">=</span> <span class="va">meta</span>,</span>
<span>  report_info <span class="op">=</span> <span class="va">report_info</span>,</span>
<span>  condition <span class="op">=</span> <span class="st">"Condition"</span>, <span class="co"># Column of meta that contains the levels.</span></span>
<span>  meta_batch_column <span class="op">=</span> <span class="st">"Batch"</span> <span class="co"># For batch effect removal in the plots</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Special print.SplineOmics function leads to selective printing</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">splineomics</span><span class="op">)</span></span>
<span><span class="co">#&gt; data:SplineOmics Object</span></span>
<span><span class="co">#&gt; -------------------</span></span>
<span><span class="co">#&gt; Number of features (rows): 10 </span></span>
<span><span class="co">#&gt; Number of samples (columns): 55 </span></span>
<span><span class="co">#&gt; Meta data columns: 5 </span></span>
<span><span class="co">#&gt; First few meta columns:</span></span>
<span><span class="co">#&gt;   sample_name Reactor Time Condition Batch</span></span>
<span><span class="co">#&gt; 1     E13_120     E13  120  constant     1</span></span>
<span><span class="co">#&gt; 2     E13_144     E13  144  constant     1</span></span>
<span><span class="co">#&gt; 3     E13_168     E13  168  constant     1</span></span>
<span><span class="co">#&gt; Condition: Condition </span></span>
<span><span class="co">#&gt; No RNA-seq data provided.</span></span>
<span><span class="co">#&gt; No annotation provided.</span></span>
<span><span class="co">#&gt; No spline parameters set.</span></span>
<span><span class="co">#&gt; P-value adjustment method: BH</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu">SplineOmics</span><span class="fu">::</span><span class="fu"><a href="../reference/explore_data.html">explore_data</a></span><span class="op">(</span></span>
<span>  splineomics <span class="op">=</span> <span class="va">splineomics</span>, <span class="co"># SplineOmics object</span></span>
<span>  report_dir <span class="op">=</span> <span class="va">report_dir</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><a href="https://drive.google.com/uc?id=1h4aOSwxUHAp5KAoAYVOlL_IRotV4EtY0" class="external-link">Here</a>
you can see the HTML report of the explore_data() function with the NOT
batch-corrected data, and <a href="https://drive.google.com/uc?id=1op3oCt7eNhrUaofIkAyNJrSqpschU78b" class="external-link">here</a>
the report for the batch-corrected data.</p>
</div>
<div class="section level2">
<h2 id="run-limma-spline-analysis">Run limma spline analysis<a class="anchor" aria-label="anchor" href="#run-limma-spline-analysis"></a>
</h2>
<p>In this example, we are skipping finding the best hyperparameters
with the screen_limma_hyperparams() function, because we already have a
clear idea of what do use.</p>
<p>Lets define our parameters and put them into the
<code>SplineOmics</code> object:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">splineomics</span> <span class="op">&lt;-</span> <span class="fu">SplineOmics</span><span class="fu">::</span><span class="fu"><a href="../reference/update_splineomics.html">update_splineomics</a></span><span class="op">(</span></span>
<span>  splineomics <span class="op">=</span> <span class="va">splineomics</span>,</span>
<span>  use_array_weights <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="co"># Reactor as random effect.</span></span>
<span>  design <span class="op">=</span> <span class="st">"~ 1 + Condition*Time + Batch + (1|Reactor)"</span>, </span>
<span>  mode <span class="op">=</span> <span class="st">"integrated"</span>, <span class="co"># means limma uses the full data for each condition.</span></span>
<span>  spline_params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    spline_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"n"</span><span class="op">)</span>, <span class="co"># natural cubic splines (take these if unsure)</span></span>
<span>    dof <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2L</span><span class="op">)</span> <span class="co"># If you are unsure about which dof, start with 2 and increase</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The term ‘(1|Reactor)’ in the design formula are the random effects.
Therefore, the linear mixed models that we use will model Reactor as a
random effect.</p>
<p>Run the <code><a href="../reference/run_limma_splines.html">run_limma_splines()</a></code> function with the updated
SplineOmics object:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">splineomics</span> <span class="op">&lt;-</span> <span class="fu">SplineOmics</span><span class="fu">::</span><span class="fu"><a href="../reference/run_limma_splines.html">run_limma_splines</a></span><span class="op">(</span></span>
<span>  splineomics <span class="op">=</span> <span class="va">splineomics</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Hint: The data contains negative values. This may occur if the data has been transformed (e.g. log-transformed or normalized) and is valid in such cases. Ensure that the data preprocessing aligns with your analysis requirements.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Make sure that the design formula contains no interaction between the condition and time for mode == isolated, and that it contains an interaction for mode == integrated. Otherwise, you will get an uncaught error of 'coefficients not estimable' or 'subscript out of bounds' (or any other hard to understand error).</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Fitting global model...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; NOTE: If you manually stop run_limma_splines() in RStudio and used, parallelization for variancePartition::dream(), then those parallelized processes may continue running. Use your system's process manager to terminate them manually!</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">Info</span> Finished limma spline analysis in 0.0 min</span></span></code></pre></div>
<p>Note that for our dataset, the Wilcoxon signed rank exact test
revealed that we have a problem of heteroscedasticity for our data. This
means that the variance between the samples is significantly higher in
the stationary phase compared to the exponential phase. This is a
violation for the assumption of homoscedasticity of linear models, which
means that the derived p-values are not trustworthy. To adress this
issue, limma has a function (arrayWeights) which downweights samples
with higher variance, which we then use here automatically. Further, the
eBayes function of limma with the robust = TRUE argument avoids using
variance “outliers” for the variance shrinkage.</p>
</div>
<div class="section level2">
<h2 id="build-limma-report">Build limma report<a class="anchor" aria-label="anchor" href="#build-limma-report"></a>
</h2>
<p>The topTables of all three limma result categories can be used to
generate p-value histograms an volcano plots.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu">SplineOmics</span><span class="fu">::</span><span class="fu"><a href="../reference/create_limma_report.html">create_limma_report</a></span><span class="op">(</span></span>
<span>  splineomics <span class="op">=</span> <span class="va">splineomics</span>,</span>
<span>  report_dir <span class="op">=</span> <span class="fu">withr</span><span class="fu">::</span><span class="fu"><a href="https://withr.r-lib.org/reference/with_tempfile.html" class="external-link">local_tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  <span style="color: #00BB00;">Info</span> Limma report generation completed successfully.</span></span>
<span><span class="co">#&gt;  Your HTML reports are located in the directory:  /tmp/RtmpAlTTrf/file5b2f466b6550 .</span></span>
<span><span class="co">#&gt;  Please note that due to embedded files, the reports might be flagged as</span></span>
<span><span class="co">#&gt;  harmful by other software. Rest assured that they provide no harm.</span></span></code></pre></div>
<p>You can view the generated analysis report of the create_limma_report
function <a href="https://drive.google.com/uc?id=1jgE8d5Gs12xnMATaitXxp28j2nAKCX1Y" class="external-link">here</a>.</p>
</div>
<div class="section level2">
<h2 id="cluster-the-hits-significant-features">Cluster the hits (significant features)<a class="anchor" aria-label="anchor" href="#cluster-the-hits-significant-features"></a>
</h2>
<p>After we obtained the limma spline results, we can cluster the hits
based on their temporal pattern (their spline shape). We define what a
hit is by setting an adj. p-value threshold for every level. Hits are
features (e.g. proteins) that have an adj. p-value below the threshold.
Hierarchical clustering is used to place every hit in one of as many
clusters as we have specified for that specific level.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">adj_pthresholds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="co"># 0.05 for both levels</span></span>
<span>  <span class="fl">0.05</span>, <span class="co"># exponential</span></span>
<span>  <span class="fl">0.05</span> <span class="co"># stationary</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">nr_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  constant <span class="op">=</span> <span class="fl">2</span>, </span>
<span>  tshifted <span class="op">=</span> <span class="fl">2</span>   </span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  y_axis_label <span class="op">=</span> <span class="st">"Glycan fractional abundance"</span>,</span>
<span>  time_unit <span class="op">=</span> <span class="st">"hours"</span>,</span>
<span>  treatment_labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    tshifted <span class="op">=</span> <span class="st">"temp shift"</span>,  </span>
<span>    double_spline_plots <span class="op">=</span> <span class="st">"temp shift"</span> </span>
<span>    <span class="op">)</span>,</span>
<span>  treatment_timepoints <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    tshifted <span class="op">=</span> <span class="fl">146</span>,  </span>
<span>    double_spline_plots <span class="op">=</span> <span class="fl">146</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Those are not genes for this glycan analysis here, but this argument expects</span></span>
<span><span class="co"># the feature names (which usually are gene names, which is why it is called </span></span>
<span><span class="co"># like this.</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot_options</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="co"># When meta_replicate_column is not there, all datapoints are blue.</span></span>
<span>  meta_replicate_column <span class="op">=</span> <span class="st">"Reactor"</span>, <span class="co"># Colors the data points based on Reactor</span></span>
<span>  cluster_heatmap_columns <span class="op">=</span> <span class="cn">FALSE</span> </span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">clustering_results</span> <span class="op">&lt;-</span> <span class="fu">SplineOmics</span><span class="fu">::</span><span class="fu"><a href="../reference/cluster_hits.html">cluster_hits</a></span><span class="op">(</span></span>
<span>  splineomics <span class="op">=</span> <span class="va">splineomics</span>,</span>
<span>  adj_pthresholds <span class="op">=</span> <span class="va">adj_pthresholds</span>,</span>
<span>  nr_clusters <span class="op">=</span> <span class="va">nr_clusters</span>,</span>
<span>  genes <span class="op">=</span> <span class="va">genes</span>,</span>
<span>  plot_info <span class="op">=</span> <span class="va">plot_info</span>,</span>
<span>  plot_options <span class="op">=</span> <span class="va">plot_options</span>,</span>
<span>  report_dir <span class="op">=</span> <span class="fu">withr</span><span class="fu">::</span><span class="fu"><a href="https://withr.r-lib.org/reference/with_tempfile.html" class="external-link">local_tempdir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  adj_pthresh_avrg_diff_conditions <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  adj_pthresh_interaction_condition_time <span class="op">=</span> <span class="fl">0.05</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Hint: The data contains negative values. This may occur if the data has been transformed (e.g. log-transformed or normalized) and is valid in such cases. Ensure that the data preprocessing aligns with your analysis requirements.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Make sure that the design formula contains no interaction between the condition and time for mode == isolated, and that it contains an interaction for mode == integrated. Otherwise, you will get an uncaught error of 'coefficients not estimable' or 'subscript out of bounds' (or any other hard to understand error).</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Performing the clustering...</span></span>
<span><span class="co">#&gt; For the level:  constant</span></span>
<span><span class="co">#&gt; For the level:  tshifted</span></span>
<span><span class="co">#&gt; Generating heatmap...</span></span>
<span><span class="co">#&gt; Generating cluster mean splines for level:  constant</span></span>
<span><span class="co">#&gt; Generating spline plots...</span></span>
<span><span class="co">#&gt; Generating cluster mean splines for level:  tshifted</span></span>
<span><span class="co">#&gt; Generating spline plots...</span></span>
<span><span class="co">#&gt; Generating report. This takes a few seconds.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  <span style="color: #00BB00;">Info</span> Clustering the hits completed successfully.</span></span>
<span><span class="co">#&gt;  Your HTML reports are located in the directory:  /tmp/RtmpAlTTrf/file5b2f77c3432f .</span></span>
<span><span class="co">#&gt;  Please note that due to embedded files, the reports might be flagged as</span></span>
<span><span class="co">#&gt;  harmful by other software. Rest assured that they provide no harm.</span></span>
<span><span class="co">#&gt; 0 clusters for the condition effect (interaction between condition and time)</span></span>
<span><span class="co">#&gt; Running this function took 0.1 min</span></span></code></pre></div>
<p>You can view the generated analysis report of the cluster_hits
function <a href="https://drive.google.com/uc?id=1HXKoyOniO6-tDC-uCiGgpaQcjUQ9-zyl" class="external-link">here</a>.</p>
<p>As discussed before, there are three limma result categories. The
cluster_hits() report shows the results of all three, if they are
present (category 2 and 3 can only be generated when the design formula
contains an interaction effect).</p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=de_AT.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=de_AT.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=de_AT.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=de_AT.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Europe/Vienna</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices datasets  utils     methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] compositions_2.0-9 knitr_1.50         dplyr_1.1.4        here_1.0.2        </span></span>
<span><span class="co">#&gt; [5] SplineOmics_0.3.4 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RColorBrewer_1.1-3       tensorA_0.36.2.1         rstudioapi_0.17.1       </span></span>
<span><span class="co">#&gt;   [4] jsonlite_2.0.0           shape_1.4.6.1            magrittr_2.0.4          </span></span>
<span><span class="co">#&gt;   [7] farver_2.1.2             nloptr_2.2.1             rmarkdown_2.29          </span></span>
<span><span class="co">#&gt;  [10] GlobalOptions_0.1.2      fs_1.6.6                 ragg_1.4.0              </span></span>
<span><span class="co">#&gt;  [13] vctrs_0.6.5              minqa_1.2.8              base64enc_0.1-3         </span></span>
<span><span class="co">#&gt;  [16] htmltools_0.5.8.1        progress_1.2.3           broom_1.0.10            </span></span>
<span><span class="co">#&gt;  [19] Formula_1.2-5            variancePartition_1.38.1 sass_0.4.10             </span></span>
<span><span class="co">#&gt;  [22] KernSmooth_2.23-26       bslib_0.9.0              htmlwidgets_1.6.4       </span></span>
<span><span class="co">#&gt;  [25] desc_1.4.3               pbkrtest_0.5.5           plyr_1.8.9              </span></span>
<span><span class="co">#&gt;  [28] cachem_1.1.0             lifecycle_1.0.4          iterators_1.0.14        </span></span>
<span><span class="co">#&gt;  [31] pkgconfig_2.0.3          Matrix_1.7-4             R6_2.6.1                </span></span>
<span><span class="co">#&gt;  [34] fastmap_1.2.0            rbibutils_2.3            clue_0.3-66             </span></span>
<span><span class="co">#&gt;  [37] digest_0.6.37            numDeriv_2016.8-1.1      colorspace_2.1-1        </span></span>
<span><span class="co">#&gt;  [40] S4Vectors_0.46.0         rprojroot_2.1.1          textshaping_1.0.3       </span></span>
<span><span class="co">#&gt;  [43] labeling_0.4.3           abind_1.4-8              compiler_4.5.1          </span></span>
<span><span class="co">#&gt;  [46] aod_1.3.3                withr_3.0.2              doParallel_1.0.17       </span></span>
<span><span class="co">#&gt;  [49] S7_0.2.0                 backports_1.5.0          BiocParallel_1.42.2     </span></span>
<span><span class="co">#&gt;  [52] carData_3.0-5            gplots_3.2.0             MASS_7.3-65             </span></span>
<span><span class="co">#&gt;  [55] bayesm_3.1-6             rjson_0.2.23             corpcor_1.6.10          </span></span>
<span><span class="co">#&gt;  [58] gtools_3.9.5             caTools_1.18.3           tools_4.5.1             </span></span>
<span><span class="co">#&gt;  [61] zip_2.3.3                remaCor_0.0.20           glue_1.8.0              </span></span>
<span><span class="co">#&gt;  [64] nlme_3.1-168             grid_4.5.1               cluster_2.1.8.1         </span></span>
<span><span class="co">#&gt;  [67] reshape2_1.4.4           generics_0.1.4           gtable_0.3.6            </span></span>
<span><span class="co">#&gt;  [70] tidyr_1.3.1              hms_1.1.3                car_3.1-3               </span></span>
<span><span class="co">#&gt;  [73] BiocGenerics_0.54.0      ggrepel_0.9.6            foreach_1.5.2           </span></span>
<span><span class="co">#&gt;  [76] pillar_1.11.0            stringr_1.5.2            limma_3.64.3            </span></span>
<span><span class="co">#&gt;  [79] robustbase_0.99-6        circlize_0.4.16          splines_4.5.1           </span></span>
<span><span class="co">#&gt;  [82] lattice_0.22-5           renv_1.1.5               gmp_0.7-5               </span></span>
<span><span class="co">#&gt;  [85] tidyselect_1.2.1         ComplexHeatmap_2.24.1    pbapply_1.7-4           </span></span>
<span><span class="co">#&gt;  [88] reformulas_0.4.1         IRanges_2.42.0           svglite_2.2.1           </span></span>
<span><span class="co">#&gt;  [91] RhpcBLASctl_0.23-42      stats4_4.5.1             xfun_0.53               </span></span>
<span><span class="co">#&gt;  [94] Biobase_2.68.0           statmod_1.5.0            matrixStats_1.5.0       </span></span>
<span><span class="co">#&gt;  [97] DEoptimR_1.1-4           stringi_1.8.7            yaml_2.3.10             </span></span>
<span><span class="co">#&gt; [100] boot_1.3-32              evaluate_1.0.5           codetools_0.2-19        </span></span>
<span><span class="co">#&gt; [103] tibble_3.3.0             BiocManager_1.30.26      cli_3.6.5               </span></span>
<span><span class="co">#&gt; [106] systemfonts_1.2.3        Rdpack_2.6.4             jquerylib_0.1.4         </span></span>
<span><span class="co">#&gt; [109] Rcpp_1.1.0               EnvStats_3.1.0           png_0.1-8               </span></span>
<span><span class="co">#&gt; [112] parallel_4.5.1           pkgdown_2.1.3            ggplot2_4.0.0           </span></span>
<span><span class="co">#&gt; [115] prettyunits_1.2.0        ClusterR_1.3.4           bitops_1.0-9            </span></span>
<span><span class="co">#&gt; [118] lme4_1.1-37              mvtnorm_1.3-3            lmerTest_3.1-3          </span></span>
<span><span class="co">#&gt; [121] scales_1.4.0             purrr_1.1.0              crayon_1.5.3            </span></span>
<span><span class="co">#&gt; [124] writexl_1.5.4            fANCOVA_0.6-1            GetoptLong_1.0.5        </span></span>
<span><span class="co">#&gt; [127] rlang_1.1.6</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Thomas Rauter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
