<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>methylation-data-analysis • SplineOmics</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="methylation-data-analysis">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SplineOmics</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/get_started.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><h6 class="dropdown-header" data-toc-skip>Further use cases</h6></li>
    <li><a class="dropdown-item" href="../articles/RNA-seq-analysis.html">RNA-seq analysis</a></li>
    <li><a class="dropdown-item" href="../articles/glycan-analysis.html">Glycan analysis</a></li>
    <li><a class="dropdown-item" href="../articles/methylation_data_analysis.html">Epigenetics analysis</a></li>
    <li><a class="dropdown-item" href="../articles/PVC_analysis.html">Peak-Valley-Cliff (PVC) analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Gene_centric_multiomics_clustering.html">Gene-centric multiomics clustering</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Guides</h6></li>
    <li><a class="dropdown-item" href="../articles/limma_result_categories.html">limma result categories</a></li>
    <li><a class="dropdown-item" href="../articles/design_limma_design_formula.html">Designing a limma design formula</a></li>
    <li><a class="dropdown-item" href="../articles/Docker-instructions.html">Docker container instructions</a></li>
    <li><a class="dropdown-item" href="../articles/Docker_permission_denied.html">Docker permission denied</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../articles/faq.html">FAQ</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/CHANGELOG.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/csbg/SplineOmics/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>methylation-data-analysis</h1>
                        <h4 data-toc-skip class="author">Thomas
Rauter</h4>
            
            <h4 data-toc-skip class="date">11 July, 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/csbg/SplineOmics/blob/HEAD/vignettes/methylation_data_analysis.Rmd" class="external-link"><code>vignettes/methylation_data_analysis.Rmd</code></a></small>
      <div class="d-none name"><code>methylation_data_analysis.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="about-this-vignette">About this vignette<a class="anchor" aria-label="anchor" href="#about-this-vignette"></a>
</h2>
<p>This tutorial intends to showcase and explain the capabilities of the
<strong>SplineOmics</strong> package by walking through a real and
complete methylation data analysis example, from start to finish.
<strong>SplineOmics</strong> is explained in more detail in the
<strong>get-started</strong> vignette, where a proteomics example is
covered. This vignette is more focused on showing how methylation data
can be used, and because of this, less details about the overall package
are provided here. Further, this vignette is more minimalistic than the
other ones, because the same functionalities than in the other vignettes
are used and the explanations are not repeated here.</p>
<div class="section level3">
<h3 id="data-overview">Data Overview<a class="anchor" aria-label="anchor" href="#data-overview"></a>
</h3>
<p>This dataset originates from a <strong>time-series methylation data
experiment</strong> designed to study Chinese Hamster Ovary (CHO) cells.
The experiment involved cultivating cells in <strong>eight
bioreactors</strong>, with <strong>four bioreactors</strong> subjected
to a temperature shift after 146 hours (experimental condition) and the
remaining <strong>four bioreactors</strong> maintained without a
temperature shift (control condition).</p>
<div class="section level4">
<h4 id="timepoints">Timepoints<a class="anchor" aria-label="anchor" href="#timepoints"></a>
</h4>
<p>Samples were collected at <strong>19 distinct time points</strong>
throughout the experiment, specifically: <code>"48h"</code>,
<code>"72h"</code>, <code>"96h"</code>, <code>"120h"</code>,
<code>"144h"</code>, <code>"168h"</code>, <code>"192h"</code>,
<code>"216h"</code>, <code>"240h"</code>, <code>"264h"</code>,
<code>"268h"</code>, <code>"288h"</code>, <code>"312h"</code>,
<code>"336h"</code>, <code>"360h"</code>, <code>"384h"</code>,
<code>"408h"</code>, <code>"432h"</code>, <code>"456h"</code>, after
cultivation start. Most time point were sampled from all eight
bioreactors, resulting in a total of <strong>130 samples</strong>.</p>
</div>
<div class="section level4">
<h4 id="additonal-effects-in-the-experiment-reactor-and-group">Additonal effects in the experiment: Reactor and Group<a class="anchor" aria-label="anchor" href="#additonal-effects-in-the-experiment-reactor-and-group"></a>
</h4>
<p>In this experiment, there are two effects to consider:
<strong>Reactor</strong> and <strong>Group</strong>.</p>
<ol style="list-style-type: decimal">
<li>
<strong>Reactor</strong>:
<ul>
<li>This refers to the different bioreactors used for cell cultivation,
which can exhibit substantial variability.</li>
<li>Each reactor was assigned a single condition: either constant
temperature or temperature-shifted. As a result, reactor is a blocked
effect.</li>
<li>Reactor should not be treated as a fixed effect to simply remove its
influence. Instead, it is treated as a <strong>random effect</strong>,
which allows us to model its variability appropriately.</li>
</ul>
</li>
<li>
<strong>Group</strong>:
<ul>
<li>This refers to the two separate groups in which the bioreactors
where run. Because the apparatus containes 4 bioreactors, only that
amount can be run simultaneosly. The other 4 were run after the initial
4.</li>
<li>Group is considered a <strong>batch effect</strong> with respect to
the condition (constant temperature vs. temperature-shifted).</li>
</ul>
</li>
</ol>
<p>Since reactor is a blocked effect, the variability due to the
reactors cannot be directly separated from the condition. Instead,
<strong>linear mixed models (LMMs)</strong> are used to attribute the
reactor as a random effect, allowing us to account for its variability
while isolating the effects of the condition. This approach ensures that
the analysis appropriately handles the hierarchical structure of the
data and avoids incorrect conclusions.</p>
</div>
<div class="section level4">
<h4 id="note">Note<a class="anchor" aria-label="anchor" href="#note"></a>
</h4>
<p>The documentation of all the <strong>SplineOmics</strong> package
functions can be viewed <a href="https://csbg.github.io/SplineOmics/reference">here</a></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="load-the-packages">Load the packages<a class="anchor" aria-label="anchor" href="#load-the-packages"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://csbg.github.io/SplineOmics">SplineOmics</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org" class="external-link">readr</a></span><span class="op">)</span> <span class="co"># For reading the meta CSV file</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://here.r-lib.org/" class="external-link">here</a></span><span class="op">)</span> <span class="co"># For managing filepaths</span></span>
<span><span class="co">#&gt; here() starts at /home/thomas/Documents/PhD/projects/DGTX/SplineOmics_hub/SplineOmics</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span> <span class="co"># For data manipulation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yihui.org/knitr/" class="external-link">knitr</a></span><span class="op">)</span> <span class="co"># For Showing the head of the data and the meta tables.</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-the-files">Load the files<a class="anchor" aria-label="anchor" href="#load-the-files"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">xzfile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>    <span class="st">"extdata"</span>,</span>
<span>    <span class="st">"methylation_data.rds.xz"</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"SplineOmics"</span></span>
<span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>        <span class="st">"extdata"</span>,</span>
<span>        <span class="st">"methylation_meta.csv"</span>,</span>
<span>        package <span class="op">=</span> <span class="st">"SplineOmics"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    show_col_types <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">meta</span><span class="op">$</span><span class="va">Condition</span><span class="op">[</span><span class="va">meta</span><span class="op">$</span><span class="va">Condition</span> <span class="op">==</span> <span class="st">"Temp. Shifted"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Temp_shift"</span></span></code></pre></div>
<div class="section level3">
<h3 id="show-top-rows-of-data">Show top rows of data<a class="anchor" aria-label="anchor" href="#show-top-rows-of-data"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>,</span>
<span>    format <span class="op">=</span> <span class="st">"markdown"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<table class="table"><tbody>
<tr class="odd">
<td align="left">function (…, list = character(), package = NULL,
lib.loc = NULL,</td>
</tr>
<tr class="even">
<td align="left">verbose = getOption(“verbose”), envir = .GlobalEnv,
overwrite = TRUE)</td>
</tr>
<tr class="odd">
<td align="left">{</td>
</tr>
<tr class="even">
<td align="left">fileExt &lt;- function(x) {</td>
</tr>
<tr class="odd">
<td align="left">db &lt;- grepl(“\.[^.]+\.(gz|bz2|xz)$”, x)</td>
</tr>
<tr class="even">
<td align="left">ans &lt;- sub(“.*\.”, ““, x)</td>
</tr>
</tbody></table>
</div>
<div class="section level3">
<h3 id="show-top-rows-of-meta">Show top rows of meta<a class="anchor" aria-label="anchor" href="#show-top-rows-of-meta"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">meta</span><span class="op">)</span>,</span>
<span>    format <span class="op">=</span> <span class="st">"markdown"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<table style="width:100%;" class="table">
<colgroup>
<col width="2%">
<col width="2%">
<col width="3%">
<col width="5%">
<col width="7%">
<col width="6%">
<col width="7%">
<col width="9%">
<col width="9%">
<col width="18%">
<col width="6%">
<col width="18%">
<col width="2%">
</colgroup>
<thead><tr class="header">
<th align="right">No.</th>
<th align="right">TP</th>
<th align="right">Time</th>
<th align="left">Reactor</th>
<th align="left">Condition</th>
<th align="left">Group</th>
<th align="left">Cell Line</th>
<th align="left">Type</th>
<th align="left">Date Isolated</th>
<th align="left">Tube Label</th>
<th align="right">A260/280</th>
<th align="right">Tapestation Concentration</th>
<th align="right">DIN</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">1</td>
<td align="right">5</td>
<td align="right">48</td>
<td align="left">E13</td>
<td align="left">Constant</td>
<td align="left">CharRun1</td>
<td align="left">NISTCHO</td>
<td align="left">1x10^6 cells</td>
<td align="left">2024-11-04</td>
<td align="left">NIST CHO BR24 48H E13</td>
<td align="right">1.888</td>
<td align="right">45.4</td>
<td align="right">9.9</td>
</tr>
<tr class="even">
<td align="right">3</td>
<td align="right">5</td>
<td align="right">48</td>
<td align="left">E15</td>
<td align="left">Constant</td>
<td align="left">CharRun1</td>
<td align="left">NISTCHO</td>
<td align="left">1x10^6 cells</td>
<td align="left">2024-11-04</td>
<td align="left">NIST CHO BR24 48H E15</td>
<td align="right">1.869</td>
<td align="right">50.7</td>
<td align="right">9.9</td>
</tr>
<tr class="odd">
<td align="right">65</td>
<td align="right">5</td>
<td align="right">48</td>
<td align="left">E17</td>
<td align="left">Constant</td>
<td align="left">CharRun2</td>
<td align="left">NISTCHO</td>
<td align="left">1x10^6 cells</td>
<td align="left">2024-11-13</td>
<td align="left">NIST CHO CHARRUN2 48H E17</td>
<td align="right">1.881</td>
<td align="right">68.0</td>
<td align="right">9.8</td>
</tr>
<tr class="even">
<td align="right">67</td>
<td align="right">5</td>
<td align="right">48</td>
<td align="left">E19</td>
<td align="left">Constant</td>
<td align="left">CharRun2</td>
<td align="left">NISTCHO</td>
<td align="left">1x10^6 cells</td>
<td align="left">2024-11-13</td>
<td align="left">NIST CHO CHARRUN2 48H E19</td>
<td align="right">1.885</td>
<td align="right">39.7</td>
<td align="right">9.9</td>
</tr>
<tr class="odd">
<td align="right">2</td>
<td align="right">5</td>
<td align="right">48</td>
<td align="left">E14</td>
<td align="left">Temp_shift</td>
<td align="left">CharRun1</td>
<td align="left">NISTCHO</td>
<td align="left">1x10^6 cells</td>
<td align="left">2024-11-04</td>
<td align="left">NIST CHO BR24 48H E14</td>
<td align="right">1.874</td>
<td align="right">60.2</td>
<td align="right">9.9</td>
</tr>
<tr class="even">
<td align="right">4</td>
<td align="right">5</td>
<td align="right">48</td>
<td align="left">E16</td>
<td align="left">Temp_shift</td>
<td align="left">CharRun1</td>
<td align="left">NISTCHO</td>
<td align="left">1x10^6 cells</td>
<td align="left">2024-11-04</td>
<td align="left">NIST CHO BR24 48H E16</td>
<td align="right">1.892</td>
<td align="right">30.9</td>
<td align="right">9.9</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu">SplineOmics</span><span class="fu">::</span><span class="fu"><a href="../reference/extract_data.html">extract_data</a></span><span class="op">(</span></span>
<span>    <span class="co"># The dataframe with the numbers on the left and info on the right.</span></span>
<span>    data <span class="op">=</span> <span class="va">data_df</span>,</span>
<span>    <span class="co"># Use this annotation column for the feature names.</span></span>
<span>    feature_name_columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"feature_names"</span><span class="op">)</span>,</span>
<span>    top_row <span class="op">=</span> <span class="fl">1</span>,</span>
<span>    bottom_row <span class="op">=</span> <span class="fl">300</span>,</span>
<span>    right_col <span class="op">=</span> <span class="fl">131</span>,</span>
<span>    left_col <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="perform-eda-exploratory-data-analysis">Perform EDA (exploratory data analysis)<a class="anchor" aria-label="anchor" href="#perform-eda-exploratory-data-analysis"></a>
</h2>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">report_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    omics_data_type <span class="op">=</span> <span class="st">"EPI"</span>,</span>
<span>    data_description <span class="op">=</span> <span class="st">"Methylation data of CHO cells"</span>,</span>
<span>    data_collection_date <span class="op">=</span> <span class="st">"December 2024"</span>,</span>
<span>    analyst_name <span class="op">=</span> <span class="st">"Thomas Rauter"</span>,</span>
<span>    contact_info <span class="op">=</span> <span class="st">"thomas.rauter@plus.ac.at"</span>,</span>
<span>    project_name <span class="op">=</span> <span class="st">"DGTX"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">report_dir</span> <span class="op">&lt;-</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html" class="external-link">here</a></span><span class="op">(</span></span>
<span>    <span class="st">"results"</span>,</span>
<span>    <span class="st">"explore_data"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">splineomics</span> <span class="op">&lt;-</span> <span class="fu">SplineOmics</span><span class="fu">::</span><span class="fu"><a href="../reference/create_splineomics.html">create_splineomics</a></span><span class="op">(</span></span>
<span>    data <span class="op">=</span> <span class="va">data</span>,</span>
<span>    meta <span class="op">=</span> <span class="va">meta</span>,</span>
<span>    report_info <span class="op">=</span> <span class="va">report_info</span>,</span>
<span>    condition <span class="op">=</span> <span class="st">"Condition"</span>, <span class="co"># Column of meta that contains the levels.</span></span>
<span>    meta_batch_column <span class="op">=</span> <span class="st">"Group"</span> <span class="co"># Remove batch effect for plotting.</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu">SplineOmics</span><span class="fu">::</span><span class="fu"><a href="../reference/explore_data.html">explore_data</a></span><span class="op">(</span></span>
<span>    splineomics <span class="op">=</span> <span class="va">splineomics</span>,</span>
<span>    report_dir <span class="op">=</span> <span class="fu">withr</span><span class="fu">::</span><span class="fu"><a href="https://withr.r-lib.org/reference/with_tempfile.html" class="external-link">local_tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Making density plots...</span></span>
<span><span class="co">#&gt; Making violin plots...</span></span>
<span><span class="co">#&gt; Making PCA plots...</span></span>
<span><span class="co">#&gt; 25 rows with missing values removed before PCA.</span></span>
<span><span class="co">#&gt; Making MDS plots...</span></span>
<span><span class="co">#&gt; 25 rows with missing values removed before MDS</span></span>
<span><span class="co">#&gt; Making correlation heatmaps...</span></span>
<span><span class="co">#&gt; Making mean correlation with time plots...</span></span>
<span><span class="co">#&gt; Making lag1 differences plots...</span></span>
<span><span class="co">#&gt; Making first lag auto-correlation with time plots...</span></span>
<span><span class="co">#&gt; 25 rows with missing values removed before firstlag autocorrelation plot</span></span>
<span><span class="co">#&gt; Making cv plots...</span></span>
<span><span class="co">#&gt; Warning: Removed 367 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_density()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 177 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_density()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 190 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_density()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 177 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_ydensity()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 177 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_boxplot()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 190 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_ydensity()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 190 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_boxplot()`).</span></span>
<span><span class="co">#&gt; Making density plots...</span></span>
<span><span class="co">#&gt; Making violin plots...</span></span>
<span><span class="co">#&gt; Making PCA plots...</span></span>
<span><span class="co">#&gt; 25 rows with missing values removed before PCA.</span></span>
<span><span class="co">#&gt; Making MDS plots...</span></span>
<span><span class="co">#&gt; 25 rows with missing values removed before MDS</span></span>
<span><span class="co">#&gt; Making correlation heatmaps...</span></span>
<span><span class="co">#&gt; Making mean correlation with time plots...</span></span>
<span><span class="co">#&gt; Making lag1 differences plots...</span></span>
<span><span class="co">#&gt; Making first lag auto-correlation with time plots...</span></span>
<span><span class="co">#&gt; 25 rows with missing values removed before firstlag autocorrelation plot</span></span>
<span><span class="co">#&gt; Making cv plots...</span></span>
<span><span class="co">#&gt; Warning: Removed 367 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_density()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 177 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_density()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 190 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_density()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 177 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_ydensity()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 177 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_boxplot()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 190 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_ydensity()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 190 rows containing non-finite outside the scale range</span></span>
<span><span class="co">#&gt; (`stat_boxplot()`).</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  <span style="color: #00BB00;">Info</span> Exploratory data analysis completed successfully.</span></span>
<span><span class="co">#&gt;  Your HTML reports are located in the directory:  /tmp/Rtmpar1fXj/file7177550b7323 .</span></span>
<span><span class="co">#&gt;  Please note that due to embedded files, the reports might be flagged as</span></span>
<span><span class="co">#&gt;  harmful by other software. Rest assured that they provide no harm.</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="run-limma-spline-analysis">Run limma spline analysis<a class="anchor" aria-label="anchor" href="#run-limma-spline-analysis"></a>
</h2>
<p>In this example, we are skipping finding the best hyperparameters
with the screen_limma_hyperparams() function, because we already have a
clear idea of what do use.</p>
<p>Lets define our parameters and put them into the
<code>SplineOmics</code> object:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spline_params</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    spline_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"n"</span><span class="op">)</span>, <span class="co"># natural cubic splines</span></span>
<span>    dof <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3L</span><span class="op">)</span> <span class="co"># Degree of freedom of 3 for the splines.</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="st">"~ 1 + Condition*Time + Group + (1|Reactor)"</span></span>
<span></span>
<span><span class="va">splineomics</span> <span class="op">&lt;-</span> <span class="fu">SplineOmics</span><span class="fu">::</span><span class="fu"><a href="../reference/update_splineomics.html">update_splineomics</a></span><span class="op">(</span></span>
<span>    splineomics <span class="op">=</span> <span class="va">splineomics</span>,</span>
<span>    data <span class="op">=</span> <span class="va">data</span>,</span>
<span>    design <span class="op">=</span> <span class="va">design</span>,</span>
<span>    mode <span class="op">=</span> <span class="st">"integrated"</span>,</span>
<span>    spline_params <span class="op">=</span> <span class="va">spline_params</span>,</span>
<span>    bp_cfg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="co"># For parallel computing for variancePartition::dream (lmm)</span></span>
<span>        n_cores <span class="op">=</span> <span class="fl">1</span>, <span class="co"># For the vignette, just use one core.</span></span>
<span>        blas_threads <span class="op">=</span> <span class="fl">1</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Run the <code><a href="../reference/run_limma_splines.html">run_limma_splines()</a></code> function:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">splineomics</span> <span class="op">&lt;-</span> <span class="fu">SplineOmics</span><span class="fu">::</span><span class="fu"><a href="../reference/run_limma_splines.html">run_limma_splines</a></span><span class="op">(</span></span>
<span>    splineomics <span class="op">=</span> <span class="va">splineomics</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Hint: The data contains missing values (NA). Ensure that imputation or handling of missing values aligns with your analysis requirements. Note that limma can handle missing values (it just ignores them), and therefore SplineOmics can also handle them.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Ensure the design has no Condition*Time interaction for mode == 'isolated', and includes it for mode == 'integrated'.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Fitting global model...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; NOTE: If you manually stop run_limma_splines() in RStudio and  used parallelization for variancePartition::dream(), then those parallelized processes may continue running. Use your system's process manager to terminate them manually!</span></span>
<span><span class="co">#&gt; Warning in variancePartition::dream(exprObj = data, formula = stats::as.formula(design), : Model failed for 8 responses.</span></span>
<span><span class="co">#&gt;   See errors with attr(., 'errors')</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">Info</span> Finished limma spline analysis in 0.2 min</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="build-limma-report">Build limma report<a class="anchor" aria-label="anchor" href="#build-limma-report"></a>
</h2>
<p>The topTables of all three limma result categories can be used to
generate p-value histograms an volcano plots.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu">SplineOmics</span><span class="fu">::</span><span class="fu"><a href="../reference/create_limma_report.html">create_limma_report</a></span><span class="op">(</span></span>
<span>    splineomics <span class="op">=</span> <span class="va">splineomics</span>,</span>
<span>    report_dir <span class="op">=</span> <span class="fu">withr</span><span class="fu">::</span><span class="fu"><a href="https://withr.r-lib.org/reference/with_tempfile.html" class="external-link">local_tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  <span style="color: #00BB00;">Info</span> Limma report generation completed successfully.</span></span>
<span><span class="co">#&gt;  Your HTML reports are located in the directory:  /tmp/Rtmpar1fXj/file717712c6e521 .</span></span>
<span><span class="co">#&gt;  Please note that due to embedded files, the reports might be flagged as</span></span>
<span><span class="co">#&gt;  harmful by other software. Rest assured that they provide no harm.</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="cluster-the-hits-significant-features">Cluster the hits (significant features)<a class="anchor" aria-label="anchor" href="#cluster-the-hits-significant-features"></a>
</h2>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nr_clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    Constant   <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">6</span>,</span>
<span>    Temp_shift <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">6</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">report_dir</span> <span class="op">&lt;-</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html" class="external-link">here</a></span><span class="op">(</span></span>
<span>    <span class="st">"results"</span>,</span>
<span>    <span class="st">"clustering_reports"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span> <span class="co"># For the spline plots</span></span>
<span>    y_axis_label <span class="op">=</span> <span class="st">"beta value"</span>,</span>
<span>    time_unit <span class="op">=</span> <span class="st">"hours"</span>,</span>
<span>    treatment_labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        Temp_shift <span class="op">=</span> <span class="st">"temp shift"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    treatment_timepoints <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        Temp_shift <span class="op">=</span> <span class="fl">146</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot_options</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="co"># When meta_replicate_column is not there, all datapoints are blue.</span></span>
<span>    meta_replicate_column <span class="op">=</span> <span class="st">"Reactor"</span> <span class="co"># Colors the data points based on Reactor</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">clustering_results</span> <span class="op">&lt;-</span> <span class="fu">SplineOmics</span><span class="fu">::</span><span class="fu"><a href="../reference/cluster_hits.html">cluster_hits</a></span><span class="op">(</span></span>
<span>    splineomics <span class="op">=</span> <span class="va">splineomics</span>,</span>
<span>    adj_pthresh_time_effect <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>    adj_pthresh_avrg_diff_conditions <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>    adj_pthresh_interaction_condition_time <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>    nr_clusters <span class="op">=</span> <span class="va">nr_clusters</span>,</span>
<span>    genes <span class="op">=</span> <span class="va">genes</span>,</span>
<span>    plot_info <span class="op">=</span> <span class="va">plot_info</span>,</span>
<span>    plot_options <span class="op">=</span> <span class="va">plot_options</span>,</span>
<span>    report_dir <span class="op">=</span> <span class="fu">withr</span><span class="fu">::</span><span class="fu"><a href="https://withr.r-lib.org/reference/with_tempfile.html" class="external-link">local_tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Hint: The data contains missing values (NA). Ensure that imputation or handling of missing values aligns with your analysis requirements. Note that limma can handle missing values (it just ignores them), and therefore SplineOmics can also handle them.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Ensure the design has no Condition*Time interaction for mode == 'isolated', and includes it for mode == 'integrated'.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Performing the clustering...</span></span>
<span><span class="co">#&gt; For the level:  Constant</span></span>
<span><span class="co">#&gt; For the level:  Temp_shift</span></span>
<span><span class="co">#&gt; Generating heatmap...</span></span>
<span><span class="co">#&gt; Generating cluster mean splines for level:  Constant</span></span>
<span><span class="co">#&gt; Generating spline plots...</span></span>
<span><span class="co">#&gt; Generating cluster mean splines for level:  Temp_shift</span></span>
<span><span class="co">#&gt; Generating spline plots...</span></span>
<span><span class="co">#&gt; Generating report. This takes a few seconds.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  <span style="color: #00BB00;">Info</span> Clustering the hits completed successfully.</span></span>
<span><span class="co">#&gt;  Your HTML reports are located in the directory:  /tmp/Rtmpar1fXj/file71775381b813 .</span></span>
<span><span class="co">#&gt;  Please note that due to embedded files, the reports might be flagged as</span></span>
<span><span class="co">#&gt;  harmful by other software. Rest assured that they provide no harm.</span></span>
<span><span class="co">#&gt; Running this function took 0.3 min</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.5.2 (2025-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=de_AT.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=de_AT.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=de_AT.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=de_AT.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Europe/Vienna</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices datasets  utils     methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] knitr_1.50        dplyr_1.1.4       here_1.0.1        readr_2.1.5      </span></span>
<span><span class="co">#&gt; [5] SplineOmics_0.4.2</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RColorBrewer_1.1-3       rstudioapi_0.17.1        jsonlite_2.0.0          </span></span>
<span><span class="co">#&gt;   [4] shape_1.4.6.1            magrittr_2.0.3           farver_2.1.2            </span></span>
<span><span class="co">#&gt;   [7] nloptr_2.2.1             rmarkdown_2.29           GlobalOptions_0.1.2     </span></span>
<span><span class="co">#&gt;  [10] fs_1.6.6                 ragg_1.4.0               vctrs_0.6.5             </span></span>
<span><span class="co">#&gt;  [13] minqa_1.2.8              base64enc_0.1-3          htmltools_0.5.8.1       </span></span>
<span><span class="co">#&gt;  [16] progress_1.2.3           broom_1.0.8              Formula_1.2-5           </span></span>
<span><span class="co">#&gt;  [19] variancePartition_1.38.0 sass_0.4.10              KernSmooth_2.23-26      </span></span>
<span><span class="co">#&gt;  [22] bslib_0.9.0              htmlwidgets_1.6.4        desc_1.4.3              </span></span>
<span><span class="co">#&gt;  [25] pbkrtest_0.5.4           plyr_1.8.9               cachem_1.1.0            </span></span>
<span><span class="co">#&gt;  [28] lifecycle_1.0.4          iterators_1.0.14         pkgconfig_2.0.3         </span></span>
<span><span class="co">#&gt;  [31] Matrix_1.7-3             R6_2.6.1                 fastmap_1.2.0           </span></span>
<span><span class="co">#&gt;  [34] rbibutils_2.3            clue_0.3-66              digest_0.6.37           </span></span>
<span><span class="co">#&gt;  [37] numDeriv_2016.8-1.1      colorspace_2.1-1         S4Vectors_0.46.0        </span></span>
<span><span class="co">#&gt;  [40] rprojroot_2.1.1          textshaping_1.0.1        labeling_0.4.3          </span></span>
<span><span class="co">#&gt;  [43] abind_1.4-8              compiler_4.5.2           bit64_4.6.0-1           </span></span>
<span><span class="co">#&gt;  [46] aod_1.3.3                withr_3.0.2              doParallel_1.0.17       </span></span>
<span><span class="co">#&gt;  [49] backports_1.5.0          BiocParallel_1.42.1      carData_3.0-5           </span></span>
<span><span class="co">#&gt;  [52] gplots_3.2.0             MASS_7.3-65              rjson_0.2.23            </span></span>
<span><span class="co">#&gt;  [55] corpcor_1.6.10           gtools_3.9.5             caTools_1.18.3          </span></span>
<span><span class="co">#&gt;  [58] tools_4.5.2              zip_2.3.3                remaCor_0.0.18          </span></span>
<span><span class="co">#&gt;  [61] glue_1.8.0               nlme_3.1-168             grid_4.5.2              </span></span>
<span><span class="co">#&gt;  [64] checkmate_2.3.3          cluster_2.1.8.1          reshape2_1.4.4          </span></span>
<span><span class="co">#&gt;  [67] generics_0.1.4           gtable_0.3.6             tzdb_0.5.0              </span></span>
<span><span class="co">#&gt;  [70] tidyr_1.3.1              hms_1.1.3                car_3.1-3               </span></span>
<span><span class="co">#&gt;  [73] BiocGenerics_0.54.0      ggrepel_0.9.6            foreach_1.5.2           </span></span>
<span><span class="co">#&gt;  [76] pillar_1.11.0            stringr_1.5.1            vroom_1.6.5             </span></span>
<span><span class="co">#&gt;  [79] limma_3.64.1             circlize_0.4.16          splines_4.5.2           </span></span>
<span><span class="co">#&gt;  [82] lattice_0.22-5           renv_1.1.5               gmp_0.7-5               </span></span>
<span><span class="co">#&gt;  [85] bit_4.6.0                tidyselect_1.2.1         ComplexHeatmap_2.24.1   </span></span>
<span><span class="co">#&gt;  [88] pbapply_1.7-2            reformulas_0.4.1         IRanges_2.42.0          </span></span>
<span><span class="co">#&gt;  [91] svglite_2.2.1            RhpcBLASctl_0.23-42      stats4_4.5.2            </span></span>
<span><span class="co">#&gt;  [94] xfun_0.52                Biobase_2.68.0           statmod_1.5.0           </span></span>
<span><span class="co">#&gt;  [97] matrixStats_1.5.0        stringi_1.8.7            yaml_2.3.10             </span></span>
<span><span class="co">#&gt; [100] boot_1.3-31              evaluate_1.0.4           codetools_0.2-19        </span></span>
<span><span class="co">#&gt; [103] tibble_3.3.0             BiocManager_1.30.26      cli_3.6.5               </span></span>
<span><span class="co">#&gt; [106] systemfonts_1.2.3        Rdpack_2.6.4             jquerylib_0.1.4         </span></span>
<span><span class="co">#&gt; [109] Rcpp_1.1.0               EnvStats_3.1.0           png_0.1-8               </span></span>
<span><span class="co">#&gt; [112] parallel_4.5.2           pkgdown_2.1.3            ggplot2_3.5.2           </span></span>
<span><span class="co">#&gt; [115] prettyunits_1.2.0        ClusterR_1.3.3           bitops_1.0-9            </span></span>
<span><span class="co">#&gt; [118] lme4_1.1-37              mvtnorm_1.3-3            lmerTest_3.1-3          </span></span>
<span><span class="co">#&gt; [121] scales_1.4.0             purrr_1.1.0              crayon_1.5.3            </span></span>
<span><span class="co">#&gt; [124] writexl_1.5.4            fANCOVA_0.6-1            GetoptLong_1.0.5        </span></span>
<span><span class="co">#&gt; [127] rlang_1.1.6</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Thomas Rauter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
