<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>RNA_seq_analysis • SplineOmics</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="RNA_seq_analysis">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SplineOmics</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/design_limma_design_formula.html">Designing a Limma Design Formula</a></li>
    <li><a class="dropdown-item" href="../articles/Docker_permission_denied.html">Docker_permission_denied</a></li>
    <li><a class="dropdown-item" href="../articles/Docker-instructions.html">Docker-instructions</a></li>
    <li><a class="dropdown-item" href="../articles/get-started.html">get-started</a></li>
    <li><a class="dropdown-item" href="../articles/RNA_seq_analysis.html">RNA_seq_analysis</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/csbg/SplineOmics/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>RNA_seq_analysis</h1>
                        <h4 data-toc-skip class="author">Thomas
Rauter</h4>
            
            <h4 data-toc-skip class="date">10 January, 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/csbg/SplineOmics/blob/HEAD/vignettes/RNA_seq_analysis.Rmd" class="external-link"><code>vignettes/RNA_seq_analysis.Rmd</code></a></small>
      <div class="d-none name"><code>RNA_seq_analysis.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="about-this-vignette">About this vignette<a class="anchor" aria-label="anchor" href="#about-this-vignette"></a>
</h2>
<p>This tutorial intends to showcase and explain the capabilities of the
<strong>SplineOmics</strong> package by walking through a real and
complete RNA-seq example, from start to finish.
<strong>SplineOmics</strong> is explained in more detail in the
<strong>get-started</strong> vignette, where a proteomics example is
covered. This vignette is more focused on showing how RNA-seq data can
be used, and because of this, less details about the overall package are
provided here.</p>
<div class="section level3">
<h3 id="data-overview">Data Overview<a class="anchor" aria-label="anchor" href="#data-overview"></a>
</h3>
<p>This dataset originates from a <strong>time-series RNA-seq
experiment</strong> designed to study Chinese Hamster Ovary (CHO) cells.
The experiment involved cultivating cells in <strong>eight
bioreactors</strong>, with <strong>four bioreactors</strong> subjected
to a temperature shift after 146 hours (experimental condition) and the
remaining <strong>four bioreactors</strong> maintained without a
temperature shift (control condition).</p>
<p>Samples were collected at <strong>17 distinct time points</strong>
throughout the experiment, specifically: <code>"72h"</code>,
<code>"76h"</code>, <code>"96h"</code>, <code>"120h"</code>,
<code>"124h"</code>, <code>"144h"</code>, <code>"148h"</code>,
<code>"152h"</code>, <code>"168h"</code>, <code>"192h"</code>,
<code>"216h"</code>, <code>"220h"</code>, <code>"240h"</code>,
<code>"264h"</code>, <code>"268h"</code>, <code>"288h"</code>, and
<code>"312h"</code> after cultivation start. Each time point was sampled
from all eight bioreactors, resulting in a total of <strong>136
samples</strong>.</p>
<p>The data matrix comprises <strong>genes as rows</strong> and
<strong>samples as columns</strong>, providing gene expression
measurements for all time points. Each sample was initially sequenced in
<strong>three technical replicates</strong> across <strong>two NovaSeq X
flow cells</strong>. These technical replicates were collapsed to
generate the final dataset used in this analysis.</p>
<p>The goal of this experiment is to investigate the effect of a
temperature shift during CHO cell cultivation on gene expression
dynamics over time.</p>
<p>Note: This is not the original dataset, as it has not yet been
published at the time of this vignette’s creation. For demonstration
purposes, the genes have been randomly shuffled, and only a subset of
the data has been included to reduce the dataset size.</p>
</div>
<div class="section level3">
<h3 id="analysis-goals">Analysis Goals<a class="anchor" aria-label="anchor" href="#analysis-goals"></a>
</h3>
<p>The main objectives of this analysis are:</p>
<ul>
<li><p><strong>Identify genes with significant temporal
changes</strong>: Among the thousands of genes measured, the goal is to
identify those that exhibit significant changes in expression over
time.</p></li>
<li><p><strong>Cluster genes based on temporal patterns</strong>: Genes
showing significant temporal changes (hits) will be grouped into
clusters based on their time-dependent expression patterns.</p></li>
<li><p><strong>Perform gene set enrichment analysis</strong>: For each
cluster, a gene set enrichment analysis will be conducted to identify
whether specific biological pathways or processes are up- or
downregulated in response to feeding and how these processes are
influenced by the temperature shift.</p></li>
<li><p><strong>Assess the impact of temperature shifts on temporal
patterns</strong>: The analysis will determine whether the temporal
patterns of gene expression are affected by the temperature shift, i.e.,
whether gene expression dynamics differ over time under temperature
shift conditions compared to controls.</p></li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="load-the-packages">Load the packages<a class="anchor" aria-label="anchor" href="#load-the-packages"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://csbg.github.io/SplineOmics">SplineOmics</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: replacing previous import 'limma::topTable' by</span></span>
<span><span class="co">#&gt; 'variancePartition::topTable' when loading 'SplineOmics'</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org" class="external-link">readr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://here.r-lib.org/" class="external-link">here</a></span><span class="op">)</span>  <span class="co"># For managing filepaths</span></span>
<span><span class="co">#&gt; here() starts at /home/thomas/Documents/PhD/projects/DGTX/R_packages/SplineOmics</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span> <span class="co"># For data manipulation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-the-files">Load the files<a class="anchor" aria-label="anchor" href="#load-the-files"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">xzfile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>  <span class="st">"extdata"</span>,</span>
<span>  <span class="st">"rna_seq_data.rds.xz"</span>,</span>
<span>  package <span class="op">=</span> <span class="st">"SplineOmics"</span></span>
<span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>    <span class="st">"extdata"</span>,</span>
<span>    <span class="st">"rna_seq_meta.csv"</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"SplineOmics"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">136</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">5</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="font-weight: bold;">Delimiter:</span> ","</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">chr</span> (3): Reactor, Condition, Plate</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">dbl</span> (2): SampleNr, Time</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span></code></pre></div>
<p><strong>data</strong>: A numeric matrix where each row represents a
gene (features) and each column corresponds to a sample. The row names
of the matrix contain the gene identifiers, while the columns are
aligned with the sample metadata in meta. The matrix contains expression
values for 136 samples.</p>
<p><strong>meta</strong>: A data frame containing metadata information
about the samples in data. Each row in meta corresponds to a column in
data, ensuring a 1:1 alignment between metadata entries and expression
data samples. The columns in meta describe various attributes of the
samples, such as SampleNr, Reactor, Time, Condition, and Plate.</p>
</div>
<div class="section level2">
<h2 id="preprocess-the-data">Preprocess the data<a class="anchor" aria-label="anchor" href="#preprocess-the-data"></a>
</h2>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="perform-eda-exploratory-data-analysis">Perform EDA (exploratory data analysis)<a class="anchor" aria-label="anchor" href="#perform-eda-exploratory-data-analysis"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">report_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  omics_data_type <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>  data_description <span class="op">=</span> <span class="st">"RNA-seq data of CHO cells"</span>,</span>
<span>  data_collection_date <span class="op">=</span> <span class="st">"December 2024"</span>,</span>
<span>  analyst_name <span class="op">=</span> <span class="st">"Thomas Rauter"</span>,</span>
<span>  contact_info <span class="op">=</span> <span class="st">"thomas.rauter@plus.ac.at"</span>,</span>
<span>  project_name <span class="op">=</span> <span class="st">"DGTX"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">report_dir</span> <span class="op">&lt;-</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span></span>
<span>  <span class="st">"results"</span>,</span>
<span>  <span class="st">"explore_data"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">splineomics</span> <span class="op">&lt;-</span> <span class="fu">SplineOmics</span><span class="fu">::</span><span class="fu"><a href="../reference/create_splineomics.html">create_splineomics</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">data</span>,</span>
<span>  meta <span class="op">=</span> <span class="va">meta</span>,</span>
<span>  <span class="co"># annotation = annotation,</span></span>
<span>  report_info <span class="op">=</span> <span class="va">report_info</span>,</span>
<span>  condition <span class="op">=</span> <span class="st">"Condition"</span>, <span class="co"># Column of meta that contains the levels.</span></span>
<span>  meta_batch_column <span class="op">=</span> <span class="st">"Plate"</span> <span class="co"># For batch effect removal</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu">SplineOmics</span><span class="fu">::</span><span class="fu"><a href="../reference/explore_data.html">explore_data</a></span><span class="op">(</span></span>
<span>  splineomics <span class="op">=</span> <span class="va">splineomics</span>, </span>
<span>  report_dir <span class="op">=</span> <span class="va">report_dir</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><a href="https://drive.google.com/uc?export=view&amp;id=1mE7FFXFO9ZZ907in3M5Cf1w7yRxRFHiy" class="external-link">Here</a>
you can see the HTML report of the explore_data() function with the NOT
batch-corrected data, and <a href="https://drive.google.com/uc?export=view&amp;id=1g_Lt4kO4H79GnVh_f7KfXqIzxfTnMrmq" class="external-link">here</a>
the report for the batch-corrected data.</p>
</div>
<div class="section level2">
<h2 id="run-limma-spline-analysis">Run limma spline analysis<a class="anchor" aria-label="anchor" href="#run-limma-spline-analysis"></a>
</h2>
<p>In this example, we are skipping finding the best hyperparameters
with the screen_limma_hyperparams() function, because we already have a
clear idea of what do use.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">voom_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocess_rna_seq_data.html">preprocess_rna_seq_data</a></span><span class="op">(</span></span>
<span>  raw_counts <span class="op">=</span> <span class="va">data</span>,</span>
<span>  meta <span class="op">=</span> <span class="va">meta</span>,</span>
<span>  spline_params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>spline_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"n"</span><span class="op">)</span>,   <span class="co"># Chosen spline parameters</span></span>
<span>                       dof <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2L</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  design <span class="op">=</span> <span class="st">"~ 1 + Condition*X + Plate"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Preprocessing RNA-seq data (normalization + voom)...</span></span>
<span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">voom_obj</span><span class="op">$</span><span class="va">E</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">splineomics</span> <span class="op">&lt;-</span> <span class="fu">SplineOmics</span><span class="fu">::</span><span class="fu"><a href="../reference/update_splineomics.html">update_splineomics</a></span><span class="op">(</span></span>
<span>  splineomics <span class="op">=</span> <span class="va">splineomics</span>,</span>
<span>  data <span class="op">=</span> <span class="va">data</span>,</span>
<span>  rna_seq_data <span class="op">=</span> <span class="va">voom_obj</span>,</span>
<span>  design <span class="op">=</span> <span class="st">"~ 1 + Condition*X + Plate"</span>, </span>
<span>  mode <span class="op">=</span> <span class="st">"integrated"</span>, <span class="co"># means limma uses the full data for each condition.</span></span>
<span>  spline_params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    spline_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"n"</span><span class="op">)</span>,    <span class="co"># natural cubic splines</span></span>
<span>    dof <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2L</span><span class="op">)</span>              <span class="co"># Degree of freedom of 2 for the splines.</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Run the <code><a href="../reference/run_limma_splines.html">run_limma_splines()</a></code> function with the updated
SplineOmics object:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">splineomics</span> <span class="op">&lt;-</span> <span class="fu">SplineOmics</span><span class="fu">::</span><span class="fu"><a href="../reference/run_limma_splines.html">run_limma_splines</a></span><span class="op">(</span></span>
<span>  splineomics <span class="op">=</span> <span class="va">splineomics</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Hint: The data contains negative values. This may occur if the data has been transformed (e.g., log-transformed or normalized) and is valid in such cases. Ensure that the data preprocessing  aligns with your analysis requirements.</span></span>
<span><span class="co">#&gt; Column 'Plate' of meta will be used to remove the batch effect for the plotting</span></span>
<span><span class="co">#&gt; Make sure that the design formula contains no interaction between the condition and time (X) for mode == isolated, and that it contains an interaction for mode == integrated. Otherwise, you will get an uncaught error of 'coefficients not estimable' or 'subscript out of bounds'.</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">Info</span> limma spline analysis completed successfully</span></span></code></pre></div>
<p>The output of the function run_limma_splines() is a named list, where
each element is a specific “category” of results. Refer to <a href="https://csbg.github.io/SplineOmics/articles/limma_result_categories.html">this
document</a> for an explanation of the different result categories. Each
of those elements is a list, containing as elements the respective limma
topTables, either for each level or each comparison between two
levels.</p>
<p>The element “time_effect” is a list, where each element is the
topTable where the p-value for each feature for the respective level are
reported.</p>
<p>The element “avrg_diff_conditions” is a list that contains as
elements the topTables, that represent the comparison of the average
differences of the levels.</p>
<p>The element “interaction_condition_time” is a list that contains as
elements the topTables, that represent the interaction between the
levels (which includes both time and the average differences)</p>
</div>
<div class="section level2">
<h2 id="build-limma-report">Build limma report<a class="anchor" aria-label="anchor" href="#build-limma-report"></a>
</h2>
<p>The topTables of all three limma result categories can be used to
generate p-value histograms an volcano plots.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">report_dir</span> <span class="op">&lt;-</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span></span>
<span>  <span class="st">"results"</span>,</span>
<span>  <span class="st">"create_limma_reports"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu">SplineOmics</span><span class="fu">::</span><span class="fu"><a href="../reference/create_limma_report.html">create_limma_report</a></span><span class="op">(</span></span>
<span>  splineomics <span class="op">=</span> <span class="va">splineomics</span>,</span>
<span>  report_dir <span class="op">=</span> <span class="va">report_dir</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>You can view the generated analysis report of the create_limma_report
function <a href="https://drive.google.com/uc?export=view&amp;id=1hvKloiaHQA5CO4Y6io8qC2lOpBJ54ypJ" class="external-link">here</a>.</p>
</div>
<div class="section level2">
<h2 id="cluster-the-hits-significant-features">Cluster the hits (significant features)<a class="anchor" aria-label="anchor" href="#cluster-the-hits-significant-features"></a>
</h2>
<p>After we obtained the limma spline results, we can cluster the hits
based on their temporal pattern (their spline shape). We define what a
hit is by setting an adj. p-value threshold for every level. Hits are
features (e.g. proteins) that have an adj. p-value below the threshold.
Hierarchical clustering is used to place every hit in one of as many
clusters as we have specified for that specific level.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">adj_pthresholds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="co"># 0.05 for both levels</span></span>
<span>  <span class="fl">0.00001</span>, <span class="co"># constant (temperature)  </span></span>
<span>  <span class="fl">0.00001</span>  <span class="co"># temp_shift</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="fl">4L</span>,  <span class="co"># 4 clusters for constant</span></span>
<span>  <span class="fl">4L</span>   <span class="co"># 4 clusters for temp_shift</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">report_dir</span> <span class="op">&lt;-</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span></span>
<span>  <span class="st">"results"</span>,</span>
<span>  <span class="st">"clustering_reports"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span> <span class="co"># For the spline plots</span></span>
<span>  y_axis_label <span class="op">=</span> <span class="st">"log2 intensity"</span>,</span>
<span>  time_unit <span class="op">=</span> <span class="st">"min"</span>, <span class="co"># our measurements were in minutes</span></span>
<span>  treatment_labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"temp_shift"</span><span class="op">)</span>, <span class="co"># add this for all conditions</span></span>
<span>  treatment_timepoints <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">146</span><span class="op">)</span> <span class="co"># Feeding was at 0 minutes.</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot_options</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="co"># When meta_replicate_column is not there, all datapoints are blue.</span></span>
<span>  meta_replicate_column <span class="op">=</span> <span class="st">"Reactor"</span>, <span class="co"># Colors the data points based on Reactor</span></span>
<span>  cluster_heatmap_columns <span class="op">=</span> <span class="cn">FALSE</span> <span class="co"># Per default FALSE, just for demonstration</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">clustering_results</span> <span class="op">&lt;-</span> <span class="fu">SplineOmics</span><span class="fu">::</span><span class="fu"><a href="../reference/cluster_hits.html">cluster_hits</a></span><span class="op">(</span></span>
<span>  splineomics <span class="op">=</span> <span class="va">splineomics</span>,</span>
<span>  adj_pthresholds <span class="op">=</span> <span class="va">adj_pthresholds</span>,</span>
<span>  clusters <span class="op">=</span> <span class="va">clusters</span>,</span>
<span>  genes <span class="op">=</span> <span class="va">genes</span>,</span>
<span>  plot_info <span class="op">=</span> <span class="va">plot_info</span>,</span>
<span>  plot_options <span class="op">=</span> <span class="va">plot_options</span>,</span>
<span>  report_dir <span class="op">=</span> <span class="va">report_dir</span>,</span>
<span>  adj_pthresh_avrg_diff_conditions <span class="op">=</span> <span class="fl">0.00001</span>,</span>
<span>  adj_pthresh_interaction_condition_time <span class="op">=</span> <span class="fl">0.00001</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>You can view the generated analysis report of the cluster_hits
function <a href="https://drive.google.com/uc?export=view&amp;id=1EvPiEwHyGnBQFxRets00s87tochgyGK4" class="external-link">here</a>.</p>
<p>As discussed before, there are three limma result categories. The
cluster_hits() report shows the results of all three, if they are
present (category 2 and 3 can only be generated when the design formula
contains an interaction effect).</p>
</div>
<div class="section level2">
<h2 id="perform-gene-set-enrichment-analysis-gsea">Perform gene set enrichment analysis (GSEA)<a class="anchor" aria-label="anchor" href="#perform-gene-set-enrichment-analysis-gsea"></a>
</h2>
<p>Usually, the final step in such a bioinformatics analysis is GSEA. To
each clustered hit, the respective gene can be assigned and GSEA
performed. For this, the Enrichr databases of choice have to be
downloaded:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Specify which databases you want to download from Enrichr</span></span>
<span><span class="va">gene_set_lib</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"WikiPathways_2019_Human"</span>,</span>
<span>  <span class="st">"NCI-Nature_2016"</span>,</span>
<span>  <span class="st">"TRRUST_Transcription_Factors_2019"</span>,</span>
<span>  <span class="st">"MSigDB_Hallmark_2020"</span>,</span>
<span>  <span class="st">"GO_Cellular_Component_2018"</span>,</span>
<span>  <span class="st">"CORUM"</span>,</span>
<span>  <span class="st">"KEGG_2019_Human"</span>,</span>
<span>  <span class="st">"TRANSFAC_and_JASPAR_PWMs"</span>,</span>
<span>  <span class="st">"ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X"</span>,</span>
<span>  <span class="st">"GO_Biological_Process_2018"</span>,</span>
<span>  <span class="st">"GO_Molecular_Function_2018"</span>,</span>
<span>  <span class="st">"Human_Gene_Atlas"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">SplineOmics</span><span class="fu">::</span><span class="fu"><a href="../reference/download_enrichr_databases.html">download_enrichr_databases</a></span><span class="op">(</span></span>
<span>  gene_set_lib <span class="op">=</span> <span class="va">gene_set_lib</span>,</span>
<span>  filename <span class="op">=</span> <span class="st">"databases.tsv"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Per default the file is placed in the current working directory,
which is the root dir of the R project.</p>
<p>To run GSEA, the downloaded database file has to be loaded as a
dataframe. Further, optionally, the clusterProfiler parameters and the
report dir can be specified. The function create_gsea_report() runs GSEA
using clusterProfiler, generates an HTML report and returns the GSEA
dotplots in R.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Specify the filepath of the TSV file with the database info</span></span>
<span><span class="va">downloaded_dbs_filepath</span> <span class="op">&lt;-</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"databases.tsv"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load the file</span></span>
<span><span class="va">databases</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.delim</a></span><span class="op">(</span></span>
<span>  <span class="va">downloaded_dbs_filepath</span>,</span>
<span>  sep <span class="op">=</span> <span class="st">"\t"</span>,</span>
<span>  stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Specify the clusterProfiler parameters</span></span>
<span><span class="va">clusterProfiler_params</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  pvalueCutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  pAdjustMethod <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  minGSSize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  maxGSSize <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  qvalueCutoff <span class="op">=</span> <span class="fl">0.2</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">report_dir</span> <span class="op">&lt;-</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span></span>
<span>  <span class="st">"results"</span>,</span>
<span>  <span class="st">"gsea_reports"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The function below runs the clusterProfiler for all clusters and all
levels, and generates the HTML report:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu">SplineOmics</span><span class="fu">::</span><span class="fu"><a href="../reference/run_gsea.html">run_gsea</a></span><span class="op">(</span></span>
<span>  <span class="co"># A dataframe with three columns: feature, cluster, and gene. Feature contains</span></span>
<span>  <span class="co"># the integer index of the feature, cluster the integer specifying the cluster</span></span>
<span>  <span class="co"># number, and gene the string of the gene, such as "CLSTN2".</span></span>
<span>  levels_clustered_hits <span class="op">=</span> <span class="va">clustering_results</span><span class="op">$</span><span class="va">clustered_hits_levels</span>,</span>
<span>  databases <span class="op">=</span> <span class="va">databases</span>,</span>
<span>  clusterProfiler_params <span class="op">=</span> <span class="va">clusterProfiler_params</span>,</span>
<span>  report_info <span class="op">=</span> <span class="va">report_info</span>,</span>
<span>  report_dir <span class="op">=</span> <span class="va">report_dir</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>You can view the generated analysis report of the run_gsea function
<a href="https://drive.google.com/uc?export=view&amp;id=1EvPiEwHyGnBQFxRets00s87tochgyGK4" class="external-link">here</a>.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.3.3 (2024-02-29)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/local/R-4.3.3/lib/R/lib/libRblas.so </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=de_AT.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=de_AT.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=de_AT.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=de_AT.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Europe/Vienna</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices datasets  utils     methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] dplyr_1.1.4       here_1.0.1        readr_2.1.5       SplineOmics_0.1.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RColorBrewer_1.1-3       rstudioapi_0.16.0        jsonlite_1.8.9          </span></span>
<span><span class="co">#&gt;   [4] shape_1.4.6.1            magrittr_2.0.3           farver_2.1.2            </span></span>
<span><span class="co">#&gt;   [7] nloptr_2.1.1             rmarkdown_2.28           GlobalOptions_0.1.2     </span></span>
<span><span class="co">#&gt;  [10] fs_1.6.4                 ragg_1.3.3               vctrs_0.6.5             </span></span>
<span><span class="co">#&gt;  [13] minqa_1.2.8              base64enc_0.1-3          htmltools_0.5.8.1       </span></span>
<span><span class="co">#&gt;  [16] progress_1.2.3           broom_1.0.7              variancePartition_1.32.5</span></span>
<span><span class="co">#&gt;  [19] sass_0.4.9               KernSmooth_2.23-22       bslib_0.8.0             </span></span>
<span><span class="co">#&gt;  [22] htmlwidgets_1.6.4        desc_1.4.3               pbkrtest_0.5.3          </span></span>
<span><span class="co">#&gt;  [25] plyr_1.8.9               cachem_1.1.0             lifecycle_1.0.4         </span></span>
<span><span class="co">#&gt;  [28] iterators_1.0.14         pkgconfig_2.0.3          Matrix_1.6-5            </span></span>
<span><span class="co">#&gt;  [31] R6_2.5.1                 fastmap_1.2.0            rbibutils_2.3           </span></span>
<span><span class="co">#&gt;  [34] clue_0.3-65              digest_0.6.37            numDeriv_2016.8-1.1     </span></span>
<span><span class="co">#&gt;  [37] colorspace_2.1-1         patchwork_1.3.0          S4Vectors_0.40.2        </span></span>
<span><span class="co">#&gt;  [40] rprojroot_2.0.4          textshaping_0.4.0        fansi_1.0.6             </span></span>
<span><span class="co">#&gt;  [43] compiler_4.3.3           withr_3.0.1              bit64_4.5.2             </span></span>
<span><span class="co">#&gt;  [46] aod_1.3.3                doParallel_1.0.17        backports_1.5.0         </span></span>
<span><span class="co">#&gt;  [49] BiocParallel_1.36.0      viridis_0.6.5            dendextend_1.18.0       </span></span>
<span><span class="co">#&gt;  [52] gplots_3.2.0             MASS_7.3-60.0.1          rjson_0.2.23            </span></span>
<span><span class="co">#&gt;  [55] corpcor_1.6.10           gtools_3.9.5             caTools_1.18.3          </span></span>
<span><span class="co">#&gt;  [58] tools_4.3.3              zip_2.3.1                remaCor_0.0.18          </span></span>
<span><span class="co">#&gt;  [61] glue_1.8.0               nlme_3.1-164             grid_4.3.3              </span></span>
<span><span class="co">#&gt;  [64] cluster_2.1.6            reshape2_1.4.4           generics_0.1.3          </span></span>
<span><span class="co">#&gt;  [67] gtable_0.3.5             tzdb_0.4.0               tidyr_1.3.1             </span></span>
<span><span class="co">#&gt;  [70] hms_1.1.3                utf8_1.2.4               BiocGenerics_0.48.1     </span></span>
<span><span class="co">#&gt;  [73] ggrepel_0.9.6            foreach_1.5.2            pillar_1.9.0            </span></span>
<span><span class="co">#&gt;  [76] stringr_1.5.1            vroom_1.6.5              limma_3.58.1            </span></span>
<span><span class="co">#&gt;  [79] circlize_0.4.16          splines_4.3.3            lattice_0.22-5          </span></span>
<span><span class="co">#&gt;  [82] renv_1.0.10              bit_4.5.0                tidyselect_1.2.1        </span></span>
<span><span class="co">#&gt;  [85] locfit_1.5-9.10          ComplexHeatmap_2.18.0    knitr_1.48              </span></span>
<span><span class="co">#&gt;  [88] gridExtra_2.3            IRanges_2.36.0           edgeR_4.0.16            </span></span>
<span><span class="co">#&gt;  [91] svglite_2.1.3            RhpcBLASctl_0.23-42      stats4_4.3.3            </span></span>
<span><span class="co">#&gt;  [94] xfun_0.48                Biobase_2.62.0           statmod_1.5.0           </span></span>
<span><span class="co">#&gt;  [97] matrixStats_1.4.1        pheatmap_1.0.12          stringi_1.8.4           </span></span>
<span><span class="co">#&gt; [100] yaml_2.3.10              boot_1.3-29              evaluate_1.0.0          </span></span>
<span><span class="co">#&gt; [103] codetools_0.2-19         tibble_3.2.1             BiocManager_1.30.25     </span></span>
<span><span class="co">#&gt; [106] cli_3.6.3                systemfonts_1.1.0        Rdpack_2.6.1            </span></span>
<span><span class="co">#&gt; [109] munsell_0.5.1            jquerylib_0.1.4          Rcpp_1.0.13             </span></span>
<span><span class="co">#&gt; [112] EnvStats_3.0.0           png_0.1-8                parallel_4.3.3          </span></span>
<span><span class="co">#&gt; [115] pkgdown_2.1.1            ggplot2_3.5.1            prettyunits_1.2.0       </span></span>
<span><span class="co">#&gt; [118] bitops_1.0-9             lme4_1.1-35.5            viridisLite_0.4.2       </span></span>
<span><span class="co">#&gt; [121] mvtnorm_1.3-1            lmerTest_3.1-3           scales_1.3.0            </span></span>
<span><span class="co">#&gt; [124] openxlsx_4.2.7.1         purrr_1.0.2              crayon_1.5.3            </span></span>
<span><span class="co">#&gt; [127] fANCOVA_0.6-1            GetoptLong_1.0.5         rlang_1.1.4</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Thomas Rauter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
