<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>cluster_hits() — cluster_hits • SplineOmics</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="cluster_hits() — cluster_hits"><meta name="description" content="Performs clustering on hits from top tables generated by differential
expression analysis.
This function filters hits based on adjusted p-value thresholds, extracts
spline coefficients for
significant features, normalizes these coefficients, and applies hierarchical
clustering. The results,
including clustering assignments and normalized spline curves, are saved in a
specified directory and
compiled into an HTML report."><meta property="og:description" content="Performs clustering on hits from top tables generated by differential
expression analysis.
This function filters hits based on adjusted p-value thresholds, extracts
spline coefficients for
significant features, normalizes these coefficients, and applies hierarchical
clustering. The results,
including clustering assignments and normalized spline curves, are saved in a
specified directory and
compiled into an HTML report."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SplineOmics</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/design_limma_design_formula.html">Designing a Limma Design Formula</a></li>
    <li><a class="dropdown-item" href="../articles/Docker_permission_denied.html">Docker_permission_denied</a></li>
    <li><a class="dropdown-item" href="../articles/Docker-instructions.html">Docker-instructions</a></li>
    <li><a class="dropdown-item" href="../articles/get-started.html">get-started</a></li>
    <li><a class="dropdown-item" href="../articles/limma_result_categories.html">limma_result_categories</a></li>
    <li><a class="dropdown-item" href="../articles/RNA-seq%20analysis.html">RNA-seq analysis</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/csbg/SplineOmics/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>cluster_hits()</h1>
      <small class="dont-index">Source: <a href="https://github.com/csbg/SplineOmics/blob/HEAD/R/cluster_hits.R" class="external-link"><code>R/cluster_hits.R</code></a></small>
      <div class="d-none name"><code>cluster_hits.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Performs clustering on hits from top tables generated by differential
expression analysis.
This function filters hits based on adjusted p-value thresholds, extracts
spline coefficients for
significant features, normalizes these coefficients, and applies hierarchical
clustering. The results,
including clustering assignments and normalized spline curves, are saved in a
specified directory and
compiled into an HTML report.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">cluster_hits</span><span class="op">(</span></span>
<span>  <span class="va">splineomics</span>,</span>
<span>  <span class="va">clusters</span>,</span>
<span>  adj_pthresholds <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">)</span>,</span>
<span>  adj_pthresh_avrg_diff_conditions <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  adj_pthresh_interaction_condition_time <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plot_info <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>y_axis_label <span class="op">=</span> <span class="st">"Value"</span>, time_unit <span class="op">=</span> <span class="st">"min"</span>, treatment_labels <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>    treatment_timepoints <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span>,</span>
<span>  plot_options <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>cluster_heatmap_columns <span class="op">=</span> <span class="cn">FALSE</span>, meta_replicate_column <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span>,</span>
<span>  raw_data <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  report_dir <span class="op">=</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  report <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-splineomics">splineomics<a class="anchor" aria-label="anchor" href="#arg-splineomics"></a></dt>
<dd><p>An S3 object of class `SplineOmics` that contains all the
necessary data and parameters for the analysis, including:</p><ul><li><p><code>data</code>: The original expression dataset used for differential
  expression analysis.</p></li>
<li><p><code>meta</code>: A dataframe containing metadata corresponding to the
  <code>data</code>, must include a 'Time' column and any columns specified by
  <code>conditions</code>.</p></li>
<li><p><code>design</code>: A character of length 1 representing the limma
  design formula.</p></li>
<li><p><code>condition</code>: Character of length 1 specifying the column name
  in <code>meta</code> used to define groups for analysis.</p></li>
<li><p><code>spline_params</code>: A list of spline parameters for the analysis.</p></li>
<li><p><code>meta_batch_column</code>: A character string specifying the column
  name in the metadata used for batch effect removal.</p></li>
<li><p><code>meta_batch2_column</code>: A character string specifying the second
  column name in the metadata used for batch effect removal.</p></li>
<li><p><code>limma_splines_result</code>: A list of data frames, each representing
   a top table from differential expression analysis, containing at least
   'adj.P.Val' and expression data columns.</p></li>
</ul></dd>


<dt id="arg-clusters">clusters<a class="anchor" aria-label="anchor" href="#arg-clusters"></a></dt>
<dd><p>Character or integer vector specifying the number of clusters</p></dd>


<dt id="arg-adj-pthresholds">adj_pthresholds<a class="anchor" aria-label="anchor" href="#arg-adj-pthresholds"></a></dt>
<dd><p>Numeric vector of p-value thresholds for filtering
hits in each top table.</p></dd>


<dt id="arg-adj-pthresh-avrg-diff-conditions">adj_pthresh_avrg_diff_conditions<a class="anchor" aria-label="anchor" href="#arg-adj-pthresh-avrg-diff-conditions"></a></dt>
<dd><p>p-value threshold for the results
from the average difference of the condition limma result. Per default 0 (
turned off).</p></dd>


<dt id="arg-adj-pthresh-interaction-condition-time">adj_pthresh_interaction_condition_time<a class="anchor" aria-label="anchor" href="#arg-adj-pthresh-interaction-condition-time"></a></dt>
<dd><p>p-value threshold for the
results from the interaction of condition and time limma result. Per default
0 (turned off).</p></dd>


<dt id="arg-genes">genes<a class="anchor" aria-label="anchor" href="#arg-genes"></a></dt>
<dd><p>A character vector containing the gene names of the features to
be analyzed.</p></dd>


<dt id="arg-plot-info">plot_info<a class="anchor" aria-label="anchor" href="#arg-plot-info"></a></dt>
<dd><p>List containing the elements y_axis_label (string),
time_unit (string), treatment_labels (character vector),
treatment_timepoints (integer vector). All can also be NA.
This list is used to add this info to the spline plots.
time_unit is used to label the x-axis, and treatment_labels
and -timepoints are used to create vertical dashed lines,
indicating the positions of the treatments (such as
feeding, temperature shift, etc.).</p></dd>


<dt id="arg-plot-options">plot_options<a class="anchor" aria-label="anchor" href="#arg-plot-options"></a></dt>
<dd><p>List with specific fields (cluster_heatmap_columns =
Bool) that allow for customization of plotting behavior.</p></dd>


<dt id="arg-raw-data">raw_data<a class="anchor" aria-label="anchor" href="#arg-raw-data"></a></dt>
<dd><p>Optional. Data matrix with the raw (unimputed) data, still
containing NA values. When provided, it highlights the datapoints in the
spline plots that originally where NA and that were imputed.</p></dd>


<dt id="arg-report-dir">report_dir<a class="anchor" aria-label="anchor" href="#arg-report-dir"></a></dt>
<dd><p>Character string specifying the directory path where the
HTML report and any other output files should be saved.</p></dd>


<dt id="arg-report">report<a class="anchor" aria-label="anchor" href="#arg-report"></a></dt>
<dd><p>Boolean TRUE or FALSE value specifing if a report should be
generated.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list where each element corresponds to a group factor and contains
the clustering results,
        including `clustered_hits` data frame, hierarchical clustering object
        `hc`, `curve_values`
        data frame with normalized spline curves, and `top_table` with
        cluster assignments.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="https://rdrr.io/pkg/limma/man/toptable.html" class="external-link">topTable</a></code>, <code><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">hclust</a></code></p></div>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Thomas Rauter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

