<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>explore_data() — explore_data • SplineOmics</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="explore_data() — explore_data"><meta name="description" content="This function automatically generates exploratory data analysis (EDA) plots
from the provided data. These include density plots, boxplots, PCA plots,
MDS plots, variance explained plots, violin plots, mean correlation with
time, first lag autocorrelation, lag1 differences, and coefficient of
variation. The function returns all EDA plots as a list and, by default,
creates an HTML report containing the plots, saving it to the specified
report directory."><meta property="og:description" content="This function automatically generates exploratory data analysis (EDA) plots
from the provided data. These include density plots, boxplots, PCA plots,
MDS plots, variance explained plots, violin plots, mean correlation with
time, first lag autocorrelation, lag1 differences, and coefficient of
variation. The function returns all EDA plots as a list and, by default,
creates an HTML report containing the plots, saving it to the specified
report directory."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SplineOmics</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.7</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/CHANGELOG.html">Changelog</a></li>
    <li><a class="dropdown-item" href="../articles/design_limma_design_formula.html">Designing a Limma Design Formula</a></li>
    <li><a class="dropdown-item" href="../articles/Docker_permission_denied.html">Docker_permission_denied</a></li>
    <li><a class="dropdown-item" href="../articles/Docker-instructions.html">Docker-instructions</a></li>
    <li><a class="dropdown-item" href="../articles/get-started.html">get-started</a></li>
    <li><a class="dropdown-item" href="../articles/glycan-analysis.html">Glycan-analysis</a></li>
    <li><a class="dropdown-item" href="../articles/limma_result_categories.html">limma_result_categories</a></li>
    <li><a class="dropdown-item" href="../articles/methylation_data_analysis.html">methylation-data-analysis</a></li>
    <li><a class="dropdown-item" href="../articles/PVC_analysis.html">PVC_analysis</a></li>
    <li><a class="dropdown-item" href="../articles/RNA-seq-analysis.html">RNA-seq-analysis</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../articles/CHANGELOG.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/csbg/SplineOmics/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>explore_data()</h1>
      <small class="dont-index">Source: <a href="https://github.com/csbg/SplineOmics/blob/HEAD/R/explore_data.R" class="external-link"><code>R/explore_data.R</code></a></small>
      <div class="d-none name"><code>explore_data.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function automatically generates exploratory data analysis (EDA) plots
from the provided data. These include density plots, boxplots, PCA plots,
MDS plots, variance explained plots, violin plots, mean correlation with
time, first lag autocorrelation, lag1 differences, and coefficient of
variation. The function returns all EDA plots as a list and, by default,
creates an HTML report containing the plots, saving it to the specified
report directory.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">explore_data</span><span class="op">(</span><span class="va">splineomics</span>, report_dir <span class="op">=</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="op">)</span>, report <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-splineomics">splineomics<a class="anchor" aria-label="anchor" href="#arg-splineomics"></a></dt>
<dd><p>A named list containing all required inputs for the
splineomics workflow. Must contain the following elements:</p><dl><dt>data</dt>
<dd><p>The data matrix with the values. The columns are the
samples (timepoint + replicate combo) and the rows are the features
(e.g. genes or proteins).</p></dd>

<dt>meta</dt>
<dd><p>A dataframe containing metadata corresponding to the
<code>data</code>, must include a 'Time' column and any columns specified by
<code>conditions</code>. In general, the columns of meta correspond to the
different types of metadata, and each row corresponds to a column of
data (contains the metadata for that sample).</p></dd>

<dt>annotation</dt>
<dd><p>A dataframe that maps the rows of <code>data</code> to
annotation info, such as the gene name or database identifiers.</p></dd>

<dt>report_info</dt>
<dd><p>A named list describing the experiment.
Must include the following fields:
- <code>"omics_data_type"</code>
- <code>"data_description"</code>
- <code>"data_collection_date"</code>
- <code>"analyst_name"</code>
- <code>"contact_info"</code>
- <code>"project_name"</code></p>
<p>May also include the following optional fields:
- <code>"method_description"</code>
- <code>"results_summary"</code>
- <code>"conclusions"</code></p></dd>

<dt>condition</dt>
<dd><p><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a>
Character vector of length 1 specifying the column
name in <code>meta</code> used to define groups for analysis.</p></dd>

<dt>meta_batch_column</dt>
<dd><p><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a>
Character vector of length 1 specifying the column
name in <code>meta</code> that contains the info for the batch effect.</p></dd>

<dt>meta_batch2_column</dt>
<dd><p><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a>
Character vector of length 1 specifying the column
name in <code>meta</code> that contains the info for the second batch
effect.</p></dd>


</dl></dd>


<dt id="arg-report-dir">report_dir<a class="anchor" aria-label="anchor" href="#arg-report-dir"></a></dt>
<dd><p>A non-empty string specifying the report directory. The
output HTML reports will be placed there. Default is the
current working dir, determined with the here library:
here::here().</p></dd>


<dt id="arg-report">report<a class="anchor" aria-label="anchor" href="#arg-report"></a></dt>
<dd><p>A Boolean TRUE or FALSE value, specifying if a report should
be generated or not.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list of ggplot objects representing various exploratory plots.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Temporary output dir for the HTML report</span></span></span>
<span class="r-in"><span><span class="va">report_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"explore_data_demo"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="va">report_dir</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">report_dir</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## --- Load example inputs from inst/extdata ---</span></span></span>
<span class="r-in"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">xzfile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="st">"extdata"</span>, <span class="st">"proteomics_data.rds.xz"</span>,</span></span>
<span class="r-in"><span>    package <span class="op">=</span> <span class="st">"SplineOmics"</span></span></span>
<span class="r-in"><span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"proteomics_meta.csv"</span>, package <span class="op">=</span> <span class="st">"SplineOmics"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Derive the annotation table from the data frame (as in the vignette)</span></span></span>
<span class="r-in"><span><span class="va">first_na_col</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">annotation</span> <span class="op">&lt;-</span> <span class="va">data</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">(</span><span class="va">first_na_col</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu">SplineOmics</span><span class="fu">::</span><span class="fu"><a href="extract_data.html">extract_data</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="co"># The dataframe with the numbers on the left and info on the right.</span></span></span>
<span class="r-in"><span>    data <span class="op">=</span> <span class="va">data</span>,</span></span>
<span class="r-in"><span>    <span class="co"># Use this annotation column for the feature names.</span></span></span>
<span class="r-in"><span>    feature_name_columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gene_name"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    top_row <span class="op">=</span> <span class="fl">4</span>,</span></span>
<span class="r-in"><span>    bottom_row <span class="op">=</span> <span class="fl">1165</span>,</span></span>
<span class="r-in"><span>    right_col <span class="op">=</span> <span class="fl">37</span>,</span></span>
<span class="r-in"><span>    left_col <span class="op">=</span> <span class="fl">2</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Minimal report metadata</span></span></span>
<span class="r-in"><span><span class="va">report_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    omics_data_type <span class="op">=</span> <span class="st">"PTX"</span>,</span></span>
<span class="r-in"><span>    data_description <span class="op">=</span> <span class="st">"Demo proteomics dataset from extdata"</span>,</span></span>
<span class="r-in"><span>    data_collection_date <span class="op">=</span> <span class="st">"2024"</span>,</span></span>
<span class="r-in"><span>    analyst_name <span class="op">=</span> <span class="st">"Example Analyst"</span>,</span></span>
<span class="r-in"><span>    contact_info <span class="op">=</span> <span class="st">"analyst@example.org"</span>,</span></span>
<span class="r-in"><span>    project_name <span class="op">=</span> <span class="st">"DemoProject"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Build the SplineOmics input object</span></span></span>
<span class="r-in"><span><span class="va">splineomics</span> <span class="op">&lt;-</span> <span class="fu">SplineOmics</span><span class="fu">::</span><span class="fu"><a href="create_splineomics.html">create_splineomics</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    data               <span class="op">=</span> <span class="va">data</span>,</span></span>
<span class="r-in"><span>    meta               <span class="op">=</span> <span class="va">meta</span>,</span></span>
<span class="r-in"><span>    annotation         <span class="op">=</span> <span class="va">annotation</span>,</span></span>
<span class="r-in"><span>    report_info        <span class="op">=</span> <span class="va">report_info</span>,</span></span>
<span class="r-in"><span>    condition          <span class="op">=</span> <span class="st">"Phase"</span>, <span class="co"># column in `meta` defining groups</span></span></span>
<span class="r-in"><span>    meta_batch_column  <span class="op">=</span> <span class="st">"Reactor"</span> <span class="co"># optional: for batch-effect removal</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Run EDA and write the HTML report to the temp dir</span></span></span>
<span class="r-in"><span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu">SplineOmics</span><span class="fu">::</span><span class="fu">explore_data</span><span class="op">(</span></span></span>
<span class="r-in"><span>    splineomics <span class="op">=</span> <span class="va">splineomics</span>,</span></span>
<span class="r-in"><span>    report_dir  <span class="op">=</span> <span class="va">report_dir</span>,</span></span>
<span class="r-in"><span>    report      <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Making density plots...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Making violin plots...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Making PCA plots...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Making MDS plots...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Making correlation heatmaps...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Subsampled to top 1000 most variable features (after filtering rows with &gt; 20% missing) for correlation heatmap.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Making mean correlation with time plots...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Making lag1 differences plots...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Making first lag auto-correlation with time plots...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Making cv plots...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Making density plots...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Making violin plots...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Making PCA plots...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Making MDS plots...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Making correlation heatmaps...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Subsampled to top 1000 most variable features (after filtering rows with &gt; 20% missing) for correlation heatmap.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Making mean correlation with time plots...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Making lag1 differences plots...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Making first lag auto-correlation with time plots...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Making cv plots...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  <span style="color: #00BB00;">Info</span> Exploratory data analysis completed successfully.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  Your HTML reports are located in the directory:  /tmp/RtmpKGGksk/explore_data_demo .</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  Please note that due to embedded files, the reports might be flagged as</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  harmful by other software. Rest assured that they provide no harm.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Inspect what was written</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">report_dir</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "explore_batch_corrected_data_PTX_21_10_2025-14_35_50.html"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2] "explore_data_PTX_21_10_2025-14_35_50.html"                </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># `plots` is a named list of ggplot objects (e.g., plots$raw_data$pca, etc.)</span></span></span>
<span class="r-in"><span><span class="co"># print(plots$raw_data$pca)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Thomas Rauter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

