<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Download gene set annotations from Bioconductor organism databases — extract_gene_sets • SplineOmics</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Download gene set annotations from Bioconductor organism databases — extract_gene_sets"><meta name="description" content="This function extracts gene-to-ontology mappings from a specified
Bioconductor organism annotation package (e.g., org.Hs.eg.db,
org.Mm.eg.db) and saves the gene sets to a .tsv file in a
standardized format. The output includes mappings for Gene Ontology
(GO) Biological Process (BP), Molecular Function (MF), Cellular
Component (CC), and KEGG pathways. The resulting file can be used
directly with enrichment functions such as clusterProfiler::enricher()
with TERM2GENE."><meta property="og:description" content="This function extracts gene-to-ontology mappings from a specified
Bioconductor organism annotation package (e.g., org.Hs.eg.db,
org.Mm.eg.db) and saves the gene sets to a .tsv file in a
standardized format. The output includes mappings for Gene Ontology
(GO) Biological Process (BP), Molecular Function (MF), Cellular
Component (CC), and KEGG pathways. The resulting file can be used
directly with enrichment functions such as clusterProfiler::enricher()
with TERM2GENE."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SplineOmics</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/get_started.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Further use cases</h6></li>
    <li><a class="dropdown-item" href="../articles/RNA-seq-analysis.html">RNA-seq analysis</a></li>
    <li><a class="dropdown-item" href="../articles/glycan-analysis.html">Glycan analysis</a></li>
    <li><a class="dropdown-item" href="../articles/methylation_data_analysis.html">Epigenetics analysis</a></li>
    <li><a class="dropdown-item" href="../articles/PVC_analysis.html">Peak-Valley-Cliff (PVC) analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Gene_centric_multiomics_clustering.html">Gene-centric multiomics clustering</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Guides</h6></li>
    <li><a class="dropdown-item" href="../articles/limma_result_categories.html">limma result categories</a></li>
    <li><a class="dropdown-item" href="../articles/design_limma_design_formula.html">Designing a limma design formula</a></li>
    <li><a class="dropdown-item" href="../articles/Docker-instructions.html">Docker container instructions</a></li>
    <li><a class="dropdown-item" href="../articles/Docker_permission_denied.html">Docker permission denied</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../articles/faq.html">FAQ</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/CHANGELOG.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/csbg/SplineOmics/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Download gene set annotations from Bioconductor organism databases</h1>
      <small class="dont-index">Source: <a href="https://github.com/csbg/SplineOmics/blob/HEAD/R/extract_gene_sets.R" class="external-link"><code>R/extract_gene_sets.R</code></a></small>
      <div class="d-none name"><code>extract_gene_sets.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function extracts gene-to-ontology mappings from a specified
Bioconductor organism annotation package (e.g., <code>org.Hs.eg.db</code>,
<code>org.Mm.eg.db</code>) and saves the gene sets to a <code>.tsv</code> file in a
standardized format. The output includes mappings for Gene Ontology
(GO) Biological Process (BP), Molecular Function (MF), Cellular
Component (CC), and KEGG pathways. The resulting file can be used
directly with enrichment functions such as <code><a href="https://rdrr.io/pkg/clusterProfiler/man/enricher.html" class="external-link">clusterProfiler::enricher()</a></code>
with <code>TERM2GENE</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">extract_gene_sets</span><span class="op">(</span></span>
<span>  organism_db <span class="op">=</span> <span class="st">"org.Hs.eg.db"</span>,</span>
<span>  output_dir <span class="op">=</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  filename <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-organism-db">organism_db<a class="anchor" aria-label="anchor" href="#arg-organism-db"></a></dt>
<dd><p><code>character(1)</code>: A string specifying the Bioconductor
organism annotation database to use (e.g., <code>"org.Hs.eg.db"</code> for human or
<code>"org.Mm.eg.db"</code> for mouse).</p></dd>


<dt id="arg-output-dir">output_dir<a class="anchor" aria-label="anchor" href="#arg-output-dir"></a></dt>
<dd><p><code>character(1)</code>: A string specifying the output directory
where the <code>.tsv</code> file will be saved. Defaults to the current project
directory as defined by <code><a href="https://here.r-lib.org/reference/here.html" class="external-link">here::here()</a></code>.</p></dd>


<dt id="arg-filename">filename<a class="anchor" aria-label="anchor" href="#arg-filename"></a></dt>
<dd><p><code>character(1)</code> | <code>NULL</code>: An optional string specifying the
filename for the output file. If <code>NULL</code> (default), a filename is generated
automatically with a timestamp.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>data.frame</code> of gene set annotations with three columns:</p><dl><dt>DB</dt>
<dd><p>Ontology/database source, e.g. <code>"GO_BP"</code>, <code>"GO_MF"</code>, <code>"GO_CC"</code>,
or <code>"KEGG"</code> (if available).</p></dd>

<dt>Geneset</dt>
<dd><p>Ontology term ID or pathway ID (e.g. GO ID, KEGG ID).</p></dd>

<dt>Gene</dt>
<dd><p>Gene symbol (<code>SYMBOL</code>).</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The TSV has three columns:</p><dl><dt>DB</dt>
<dd><p>Ontology/database source, e.g., <code>"GO_BP"</code>, <code>"GO_MF"</code>,
<code>"GO_CC"</code>, or <code>"KEGG"</code> (if available).</p></dd>

<dt>Geneset</dt>
<dd><p>Ontology term ID or pathway ID (e.g., GO ID, KEGG ID).</p></dd>

<dt>Gene</dt>
<dd><p>Gene symbol (<code>SYMBOL</code>).</p></dd>


</dl><p>Note: Some <code>org.*.eg.db</code> packages no longer include KEGG mappings; in
such cases the KEGG section will be empty.</p>
<p>In addition to returning the <code>data.frame</code>, the function also writes the same
table to disk as a <code>.tsv</code> file in the specified <code>output_dir</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Minimal real example (runs only if org package is installed)</span></span></span>
<span class="r-in"><span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"org.Mm.eg.db"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">&amp;&amp;</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"AnnotationDbi"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>    <span class="va">gs</span> <span class="op">&lt;-</span> <span class="fu">extract_gene_sets</span><span class="op">(</span></span></span>
<span class="r-in"><span>        organism_db <span class="op">=</span> <span class="st">"org.Mm.eg.db"</span>,</span></span>
<span class="r-in"><span>        output_dir  <span class="op">=</span> <span class="va">tmp</span>,</span></span>
<span class="r-in"><span>        filename    <span class="op">=</span> <span class="st">"mm_genesets.tsv"</span></span></span>
<span class="r-in"><span>    <span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gs</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="co"># The file path:</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">tmp</span>, <span class="st">"mm_genesets.tsv"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'select()' returned 1:many mapping between keys and columns</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'select()' returned 1:many mapping between keys and columns</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Download complete! The file has been saved as: /tmp/Rtmpy5hVfh/mm_genesets.tsv</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "/tmp/Rtmpy5hVfh/mm_genesets.tsv"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># If the organism package is not installed, you can still see the TSV format:</span></span></span>
<span class="r-in"><span><span class="va">tiny</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    DB <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GO_BP"</span>, <span class="st">"GO_MF"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    Geneset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GO:0008150"</span>, <span class="st">"GO:0003674"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    Gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Trp53"</span>, <span class="st">"Egfr"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="va">tiny</span>,</span></span>
<span class="r-in"><span>    file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">tmp</span>, <span class="st">"example_genesets.tsv"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    sep <span class="op">=</span> <span class="st">"\t"</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span>, col.names <span class="op">=</span> <span class="cn">TRUE</span>, quote <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Thomas Rauter.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

